// app/event/event-page.js
/*==========================================================================*\
  Event page with a list of categories 
\*==========================================================================*/
"use strict";

var BasePage = require('app/base/base-page');
var CategoriesView = require('app/category/categories');
var ProductsView = require('app/product/products');
var SalesView = require('app/sale/sales-info');

var app = require('app');
var assign = require('lodash.assign');

module.exports = BasePage.extend({
  dictionary: 'event', // localization dictionary 
  pageTitle: '',
  pageType: 'event',
  
  template: require('app/templates').event['event-page'],
  
  props: {
    eventId: ['number', true, 0],
    noPhotos: ['boolean', true, false],
    // hasProducts: ['boolean', true, false],
  },
  
  events: {
    // 'click .sorting a': 'sort',
    'click .load-more > .label': 'loadMore',
    'click a.contact-us': 'showContactForm',
  },
  
  bindings: {
    'model.name': '[data-hook=event-name]',
    // hasProducts: {
    //   type: 'booleanClass',
    //   name: 'has-products',
    // },
    noPhotos: {
      type: 'booleanClass',
      name: 'no-photos',
    },
  },
  
  subviews: assign({}, BasePage.prototype.subviews, {
    categories: {
      container: '.categories',
      waitFor: 'model.name',
      prepareView: function (el) {
        
        this.categories = new CategoriesView({
          el: el,
          parent: this,
          collection: this.model.categories,
        });
        
        this.triggerReady();
        return this.categories;
      },
    },
    products: {
      container: '.products',
      waitFor: 'model.complete',
      prepareView: function (el) {
        if (this.model.products_list.length) {
          this.el.classList.add('has-products');
        }
        
        
        this.products = new ProductsView({
          el: el,
          parent: this,
          collection: this.model.products_list, 
        });
        return this.products;
      },
    },
    sales: {
      container: '.sales-info',
      waitFor: 'model.name',
      prepareView: function (el) {
        this.sales = new SalesView({
          el: el,
          parent: this,
          model: this.model.sales,
        });
        return this.sales;
      },
    },
  }),
  
  initialize: function (ops) {
    var self = this;
    BasePage.prototype.initialize.apply(this, arguments);
    ops = ops || {};
    
    this.eventId = this.model.id;
    
    if (!this.model.locked) {
      if (!this.model.name || !this.model.complete) {
        this.model.fetch();
      } 
      else {
        this.triggerReady();
      }
    }
    
    if (this.model.complete) this.checkForPhotos();
    else this.listenToOnce(this.model, 'change:complete', this.checkForPhotos);
    
    this.setTitle(this.model.name);
    this.listenTo(this.model, 'change:name', function () {
    //  if (this.model.categories.length === 1 && !this.model.products.length) {
        this.setTitle(this.model.name);
    //  }
    });
    
    this.render();
  },
  
  checkForPhotos: function () {
    var self = this;
    
    var noPhotos = true;
    for (var i = 0; i < this.model.categories.models.length; i++) {
      if (this.model.categories.models[i].photos.models.length) {
        noPhotos = false;
      }
    }
    
    app.ready(function () {
      if (app.config.empty_message) {
        var message = app.config.empty_message;
        message = message.replace('[[EVENT_NAME]]', self.model.name);
        message = message.replace('[[CONTACT_US]]', 
            '<a href="javascript:void(0);" class="contact-us">contact us</a>');
        
        self.query('.event-empty-message h2').innerHTML = message;
      }
    });
    
    this.noPhotos = noPhotos;
  },
  
  showContactForm: function () {
    var nav = app.view.nav;
    if (!nav.menu) nav.initMenu();
    nav.menu.showContacts();
  },
});
