// app/event/event-nav.js
/*==========================================================================*\
  Fixed header navbar for an event 
\*==========================================================================*/
"use strict";

var app = require('app');
var NavContents = require('app/nav/nav-contents');

module.exports = NavContents.extend({
  template: require('app/templates').event['event-nav'],
  
  props: {
    cartEmpty: ['boolean', true, true],
  },
  
  initialize: function () {
    NavContents.prototype.initialize.apply(this, arguments);
    
    this.listenTo(this.model.cart, 'change:is_empty', this.updateIcons);
    this.listenTo(this.model.favorites, 'change:isEmpty', this.updateIcons);
    this.listenTo(app.router, 'page', this.updateIcons);
  },
  
  switchNav: function (newView) { // show a nav view 
    this.viewSwitcher.set(newView);
  },
  
  onLinkClick: function (e) {
    e.stopPropagation();
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.cacheElements({
      favsEl: '.favs',
      cartEl: '.cart',
      menuEl: '.menu',
    });
    this.updateIcons();
  },
  
  updateIcons: function () {
    this.cartEl.classList.toggle('filled', !this.model.cart.is_empty);
    this.favsEl.classList.toggle('filled', !this.model.favorites.isEmpty);
    
    var page = app.router.currentPage;
    if (page) {
      this.cartEl.classList.toggle('current', page === 'cart');
      this.favsEl.classList.toggle('current', page === 'favorites');
    }
  },
});
