// app/cart/photo/option/cart-option-model.js
/*==========================================================================*\
  A photo's purchase option (size) in cart 
\*==========================================================================*/
"use strict";

var Addon = require('../addon/cart-addon-model');
var BaseModel = require('app/base/base-model');
var BaseCollection = require('app/base/base-collection');
var CropModel = require('./crop-model');

var app = require('app');

// var Addon = BaseModel.extend({
//   idAttribute: 'addon_id',
//   props: {
//     id: 'number', // ID of the addon
//     oid: 'number', // selected option ID
//   },
//   derived: {
//     addon: {
//       fn: function () {
//         return app.addons.get(this.id);
//       },
//     },
//   },
// });

var Addons = BaseCollection.extend({
  model: Addon,
});

module.exports = BaseModel.extend({
  idAttribute: 'size_id',
  props: {
    size_id: 'number',
    qty: 'number',
    // crop: 'object',
    // addons: 'object',
  },
  
  children: {
    crop: CropModel,
  },
  
  collections: {
    addons: Addons,
  },
  
  session: {
    price: ['number', true, 0],
    base_price: ['number', true, 0],
    discounted: ['boolean', true, false],
    bulkPricingActive: ['boolean', true, false],
  },
  
  derived: {
    size: {
      fn: function () {
        return app.sizes.get(this.size_id);
      },
    },
    photo: {
      fn: function () { return this.cart_photo.photo; }
    },
    photo_option: {
      fn: function () {
        return this.photo.buyOptions.get(this.size_id +'-'+ 
                                         this.cart_photo.color_id);
      },
    },
    
    // discounted: {
    //   deps: ['price'],
    //   fn: function () {
    //     var prices = this.photo_option.getPrices(
    //                                           this.size.no_qty ? 1 : this.qty);
    //     return prices.baseTotal !== prices.total;
    //   },
    // },
    addons_total: {
      cache: false,
      fn: function () {
        var total = 0;
        var qty = this.size.no_qty ? 1 : this.qty;
        for (var i = 0; i < this.addons.length; i++) {
          total += this.addons.at(i).price * qty;
        }
        return total;
      },
    },
    discounted_price: {
      deps: ['qty', 'addons'],
      cache: false,
      fn: function () {
        // var photo = this.cart_photo.photo;
        // var option = photo.buyOptions.get(this.size_id);
        var option = this.photo_option;
        if (!option) return 0; // product option
        // var singlePrice = option.price;
        var singlePrice = option.discounted_price;
        for (var i = 0; i < this.addons.length; i++) {
          singlePrice += this.addons.at(i).price;
        }
        return singlePrice * (this.size.no_qty ? 1 : this.qty);
      },
    },
    
    cart_photo: {
      fn: function () {
        return this.collection.parent;
      },
    },
    
    cart: {
      fn: function () {
        return this.cart_photo.cart;
      },
    },
    
    shipping_price: {
      deps: ['price'],
      cache: false,
      fn: function () {
        if (!this.photo_option) return 0; // product option
        var shipping_item = this.size.shipping.get(this.cart.shipping_type);
        return shipping_item ? shipping_item.price : 0;
      },
    },
    
    bulkPricingAvailable: {
      deps: ['photo_option'],
      fn: function () {
        if (!this.photo_option) return false; // product option
        return this.photo_option && !!this.photo_option.bulk_pricing.length;
      },
    },
  },
  
  initialize: function () {
    this.listenTo(this.addons, 'all', this.updatePrice);
    this.listenTo(this, 'change:qty', this.updatePrice);
  },
  
  updatePrice: function () {
    if (this.collection.isFromProductPhoto) return 0;
    var option = this.photo_option;
    if (!option) { // this is a product option
      return 0;
    }
    var qty = this.size.no_qty ? 1 : this.qty;
    var prices = option.getPrices(qty);
    this.prices = prices;
    // var singlePrice = option.price;
    var addonsTotal = this.addons_total;
    this.discounted = prices.total !== prices.baseTotal;
    this.base_price = prices.baseTotal + addonsTotal;
    this.bulkPricingActive = !!prices.activeBulkQty;
    this.price = prices.total + addonsTotal;

    if (this.cart) {
      this.cart.updateSummary();
      this.cart.save();
    }
  },
  
  priceForOption: function (oid) {
    var option = this.options[oid];
  },
  
  // parse: function (res) {
  // },
  
});
