// app/photo/photo.js
/*==========================================================================*\
  A single photo with its UI elements in the lightbox
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var BuyForm = require('./form/photo-form');
var ShareView = require('app/social/social');

var app = require('app');
var keyboard = require('app/util/keyboard/keyboard');
var popup = require('app/popup/popup');
var ua = require('app/util/ua/ua');

module.exports = BaseView.extend({
  template: require('app/templates').photo.photo,
  loaded: false, // whether the main image has been loaded 
  failures: 0, // number of times main image request has failed 
  imageWidth: 0, // dimensions of an image as currently displayed 
  imageHeight: 0,
  
  props: {
    buyOnly: ['boolean', true, false], // only show buy form 
    cropOnly: 'any', // only show crop interface
    
    // whether buy form is to be displayed on next to a photo, not on top
    buyFormInline: ['boolean', true, true],
    
    // whether buy form should span all of viewport without margins 
    buyFormFullscreen: ['boolean', true, true],
    
    // Prev and Next click areas get pointer-events:none;
    disabledPrevNext: ['boolean', true, false],
    
    // inCart: ['boolean', true, false],
    cropDisplayed: ['boolean', true, false],
    croppingMode: ['boolean', true, false],
    cropResize: ['boolean', true, false],
    currentCrop: 'object',
    cropCanRotate: ['boolean', true, false],
    croppingAllowed: ['boolean', true, true],
    
    cropSize: ['string', true, ''],
    
    currentVariant: 'string',
    
    nameDisplayed: ['boolean', true, false],
  },
  
  derived: {
    lightbox: {
      fn: function () {
        return this.getParent('lightbox');
      },
    },
  },
  
  bindings: {
    'model.locked': {
      type: 'booleanClass',
      name: 'locked',
    },
    
    'model.id': {
      type: function (el, val) {
        this.el.setAttribute("data-id", val);
      },
    },
    
    nameDisplayed: {
      type: 'booleanClass',
      name: 'name-displayed',
    },
    buyOnly: {
      type: 'booleanClass',
      name: 'buy-only',
    },
    cropSize: {
      type: function (el, val) {
        el.innerHTML = val;
      },
      selector: '.crop-controls .size',
    },
    croppingAllowed: {
      // type: function (el, val) {
      //   this.el.classList.toggle('cropping-disabled', !val);
      // },
      type: 'booleanClass',
      name: 'cropping-allowed',
    },
    // buyFormInline: {
    //   type: 'booleanClass',
    //   name: 'buy-form-inline',
    // },
    // buyFormFullscreen: {
    //   type: 'booleanClass',
    //   name: 'buy-form-fullscreen',
    // },
    croppingMode: {
      type: 'booleanClass',
      name: 'cropping-mode-pending',
    },
    cropResize: {
      type: 'booleanClass',
      name: 'cropping-resize',
    },
    cropDisplayed: {
      type: 'booleanClass',
      name: 'cropping-areas',
    },
    cropCanRotate: {
      type: 'booleanClass',
      name: 'cropping-rotate',
    },
    
    currentVariant: {
      type: 'class',
      // type: function (el, val) {
      //   if (val) this.imgEl.src = this.model.getUrl(val);
      // },
    },
    
    // inCart: {
    'model.in_cart': {
      type: function (el, val) {
        this.el.classList.toggle('in-cart', val);
        this.query('.action.cart .tip').innerHTML = 
            this.lz(val ? 'In Cart' : 'Add To Cart');
      },
      // type: 'booleanClass',
      // name: 'in-cart',
    },
    'model.in_favorites': {
      type: function (el, val) {
        this.el.classList.toggle('in-favorites', val);
        this.query('.action.fav .tip').innerHTML = 
            this.lz(val ? 'In Favorites' : 'Add To Favorites');
      },
    },
    'model.hidden': {
      type: function (el, val) {
        this.el.classList.toggle('hidden', val);
        this.query('.action.hide .tip').innerHTML = 
            this.lz(val ? 'Hidden' : 'Hide');
      },
    },
    'model.current_color_id': {
      type: function () {
        this.currentVariant = 
            app.colors.get(this.model.current_color_id).prefix;
      },
    }
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.bindMethods('exitCropping', 'onImageTransitionEnd');
    
    this.croppingAllowed = app.config.custom_crop !== 'view' &&
                           app.config.custom_crop !== 'none';
                           
    this.nameDisplayed = !!this.model.name && !!this.parent.namesDisplayed;
    this.listenTo(this.parent, 'change:namesDisplayed', function () {
      this.nameDisplayed = !!this.model.name && !!this.parent.namesDisplayed;
      this.resize();
    });

    this.load();
    this.render();
    this.cacheElements({
      imagesEl: '.images',
      img: '.img',
      imgEl: 'img',
      actionsEl: '.actions',
      shareMenuEl: '.share-menu',
      
      cropCoverTop: '.crop-cover.top',
      cropCoverRight: '.crop-cover.right',
      cropCoverBottom: '.crop-cover.bottom',
      cropCoverLeft: '.crop-cover.left',
      
      cropHandleTopLeft: '.crop-handle.top-left',
      cropHandleTopRight: '.crop-handle.top-right',
      cropHandleBottomLeft: '.crop-handle.bottom-left',
      cropHandleBottomRight: '.crop-handle.bottom-right',
    });
    this.currentVariant = 'original';
    this.img.addEventListener(ua.events.transitionEnd, 
                              this.onImageTransitionEnd);
    this.resize();
    
    // this.inCart = !!this.model.event.cart.photos.get(this.model.id, 'pid');
    var params = app.history.latestConfig;
    
    if (params.buy) {
      this.buyOnly = true;
      this.after(1, function () {
        this.showBuyForm();
        this.onImageTransitionEnd();
      });
    }
    if (params.crop) {
      this.buyOnly = true;
      this.cropOnly = true;
      this.after(1, function () {
        this.showBuyForm();
        this.onImageTransitionEnd();
        
        // var option;
        // if (typeof ops.crop === 'number') { // separate purchase option
        //   option = this.model.cart.photos.get(this.model.id)
        //            .options.get(ops.crop)
        // }
        // else { // a photo added to a product 
        // }
      });
    }

    this.shareView = new ShareView({
      el: this.shareMenuEl,
      parent: this,
      model: this.model
    });
  },
  
  events: {
    // click: 'onClick',
    'click .action.cart': 'showBuyForm',
    'click .action.fav': 'toggleFav',
    'click .action.hide': 'toggleHide',
    
    'mousedown .img .crop-covers': 'onDragMouseDown',
    'mousedown .img .crop-handle': 'onResizeMouseDown',
    
    'touchstart .img .crop-covers': 'onDragMouseDown',
    'touchstart .img .crop-handle': 'onResizeMouseDown',
    
    'click .crop-controls .button': 'onCropButtonClick',
    'click .crop-tips .got-it': 'hideCropTips',
    
    'wheel .img .crop-covers': 'onMouseWheel',
    
    'contextmenu .img': 'onContextMenu',
    
    'click .variant-name': 'selectVariant',
    'click .variants-overlay .variant-button': 'selectVariant',

    // 'click .crop-controls .reset': 'resetCrop',
    // 'click .crop-controls .rotate': 'rotateCrop',
    // 'click .crop-controls .cancel': 'cancelCrop',
  },
  
  // onClick: function (e) { // hide lightbox on tapping the photo on mobile 
  //   if (!e || !e.target) return;
  //   if (!ua.mobile || this.croppingMode) return;
    
  //   if (e.target.closest('.actions')) return;
  //   if (e.target.closest('.title')) return;
    
  //   this.lightbox.hide();
  // },
  
  selectVariant: function (e) {
    var id = +e.target.getAttribute('data-color-id');
    if (!id) id = app.colors.get('original', 'prefix').id;
    this.displayVariant(id);
  },
  displayVariant: function (id) {
    this.model.current_color_id = id;
    // this.currentVariant = app.colors.get(id).prefix;
  },
  loadVariants: function () {
    var black = false;
    var sepia = false;
    for (var id in this.model.pricing.colors) {
      if (app.colors.get(id).prefix === 'black') black = true;
      if (app.colors.get(id).prefix === 'sepia') sepia = true;
    }
    if (black) this.query('img.black').src = this.model.getUrl('black');
    if (sepia) this.query('img.sepia').src = this.model.getUrl('sepia');
  },
  
  disablePrevNext: function () {
    this.disabledPrevNext = true;
  },
  enablePrevNext: function () {
    this.disabledPrevNext = false;
  },
  
  enterCropping: function (crop) {
    this.currentCrop = crop;
    
    this.cropCanRotate = crop.canRotate;
    this.cropResize = crop.minWidth <= 90;
    this.croppingMode = true;
    this.applyCrop(crop);
    this.showCrop();
    
    document.body.addEventListener('mousedown', this, false);
    document.body.addEventListener('touchstart', this, false);
    
    keyboard.bind('esc', this.exitCropping);
    keyboard.bind('enter', this.exitCropping);
  },
  exitCropping: function () {
    this.currentCrop = null;
    this.cropResize = false;
    this.croppingMode = false;
    this.hideCrop();
    
    document.body.removeEventListener('mousedown', this, false);
    document.body.removeEventListener('touchstart', this, false);
    
    if (this.cropOnly) {
      this.buyForm.save();
    }
  },
  
  glimpseCropping: function (crop) {
    console.log('glimpse');
    this.croppingMode = true;
    this.applyCrop(null, crop);
    this.showCrop();
  },
  
  applyCrop: function (newCrop, tmpCrop) {
    if (newCrop) this.currentCrop = newCrop;
    var crop = tmpCrop || this.currentCrop;
    // if (!crop) return;
    // console.log('applying crop', crop.top, 
    //             crop.bottom, crop.left, crop.right);
    // console.log(crop.ratio, this.model.ratio);
    
    var top = Math.round(this.imageHeight * crop.top / 100);
    var right = Math.round(this.imageWidth * crop.right / 100);
    var bottom = Math.round(this.imageHeight * crop.bottom / 100);
    var left = Math.round(this.imageWidth * crop.left / 100);
    // if (top <= 1) top = 0;
    // if (right <= 1) right = 0;
    // if (bottom <= 1) bottom = 0;
    // if (left <= 1) left = 0;
    
    this.cropCoverTop.style[ua.css.transform] = 
        'translate3d('+ left +'px,'+ top +'px,0)';
    this.cropCoverRight.style[ua.css.transform] = 
        'translate3d('+ (-right) +'px,'+ top +'px,0)';
    this.cropCoverBottom.style[ua.css.transform] = 
        'translate3d('+ (-right) +'px,'+ (-bottom) +'px,0)';
    this.cropCoverLeft.style[ua.css.transform] = 
        'translate3d('+ left +'px,'+ (-bottom) +'px,0)';
        
    this.cropHandleTopLeft.style[ua.css.transform] = 
        'translate3d('+ left +'px,'+ top +'px,0)';
    this.cropHandleTopRight.style[ua.css.transform] = 
        'translate3d('+ (-right) +'px,'+ top +'px,0)';
    this.cropHandleBottomLeft.style[ua.css.transform] = 
        'translate3d('+ left +'px,'+ (-bottom) +'px,0)';
    this.cropHandleBottomRight.style[ua.css.transform] = 
        'translate3d('+ (-right) +'px,'+ (-bottom) +'px,0)';
    
    this.cropSize = crop.sizeStr;
  },
  
  showCrop: function () {
    if (this.buyOnly) {
      this.el.classList.remove('buy-only');
    }
    this.cropDisplayed = true;
    // this.resize();
    
    this.resize({ notReally: !this.cropOnly });
    if (!this.buyFormInline) {
      var imgRect = this.img.getBoundingClientRect();
      var newLeft = (app.state.width - this.imageWidth) / 2;
      var newTop = (app.state.height - this.imageHeight) / 2;
      
      var scaleX = this.imageWidth / imgRect.width;
      var scaleY = this.imageHeight / imgRect.height;
      
      if (!this.cropOnly) this.img.classList.remove('no-transition');
      this.img.style[ua.css.transform] = 
          'translate('+ (newLeft - imgRect.left) +'px, '+ 
          (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
    }
  },
  hideCrop: function () {
    this.cropDisplayed = false;
    if (this.cropOnly) return;
    // this.resize();
    
    this.resize({ notReally: true });
    if (!this.buyFormInline) {

      var imgRect = this.img.getBoundingClientRect();
      var titleOffset = !app.state.fullscreen && !ua.mobile && this.nameDisplayed
          ? 54 : 0;
      var newLeft = (app.state.width - this.imageWidth) / 2;
      var newTop = (app.state.height - this.imageHeight - titleOffset) / 2;
      // var newTop = (app.state.height - this.imageHeight) / 2;
      
      var scaleX = this.imageWidth / imgRect.width;
      var scaleY = this.imageHeight / imgRect.height;
      
      if (!this.cropOnly) this.img.classList.remove('no-transition');
      this.img.style[ua.css.transform] = 
          'translate('+ (newLeft - imgRect.left) +'px, '+ 
          (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
    }
  },
  
  hideCropTips: function () {
    app.storage.cropTipUsed = '1';
    this.query('.crop-tips').classList.add('hidden');
  },
  
  onContextMenu: function (e) {
    // if (this.croppingMode) {
      e.preventDefault();
      return false;
    // }
  },
  onCropButtonClick: function (e) {
    if (!this.currentCrop) return;
    this.currentCrop.onButtonClick(e);
  },
  onMouseWheel: function (e) {
    if (!this.currentCrop) return;
    if (e.deltaY) {
      if (e.deltaY < 0) this.currentCrop.scale(1.015);
      else this.currentCrop.scale(0.985);
      this.applyCrop();
    }
  },
  
  resetCrop: function () {
    if (!this.currentCrop) return;
    this.currentCrop.reset();
    this.applyCrop(this.currentCrop);
  },
  rotateCrop: function () {
    if (!this.currentCrop) return;
    // this.currentCrop.reset({ rotate: true });
    this.currentCrop.rotate();
    this.applyCrop(this.currentCrop);
  },
  
  onDragMouseDown: function (e) {
    if (app.config.custom_crop === 'none' ||
        app.config.custom_crop === 'view') return;
    if (e.target.closest('.crop-cover')) return;
    if (e.button === 2) return;
    if (!this.currentCrop || !this.croppingMode) return;
    
    this.currentCrop.onDragStart(e);
  },
  onResizeMouseDown: function (e) {
    if (e.button === 2) return;
    if (!this.currentCrop || !this.croppingMode) return;
    
    this.currentCrop.onResizeStart(e);
  },

  load: function () {
    if (!this.image) {
      this.image = new Image();
      this.image.addEventListener('load', this, false);
      this.image.addEventListener('error', this, false);
    }
    this.image.src = this.model.getUrl(
                                    this.model.current_color_id || 'original');
  },
  
  handleEvent: function (e) {
    switch (e.type) {
      case 'mousedown': 
      case 'touchstart': 
        if (this.croppingMode) {
          if (e.button === 2 || !e.target.closest('.img') &&
              !e.target.closest('.purchase-option.selected')) {
            // e.preventDefault();
            this.exitCropping();
          }
        }
        break;
      case 'load':
        this.loaded = true;
        this.trigger('load');
        this.el.classList.add('loaded');
        this.parent.preloading = false;
        if (!this.buyOnly) this.parent.preloadMore();
        
        this.rq = null;
        this.image.removeEventListener('load', this, false);
        this.image.removeEventListener('error', this, false);
        this.image = null;
        break;
      case 'error':
        console.log('failed to load photo: '+ this.model.img_url, e);
        if (this.failures < 5) {
          this.image.src = this.model.img_url + 
              (this.model.img_url.indexOf('?') > -1 ? '&x=' : '?') + Date.now();
        }
        else {
          this.failed = true;
          this.trigger('error');
          this.rq = null;
        }
        this.failures++;
        break;
    }
  },
  
  resize: function (ops) {
    ops = ops || {};
    // this.buyFormInline = !this.cropOnly && app.state.width >= 1150;
    this.buyFormInline = !this.cropOnly && app.state.width >= 1250;
    this.buyFormFullscreen = !!this.cropOnly || app.state.width <= 550;
    
    var pWidth, pHeight;
    var screenRatio = app.state.width / app.state.height;
    if (app.state.fullscreen || ua.mobile) {
      pHeight = app.state.height;
      pWidth = app.state.width;
    }
    else {
      pHeight = app.state.height - 88 - (this.nameDisplayed ? 54 : 0);
      pWidth = app.state.width - 108;
    }
    if (this.parent.buyFormDisplayed && this.buyFormInline) {
      pWidth = Math.floor(app.state.width / 2);
      pHeight = app.state.height - 60;
      // pHeight = app.state.fullscreen || ua.mobile ? app.state.height 
      //                                             : app.state.height - 60;
    }
    else if (this.parent.buyFormDisplayed && !this.buyFormInline &&
            this.croppingMode) {
      pWidth = app.state.width - 60;
      pHeight = app.state.height - 60;
    }
    var spaceRatio = pWidth / pHeight;
    if (this.model.ratio > spaceRatio) {
      this.imageWidth = pWidth;
      this.imageHeight = this.imageWidth / this.model.ratio;
    }
    else {
      this.imageHeight = pHeight;
      this.imageWidth = this.imageHeight * this.model.ratio;
    }
    if (this.imageWidth > this.model.width ||
        this.imageHeight > this.model.height) {
      this.imageWidth = this.model.width;
      this.imageHeight = this.model.height;
    }
    if (!ops.notReally) {
      this.imgEl.style.top = 0;
      this.imgEl.style.left = 0;
    }
    
    // fill the screen if close enough 
    if (!this.parent.buyFormDisplayed && (app.state.fullscreen || ua.mobile)) {
      if (screenRatio > this.model.ratio && 
          app.state.width - this.imageWidth < 88) {
        this.imageWidth = app.state.width;
        this.imageHeight = Math.ceil(this.imageWidth / this.model.ratio);
        if (!ops.notReally) {
          this.imgEl.style.top = Math.ceil((
              app.state.height - this.imageHeight) / 2) +'px';
        }
      }
      else if (screenRatio < this.model.ratio && 
               app.state.height - this.imageHeight < 108) {
        this.imageHeight = app.state.height;
        this.imageWidth = Math.ceil(this.imageHeight * this.model.ratio);
        if (!ops.notReally) {
          this.imgEl.style.left = Math.ceil(
              (app.state.width - this.imageWidth) / 2) +'px';
        }
      }
    }
    
    if (this.buyForm && this.parent.buyFormDisplayed) {
      // var formWidth = app.state.width - (this.buyFormFullscreen ? 0 : 60);
      var formWidth = app.state.width - (this.buyFormFullscreen ? 0 : 20);
      // var formLeft = this.buyFormFullscreen ? 0 : 30;
      var formLeft = 0;
      if (this.buyFormInline) {
        formWidth -= this.imageWidth + 30;
      }
      else {
        formLeft = 0;
        formWidth = app.state.width;
        // if (formWidth > 650) {
        //   formLeft += Math.round((formWidth - 650) / 2);
        //   formWidth = 650;
        // }
      }
      
      if (this.parent.buyFormDisplayed && 
         (this.buyFormInline || this.croppingMode)) {
        // this.imagesRight = 30;
        this.imagesRight = 15;
      }
      else {
        this.imagesRight = app.state.fullscreen || ua.mobile ? 0 : 54;
      }
      
      if (formWidth > 850 && this.buyFormInline) {
        var excessSpace = formWidth - 850;
        formWidth = 850;
        formLeft += Math.ceil(excessSpace / 2);
        this.imagesRight += Math.floor(excessSpace / 2);
      }
      this.buyForm.el.style.width = formWidth +'px';
      this.buyForm.el.style.left = formLeft +'px';
    }
    else {
      this.imagesRight = app.state.fullscreen || ua.mobile ? 0 : 54;
    }
    
    if (!ops.notReally) {
      this.img.style.width = this.imageWidth +'px';
      this.img.style.height = this.imageHeight +'px';
      this.imgEl.style.width = this.imageWidth +'px';
      this.imgEl.style.height = this.imageHeight +'px';
      this.imagesEl.style.right = this.imagesRight +'px';
      
      if (this.currentCrop) this.applyCrop();
      
      var focus = this.query('input:focus');
      if (focus) focus.scrollIntoView();
    }
    
    if (!ops.notReally || this.buyFormInline) {
      this.el.classList.toggle('cropping-mode', this.croppingMode);
      this.el.classList.toggle('buy-form-fullscreen', this.buyFormFullscreen);
      this.el.classList.toggle('buy-form-inline', this.buyFormInline);
    }
  },
  
  remove: function () {
    BaseView.prototype.remove.apply(this);
    this.imgEl.removeAttribute('src');
    this.el.innerHTML = '';
    this.stopListening();
    if (this.rq) this.rq.abort();
    if (this.image) {
      this.image.removeEventListener('load', this, false);
      this.image.removeEventListener('error', this, false);
      this.image.removeAttribute('src');
      this.image = null;
    }
  },
  
  showBuyForm: function () {
    if (!this.buyForm) {
      this.buyForm = new BuyForm({
        model: this.model,
        parent: this,
        cropOnly: this.cropOnly,
        // el: this.query('.buy-form'),
      });
      this.el.appendChild(this.buyForm.el);
      
      this.buyForm.updatePositions();
    }
    else {
      this.buyForm.saveAsDefault = false;
    }
    
    this.loadVariants();
    
    this.buyForm.el.classList.remove('hidden');
    this.parent.buyFormDisplayed = true;
    
    this.resize({ notReally: true });
    if (this.buyFormInline) {
      var imgRect = this.img.getBoundingClientRect();
      var newLeft = app.state.width - this.imagesRight - this.imageWidth;
      var newTop = (app.state.height - this.imageHeight) / 2;
      
      var scaleX = this.imageWidth / imgRect.width;
      var scaleY = this.imageHeight / imgRect.height;
      
      this.img.classList.remove('no-transition');
      this.img.style[ua.css.transform] = 
          'translate('+ (newLeft - imgRect.left) +'px, '+ 
          (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
    }
  },
  
  hideBuyForm: function () {
    if (!this.parent.buyFormDisplayed) return;
    
    this.parent.buyFormDisplayed = false;
    
    if (this.buyOnly || this.cropOnly) {
      this.parent.close();
      return;
    }
    
    this.resize({ notReally: true });
    if (this.buyFormInline) {
      var imgRect = this.img.getBoundingClientRect();
      var titleOffset = !app.state.fullscreen && !ua.mobile && this.nameDisplayed
          ? 54 : 0;
      var newLeft = (app.state.width - this.imageWidth) / 2;
      var newTop = (app.state.height - this.imageHeight - titleOffset) / 2;
      // var newTop = (app.state.height - this.imageHeight) / 2;
      
      var scaleX = this.imageWidth / imgRect.width;
      var scaleY = this.imageHeight / imgRect.height;
      
      this.img.classList.remove('no-transition');
      this.img.style[ua.css.transform] = 
          'translate('+ (newLeft - imgRect.left) +'px, '+ 
          (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
    }
    else {
      setTimeout(this.onImageTransitionEnd, 300);
    }
    
  },
  
  toggleFav: function () {
    var favs = this.model.event.favorites;
    var favItem = favs.photos.get(this.model.id);
    if (favItem) {
      if (this.parent.collection === favs.photos) {
        if (this.parent.collection.length === 1) {
          this.parent.close();
        }
        else {
          this.parent.forward();
        }
      }
      favs.photos.remove(favItem);
    } 
    else {
      favs.photos.add({
        pid: this.model.id,
        cat_id: this.model.cat_id,
      }, { 
        parse: true,
        at: favs.photos.length,
      });
    }
    favs.save();
  },
  toggleHide: function () {
    var hides = this.model.event.hides;
    var item = hides.photos.get(this.model.id);
    if (!item) {
      hides.photos.add({
        pid: this.model.id,
        cat_id: this.model.cat_id,
      });
      
      if (this.parent.collection.parent.visible_count === 1) {
        this.parent.close();
      }
      else {
        this.parent.forward({ hide: true });
      }
    }
    else {
      hides.photos.remove(this.model.id);
    }
    hides.save();
  },
  
  onImageTransitionEnd: function () {
    this.imagesEl.classList[this.parent.buyFormDisplayed ? 'add' : 'remove'](
        'pushed-right');
    if (!this.parent.buyFormDisplayed) this.buyForm.el.classList.add('hidden');
    this.img.classList.add('no-transition');
    this.img.style[ua.css.transform] = 'none';
    this.resize();
    
    if (this.buyOnly && !this.croppingMode) {
      this.el.classList.add('buy-only');
    }
    
    // if (!this.parent.buyFormDisplayed && !this.inCart) {
    if (!this.parent.buyFormDisplayed) {
      this.buyForm.remove();
      this.buyForm.stopListening();
      this.buyForm = null;
    }
  },
});
