// app/photo/thumb.js
/*==========================================================================*\
  A single photo on the grid
\*==========================================================================*/
"use strict";
var BaseView = require('app/base/base-view');
var ShareView = require('app/social/social');

var app = require('app');
var ua = require('app/util/ua/ua');

module.exports = BaseView.extend({
  viewType: 'thumb',
  dictionary: 'thumb',
  template: require('app/templates').photo.thumb,
  displayed: false, // whether the photo has been or just about to be displayed
  
  props: {
    animatedPosition: ['boolean', true, false], // animate coordinates change
    pendingRemoval: ['boolean', true, false],
    width: ['number', true, 0],
    height: ['number', true, 0],
    x: ['number', true, 0],
    y: ['number', true, 0],
  },
  
  derived: {
    onFavPage: {
      fn: function () {
        return !!this.getPage().isFavPage;
      },
    },
  },

  initialize: function () {
    this.bindMethods('drawImage', 'onLoad');

    BaseView.prototype.initialize.apply(this, arguments);

    this.render();
    this.cacheElements({
      imageContainer: '.image',
      shareMenuContainer: '.share-menu'
    });

    this.shareView = new ShareView({
      el: this.shareMenuContainer,
      parent: this,
      model: this.model
    });
  },
  
  events: {
    'click .icon.fav': 'toggleFav',
    'click .icon.hide': 'toggleHide',
    'click .button.remove': 'removePhoto',
    'click .button.restore': 'restore'
  },
  
  bindings: {
    animatedPosition: {
      type: 'booleanClass',
      name: 'animated-position',
    },
    pendingRemoval: {
      type: 'booleanClass',
      name: 'removed',
    },
    x: {
      type: function () {
        this.el.style[ua.css.transform] = 
            'translate('+this.x+'px,'+this.y+'px)';
      },
    },
    y: {
      type: function () {
        this.el.style[ua.css.transform] = 
            'translate('+this.x+'px,'+this.y+'px)';
      },
    },
    width: {
      type: function () {
        this.el.style.width = this.width+'px';
      },
    },
    height: {
      type: function () {
        this.el.style.height = this.height+'px';
      },
    },
    'model.in_cart': {
      type: function (el, val) {
        this.el.classList.toggle('in-cart', val);
        this.query('.icon.cart .tip').innerHTML = 
            this.lz(val ? 'In Cart' : 'To Cart');
      },
      // type: 'booleanClass',
      // name: 'in-cart',
    },
    'model.in_favorites': {
      type: function (el, val) {
        this.el.classList.toggle('in-favorites', val);
        this.query('.icon.fav .tip').innerHTML = 
            this.lz(val ? 'In Favorites' : 'To Favorites');
      },
    },
    'model.hidden': {
      type: function (el, val) {
        // console.log('hidden change', this.model.id, val);
        //this.el.classList.toggle('hidden', val);
        this.query('.icon.hide .tip').innerHTML = 
            this.lz(val ? 'Hidden' : 'Hide');
      },
    },
  },
  
  loadImage: function () {
    if (this.loading || this.loaded) return;
    this.loading = true;
    
    if (!this.image) this.image = document.createElement('img');
    this.image.addEventListener('load', this, false);
    this.image.addEventListener('error', this, false);
    this.image.src = this.model.thumb_url;
    
    this.imageContainer.appendChild(this.image);
  },
  
  handleEvent: function (e) {
    if (!this.loading) return;
    switch (e.type) {
      case 'load':
        this.loaded = true;
        app.loadedImages[this.model.thumb_url] = true;
        this.trigger('load');
        this.drawImage();
       
        this.image.removeEventListener('load', this, false);
        this.image.removeEventListener('error', this, false);
        break;
      case 'error':
        console.log('failed to load photo: '+ this.model.img_url, arguments);
        if (this.failures < 5) {
          this.load();
          // this.image.src = this.model.url + 
          //     (this.model.url.indexOf('?') > -1 ? '&x=' : '?') + Date.now();
        }
        else {
          this.failed = true;
          this.trigger('error');
          // this.rq = null;
          this.image.removeEventListener('load', this, false);
          this.image.removeEventListener('error', this, false);
          this.image = null;
        }
        this.failures++;
        break;
    }
    
    this.parent.loadingCount--;
    this.parent.loadPhotos();
  },
  
  onLoad: function () {
    app.loadedImages[this.model.thumb_url] = true;
  },
  
  removePhoto: function () {
    this.parent.removePhoto(this.model);
  },
  
  restore: function () {
    
  },
  
  toggleFav: function () {
    this.model.event.trigger('fav-toggle', this.model.id, this.model.cat_id);
  },
  toggleHide: function () {
    this.model.event.trigger('hide-toggle', this.model.id, this.model.cat_id);
  },
  
  drawImage: function () {
    this.write(function () {
      this.el.classList.add('loaded');
      this.el.firstChild.style.backgroundImage = 
          'url('+this.model.thumb_url+')';
    }, 'high');
  },
});
