// app/models/product/product-model.js
/*==========================================================================*\
  Product Model 
\*==========================================================================*/
"use strict";

var OptionsCollection = require('./product-options-collection');

module.exports = require('app/models/base-model').extend({
  urlRoot: '/api/products',
  props: {
    id: 'number',
    name: ['string', true, ''],
    description: ['string', true, ''],
    
    img_url: ['string', true, ''],   
    img_width: ['number', true, 0],  
    img_height: ['number', true, 0], 
    
    'sales_tax': ['boolean', true, true], // sales tax should be charged 
    
    // product|gift_cert|album|full_download
    type: ['string', true, 'product'], 
    
    release: 'array',
  },

  collections: {
    options: OptionsCollection,
  },

  derived: {
    includesPrints: {
      fn: function () {
        var option;
        for (var i = 0; i < this.options.length; i++) {
          option = this.options.at(i);
          if (option.required_size_list.length) return true;
        }
        return false;
      },
    }
  },
});
