// app/category/category-footer.js
/*==========================================================================*\
  Fixed footer for the event category page
\*==========================================================================*/
"use strict";
var BaseView = require('app/base/base-view');

module.exports = BaseView.extend({
  
  template: require('app/templates').category['category-footer'],
  
  events: {
    'click a': 'onLinkClick'
  },
  
  'model.next': {
    hook: 'next-category',
    type: function (el, model) {
      el.style.display = model ? 'flex' : 'none';
      if (model) {
        el.href = '#events/'+ model.collection.parent.id +'/'+ model.id;
        el.innerHTML = model.name;
      }
    },
  },
  'model.prev': {
    hook: 'prev-category',
    type: function (el, model) {
      el.style.display = model ? 'flex' : 'none';
      if (model) {
        el.href = '#events/'+ model.collection.parent.id +'/'+ model.id;
        el.innerHTML = model.name;
      }
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.renderWithTemplate();
  },
  
  onLinkClick: function (e) {
    e.stopPropagation();
  },
});
