// app/category/category-page.js
/*==========================================================================*\
  A single event category page
\*==========================================================================*/
"use strict";

var assign = require('lodash.assign');
var ua = require('app/util/ua/ua');
var Page = require('app/base/base-page');
var CategoryView = require('./category');
var FixedFooter = require('./category-footer');
var ProductsView = require('app/product/products');
var SalesView = require('app/sale/sales-info');

module.exports = Page.extend({
  dictionary: 'category', // localization dictionary 
  pageTitle: '',
  pageType: 'event-category',
  template: require('app/templates').category['category-page'],
  complete: false, // whether all photos are displayed 
  
  props: {
    eventId: ['number', true, 0],
  },
  
  events: {
    // 'click .sorting a': 'sort',
    'click .load-more > .label': 'showMore',
    // 'scroll': 'onScroll',
  },
  
  bindings: {
    'model.name': '[data-hook=event-name]',
  },
  
  subviews: assign({}, Page.prototype.subviews, {
    category: {
      container: '.categories',
      waitFor: 'model.complete',
      prepareView: function (el) {
        this.triggerReady();
        this.category = new CategoryView({
          el: el,
          parent: this,
          model: this.model.categories.get(this.categoryID),
        });
        this.fixedFooter.model = this.category.model;
        this.write(function () {
          this.queryByHook('category-name').innerHTML= this.category.model.name;
          this.fixedFooter.render();
        });
        return this.category;
      },
    },
    products: {
      container: '.products',
      waitFor: 'model.complete',
      prepareView: function (el) {
        if (this.model.products_list.length) {
          this.el.classList.add('has-products');
        }
        this.products = new ProductsView({
          el: el,
          parent: this,
          collection: this.model.products_list, 
        });
        return this.products;
      },
    },
    sales: {
      container: '.sales-info',
      waitFor: 'model.categories',
      prepareView: function (el) {
        this.sales = new SalesView({
          el: el,
          parent: this,
          model: this.model.sales,
        });
        return this.sales;
      },
    },
  }),
  
  
  compare: function (page) {
    if (this.category && this.category.model && 
        page.category && page.category.model && 
        this.model.categories === page.category.model.collection) {
      
      if (this.model.categories.indexOf(this.category.model) > 
          page.category.model.collection.indexOf(page.category.model)){
        return 'right-transition';
      }
      else {
        return 'left-transition';
      }  
    }
    return 'normal-transition';
  },
  
  initialize: function (ops) {
    Page.prototype.initialize.apply(this, arguments);
    // var self = this;
    ops = ops || {};
    
    this.eventId = this.model.id;
    
    this.categoryID = ops.cid;
    if (!this.model.complete) {
      this.model.fetch();
    } 
    else {
      this.triggerReady();
    }
    // if (!this.model.name || this.model.categories.get(this.categoryID).photos) {
    //   this.model.fetch();
    // } 
    // else {
    //   this.triggerReady();
    // }
    
    this.fixedFooter = new FixedFooter({
      parent: this,
      model: this.category ? this.model.categories.get(this.categoryID) :void 0,
    });
    
    this.setTitle(this.model.name);
    this.listenTo(this.model, 'change:complete', function () {
      this.fixedFooter.model = this.model.categories.get(this.categoryID);
      this.setTitle(this.model.categories.get(this.categoryID).name + ' - '+
                    this.model.name);
      this.write(function () {
        this.fixedFooter.render();
      });
    });
    
    this.render();
  },
  
  reset: function (ops) {
    if (ops.showMore && !this.category.photos.displayedCount) {
      this.category.showMore();
    }
    return this;
  },
  
  onNearBottom: function () {
    if (ua.mobile || this.complete) return;
    this.category.showMore();
    return this;
  },
  onShow: function () {
    if (this.category) this.category.onPageShow();
  },
  
  showMore: function () {
    if (!this.complete) this.category.showMore();
    return this;
  },
  
});
