// app/splash/splash.js
/*==========================================================================*\
  Splash Screen
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var EventModel = require('app/event/event-model');

var app = require('app');
var ua = require('app/util/ua/ua');
var templates = require('app/templates');
var xhr = require('xhr');

module.exports = BaseView.extend({
  
  rafID: null,
  lastFrameTime: null,
  frameRates: null,
  
  props: {
    loginForm: ['boolean', true, false],
    homeLogin: ['boolean', true, false],
    loggingIn: ['boolean', true, false],
    wrongData: ['boolean', true, false],
    loginFormEid: ['number', true, 0],
    wrongDataEid: ['number', true, 0],
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    
    var self = this;
    this.frameRates = [];
    this.onAppReady = this.onAppReady.bind(this);
    this.onTransitionEnd = this.onTransitionEnd.bind(this);
    this.measureFPS = this.measureFPS.bind(this);
    
    app.state.splashState = 'locked';
    app.ready(function () {
      self.write(self.onAppReady);
    });
    
    this.listenTo(app, 'event-401', function () {
      this.after(1, this.onAppReady);
    });
  },
  
  bindings: {
    loginForm: {
      type: 'booleanClass',
      name: 'login',
    },
    loggingIn: {
      type: 'booleanClass',
      name: 'logging-in',
    },
    wrongData: {
      type: 'booleanClass',
      name: 'wrong-data',
    },
    'model.email': {
      type: 'booleanClass',
      name: 'email-required',
    },
    'model.password': {
      type: 'booleanClass',
      name: 'password-required',
    },
    homeLogin: {
      type: 'booleanClass',
      name: 'home-login',
    },
  },
  
  events: {
    'click': 'hide',
    'mousewheel': 'hide',
    'wheel': 'hide',
    'scroll': 'hide',
    'touchstart': 'touchstart',
    'touchmove': 'touchmove',
    'touchend': 'touchend',
    'click .back': 'goBack',
    'click .submit': 'login',
    'keypress input': 'onKeyPress',
    'click .subscribe': 'toggleSubscribe',
  },
  
  hide: function (e) {
    if (e !== 'force' && 
        (!app.isReady || this.unlocked || this.homeLogin || 
         this.model && this.model.locked)) {
      return;
    }
    // if (!this.rafID) this.rafID = ua.raf(this.measureFPS);
    this.loginForm = false;
    this.homeLogin = false;
    this.unlocked = true;
    this.loggingIn = false;
    this.wrongData = false;
    this.model = null;
    if (ua.desktop) this.write(function () {
      app.state.splashState = 'unlocked';
    }, 'high');
    else app.state.splashState = 'unlocked';

    setTimeout(this.onTransitionEnd, 1000); // a fallback just in case
    
    if (e && e.preventDefault) e.preventDefault();
    if (e && e.stopPropagation) e.stopPropagation();
    
    return false;
  },
  
  goBack: function () {
    app.navigate('events');
    this.loggingIn = false;
    this.wrongData = false;
    this.model = null;
    this.loginForm = false;
    this.homeLogin = false;
    this.hide();
  },
  
  showLoginForm: function (event) {
    if (this.loggingIn && this.model === event) {
      this.wrongData = true;
      this.wrongDataEid = event.id;
      this.query('.password input').value = '';
      if (event.email) this.query('.email input').select();
      else this.query('.password input').select();
      return;
    }
    else {
      if (this.wrongDataEid !== event.id) this.wrongData = false;
    }
    
    if (this.loginForm && this.loginFormEid === event.id) return;
    
    this.homeLogin = false;
    this.loggingIn = false;
    this.model = event;
    app.state.splashState = 'locked';
    this.unlocked = false;
    this.el.classList.add('event');
    
    if (event.display_image) {
      var thumbImage = this.query('#splash-thumb-image');
      thumbImage.style.backgroundImage = 'url('+ event.thumb_url +')';
      
      var pos = this.model.img_position || 'center center';
      // thumbImage.style.backgroundPosition = pos;
      
      var height = app.state.height;
      height -= app.state.width <= 700 || app.state.height <= 700 ? 50 : 100;
      // var path = 'https://secure.instaproofs.com/api/rest2/img_ob.php?'+
      //         'eid='+ event.id +
      //         '&w='+ app.state.width +'&h='+ app.state.height +'&fill=1';
      this.query('.image').style.backgroundImage = 'url('+
          event.img_url.replace(/&w=\d+&h=\d+/g, '') + 
          '&w='+ app.state.width +
          '&h='+ height +
          '&fill=1' +
          '&position='+ encodeURIComponent(pos) +
          ')';
    }
    else {
      //Display a blank background when `display_image` is false
      this.query('#splash-thumb-image').style.backgroundImage = '';
      this.query('.image').style.backgroundImage = '';
    }
    
    // if (this.loginForm) return;
    
    if (this.model.email && this.model.password) {
      this.loginErrorMessage = this.lz('Login data incorrect');
    }
    else if (this.model.email) {
      this.loginErrorMessage = this.lz('Wrong email');
    }
    else {
      this.loginErrorMessage = this.lz('Wrong password');
    }
    
    this.el.classList.add('loaded');
    if (app.isReady) this.el.classList.add('fade-in');
    this.loginForm = true;
    this.loginFormEid = event.id;
    this.query('.label').innerHTML = templates.auth.login(this);
    // this.query('.image').style.backgroundImage = 'url('+
    //     event.img_url.replace(/w=\d+/, 'w='+ app.state.width)
    //     .replace(/h=\d+/, 'h='+ (app.state.height - 100)) +')';

    if (this.model.email && app.storage.lastUsedEmail) {
      this.query('.email input').value = app.storage.lastUsedEmail;
    }
    // else if (this.model.email) this.query('.email input').focus();
    // else this.query('.password input').focus();

  },
  
  showHomeLoginForm: function () {
    if (this.loggingIn && (!this.model || !this.model.categories)) {
      this.wrongData = true;
      this.wrongDataEid = 0;
      this.query('.password input').value = '';
      this.query('.email input').select();
      return;
    }
    else {
      if (this.wrongDataEid !== 0) this.wrongData = false;
    }
    
    if (this.loginForm && this.loginFormEid === 0) return;
    
    this.homeLogin = true;
    this.loggingIn = false;
    this.model = app.config.photographer;
    app.state.splashState = 'locked';
    this.unlocked = false;
    this.el.classList.add('event');
    
    this.query('#splash-thumb-image').style.backgroundImage = '';
    this.query('.image').style.backgroundImage = 'url('+
        'https://secure.instaproofs.com/api/rest2/img_store.php?'+
        'store='+ window.SITE_URL +
        '&w='+ app.state.width +'&h='+ app.state.height +'&fill=1)';
    
    this.loginErrorMessage = this.lz('Login data incorrect');
    
    this.el.classList.add('loaded');
    if (app.isReady) this.el.classList.add('fade-in');
    this.loginForm = true;
    this.loginFormEid = 0;
    this.query('.label').innerHTML = templates.auth.login(this);

    if (app.storage.lastUsedEmail) {
      this.query('.email input').value = app.storage.lastUsedEmail;
    }
  },
  
  login: function () {
    var email = this.query('.email input');
    var password = this.query('.password input');
    var sub = this.query('.subscribe');
    var emailRequired = this.homeLogin || this.model.email;
    var passwordRequired = this.homeLogin || this.model.password;
    if (emailRequired && !email.value) {
      email.focus();
      return;
    }
    if (passwordRequired && !password.value) {
      password.focus();
      return;
    }
    
    email = emailRequired ? email.value : '';
    password = passwordRequired ? password.value : '';
    
    // If the display_opt_in config setting is `true`, use the checkbox's 
    // setting. Otherwise set the optIn login value to true.
    
    var optIn = true;
    if (sub && app.config.display_opt_in) {
      optIn = sub.classList.contains('checked');
    }
    
    if (this.homeLogin) {
      this.loginFromHome(email, password, optIn);
    }
    else {
      app.keychain.saveEventKeys(this.model, email, password, optIn);
      
      this.loggingIn = true;
      this.model.fetch();
      this.listenToOnce(this.model, 'sync', function () {
        this.loggingIn = false;
      });
    }
  },
  

  loginFromHome: function (email, password, optIn) {
    var self = this;
    this.loggingIn = true;
    xhr({
      uri: app.apiUrl + 'events/-100?url=' + app.url +
          '&email=' +
          encodeURIComponent(email) +
          '&password=' +
          encodeURIComponent(password) +
          (optIn ? '&opt_in=1' : ''),
    }, function (err, resp, body) {
      body = JSON.parse(body);
      // console.log('login resp', body);
      // body = JSON.parse(app.storage.eventData);
      if (!err) {
        if (body.status === 200) {
          var event = app.eventModels[body.data.id] = 
                      new EventModel({ id: body.data.id });
          var data = event.parse(body);
          event.set(data);
          
          app.keychain.saveEventKeys(event, email, password, optIn);
          app.keychain.confirmKeys(event.id, body.permissions, body.token);
          
          app.navigate(event.alt_homepage || 'events/'+ event.id);
          
          self.homeLogin = false;
          self.hide();
        }
        else {
          self.showHomeLoginForm();
        }
      }
      else {
        self.alert('Failed to connect with the server. Please check your '+
                   'internet connection and try again.');
      }
      self.loggingIn = false;
    });
  },
  
  toggleSubscribe: function () {
    var sub = this.query('.block.subscribe');
    if (sub) sub.classList.toggle('checked');
  },
  
  onKeyPress: function (e) {
    this.wrongData = false;
    if (e.keyCode === 10 || e.keyCode === 13) {
      this.login();
    }
  },
  
  onAppReady: function () {
    // console.log('onAppReady');
    // console.trace();
    // if (app.currentPage.pageType.substr(0, 5) === 'event') {
    this.listenTo(app, 'event-fetch-error', this.onEventFetchError);
    this.listenTo(app, 'event-access-denied', this.onAccessDenied);
      
    if (app.view.currentPage.eventId) {
      // var event = app.view.currentPage.model;
      var event = app.eventModels[app.view.currentPage.eventId];
      if (app.view.currentPage.pageType === 'paypal-confirmation' ||
          app.view.currentPage.pageType === 'event-checkout' ||
          app.view.currentPage.pageType === 'event-confirmation') {
        this.hide();
      }
      else if (event.locked) {
        this.showLoginForm(event);
      }
      else if (this.loginForm) {
        this.hide();
      }
      else {
        this.query('.label').innerHTML = 
            '<h1>'+ event.name +'</h1>'+
            (event.date ? 
              '<div class="date">'+ event.date.format("MMMM Do, YYYY") +
              '</div>' : '')+
            '<div class="type">'+ event.type +'</div>'+
            '<div class="arrow"></div>';
      }
    }
    else {
      if (app.config.style === 'login') {
        this.showHomeLoginForm();
      }
      if (!app.config.splash_exists || 
          this.el.classList.contains('no-splash-image')) {
        this.hide();
      }
      else {
        this.query('.prompt').innerHTML = ua.mobile
            ? '<div><span>Slide Up</span></div>'
            : '<div><span>Scroll Down</span></div>';
      }
    }
    
    if (!this.unlocked) this.el.classList.add('loaded');
    else {
      this.writeAfter(2000, function () {
        this.el.classList.add('loaded');
      });
    }
    
  },
  
  onEventFetchError: function (eid) {
    if (!this.model || this.model.id !== eid) return;
    
    if (this.loggingIn) {
      this.alert('Failed to connect with the server. Please check your '+
                 'internet connection and try again.');
    }
    this.loggingIn = false;
  },
  
  onTransitionEnd: function (e) {
    if (!this.unlocked || (e && e.target && e.target.id !== 'splash')) return;
    if (!app.state.splashState) return;
    // this.remove();
    app.state.splashState = '';
    this.loginForm = false;
    this.homeLogin = false;
    
    // avoid a webkit bug with disappearing scrollbar
    if (!app.state.noTransitions) {
      app.view.currentPage.el.style[ua.css.transform] = 'translateY(0)';
    }
    
  },
  
  measureFPS: function (time) {
    time = typeof time === 'number' ? time : ua.now();
    var rate;
    if (app.state.splashState) this.rafID = ua.raf(this.measureFPS);
    else {
      var sum = 0;
      var min = 0;
      this.frameRates = this.frameRates.slice(2, -1);
      for (var i = 0; i < this.frameRates.length; i++) {
        sum += this.frameRates[i];
        if (this.frameRates[i] > min) min = this.frameRates[i];
      }
      rate = sum / this.frameRates.length;
      console.log('Average FPS: '+ (Math.round(1000 / rate * 100) / 100));
      console.log('Minimum FPS: '+ (Math.round(1000 / min * 100) / 100));
      return;
    }
    if (!this.lastFrameTime) {
      this.lastFrameTime = time;
      return;
    }
    rate = time - this.lastFrameTime;
    this.lastFrameTime = time;
    if (rate < 80) this.frameRates.push(rate);
    console.log('FPS: '+ (Math.round(100 * 1000 / rate) / 100));
  },
  
  touchstart: function (e) {
    if (this.loginForm) return;
    if (app.isReady && !this.unlocked) {
      this.gesture = {
        startY: e.touches[0].clientY,
        y: e.touches[0].clientY,
        winHeight: document.documentElement.clientHeight,
      };
      this.el.style[ua.css.transition] = 'none';
      app.view.currentPage.el.style[ua.css.transition] = 'none';
    }
    // e.preventDefault();
  },
  touchmove: function (e) {
    if (this.loginForm) return;
    if (this.unlocked) return e.preventDefault();
    var dy = e.touches[0].clientY - this.gesture.startY;
    this.gesture.direction = e.touches[0].clientY - this.gesture.y;
    this.gesture.y = e.touches[0].clientY;
    if (dy > 0) dy = 0;
    this.el.style[ua.css.transform] = 'translate3d(0,'+ dy +'px,0)';
    // app.view.currentPage.el.style.opacity = 
    //     0.5 + Math.abs(0.5 * dy / this.gesture.winHeight);
    if (app.state.noTransitions) {
      app.view.currentPage.el.style.opacity = Math.abs(dy / this.gesture.winHeight);
    }
    else {
      app.view.currentPage.el.style[ua.css.transform] = 
          'scale('+ (0.5 + Math.abs(0.5 * dy / this.gesture.winHeight)) +')';
    }
    
    // app.view.currentPage.el.style[ua.css.transform] = 
    //     'translate3d(0,'+ (200 + 200 * dy / this.gesture.winHeight) +'px,0)';
    e.preventDefault();
  },
  touchend: function (e) {
    // if (this.unlocked) return e.preventDefault();
    if (this.unlocked) return;
    var g = this.gesture;
    if (!g) return;
    // this.el.removeAttribute('style');
    this.el.style[ua.css.transform] = null;
    this.el.style[ua.css.transition] = null;
    app.view.currentPage.el.removeAttribute('style');
    if (g.direction < 0) {
      this.el.style[ua.css.transitionDuration] = 
          ((g.winHeight - Math.abs(g.y - g.startY)) / g.winHeight * 400) +'ms';
      this.hide();
    }
    else {
      this.el.style[ua.css.transitionDuration] = 
          (Math.abs(g.y - g.startY) / g.winHeight * 400) +'ms';
    }
    // e.preventDefault();
  },
  
});

module.exports.prototype.events[ua.events.transitionEnd] = 'onTransitionEnd';
