// app/social/social.js
/*==========================================================================*\
  A popup with social sharing options
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var EmailPhotoPopup = require('app/misc-popup/email-photo-popup');

var app = require('app');

module.exports = BaseView.extend({
  viewType: 'social',
  autoRender: true,
  template: require('app/templates').social.social,

  events: {
    'click .social-button.facebook': 'shareOnFacebook',
    'click .social-button.googleplus': 'shareOnGoogleplus',
    'click .social-button.pinterest': 'shareOnPinterest',
    'click .social-button.stumbleupon': 'shareOnStumbleupon',
    'click .social-button.twitter': 'shareOnTwitter',
    'click .social-button.email': 'shareOnEmail'
  },

  shareOnFacebook: function (e) {
    return app.sharing.popup(e, 'facebook', this.model.id, this.model.eid);
  },

  shareOnGoogleplus: function (e) {
    return app.sharing.popup(e, 'googleplus', this.model.id, this.model.eid);
  },

  shareOnPinterest: function (e) {
    return app.sharing.popup(e, 'pinterest', this.model.id, this.model.eid);
  },

  shareOnStumbleupon: function (e) {
    return app.sharing.popup(e, 'stumbleupon', this.model.id, this.model.eid);
  },

  shareOnTwitter: function (e) {
    return app.sharing.popup(e, 'twitter', this.model.id, this.model.eid);
  },

  shareOnEmail: function (e) {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    this.popup(new EmailPhotoPopup({
      model: this.model
    }));

    return false;
  }
});
