// app/photo/form/addon-option.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var popup = require('app/popup/popup');

module.exports = BaseView.extend({
  template: require('app/templates').photo.form['addon-option'],
  
  props: {
    selected: ['boolean', true, false],
  },
  
  derived: {
    size: function () {
      return this.getParent('size');
    },
    size_id: {
      fn: function () {
        return this.size.model.size_id;
      },
    }
  },
  
  bindings: {
    selected: {
      type: 'booleanClass',
      name: 'selected',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.singlePrice = this.model.price;
    this.addonView = this.getParent('addon');
    this.selected = this.addonView.selected_id === this.model.id;
    this.listenTo(this, 'change:active', this.updateFolding);
  },
  
  events: {
    click: 'onClick',
  },
  
  onClick: function () {
    for (var i = 0; i < this.parent.views.length; i++) {
      this.parent.views[i].selected = this.parent.views[i] === this;
    }
    this.addonView.selected_id = this.model.id;
    this.addonView.updatePrice();
    popup.hide();
  },

});
