// app/photo/form/addon-option.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var popup = require('app/popup/popup');

module.exports = BaseView.extend({
  viewType: 'addon-option',
  template: require('app/templates').photo.form['addon-option'],
  
  props: {
    selected: ['boolean', true, false],
  },
  
  derived: {
    size: function () {
      return this.getParent('size');
    },
    size_id: {
      fn: function () {
        return this.size.model.size_id;
      },
    },
  },
  
  bindings: {
    selected: {
      type: 'booleanClass',
      name: 'selected',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.singlePrice = this.model.price;
    this.addonView = this.getParent('addon');
    this.selected = this.addonView.selected_id === this.model.id;
    this.listenTo(this, 'change:active', this.updateFolding);
  },
  
  events: {
    click: 'onClick',
  },
  
  onClick: function () {
    var options = this.getParent('addon-options').optionsView;
    for (var i = 0; i < options.views.length; i++) {
      options.views[i].selected = options.views[i] === this;
    }
    this.addonView.selected_id = this.model.id;
    this.addonView.updatePrice();
    popup.hide();
  },
  
  formatPrice: function () { // output price with discount 
    var f = this.model.formatMoney;
    var price = this.model.getPrice(this.size_id);
    var e = this.size.event;
    var discount = e.sales.allProducts;
    var discountAmount = (discount && discount.discount) 
                         ? (discount.discount * price) : 0;
    var originalPrice;
    if (discountAmount) {
      originalPrice = price;
      price = Math.round(price - discountAmount);
    }
    if (originalPrice) {
      return '<span class="linethrough">'+f(originalPrice)+'</span> '+ f(price);
    }
   else return f(price);
  },

});
