// app/photo/form/addon-options.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var OptionView = require('./addon-option');

module.exports = BaseView.extend({
  viewType: 'addon-options',
  template: require('app/templates').photo.form['addon-options'],
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.render();
  },
  
  events: {
  },
  
  derived: {
    size: {
      fn: function () {
        return this.getParent('size');
      },
    },
  },
  
  render: function () {
    this.renderWithTemplate(this);
    var sizeId = this.size.model.size_id;
    this.optionsView = 
                this.renderCollection(this.collection, OptionView, '.options', {
      filter: function (option) {
        return !option.sizes_shipping || !!option.sizes[sizeId];
      },
    });
  },

});
