// app/photo/form/photo-form.js
/*==========================================================================*\
  Buy Image Form
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var CartPhoto = require('app/cart/photo/cart-photo-model');
var OptionView = require('./size');

var ProductView = require('./product-sizes');
var ProductPopup = require('app/product/product-popup');
var ProductsCollection = require('app/product/products-collection');
var ProductsPopup = require('./products-popup');

var app = require('app');
var clone = require('lodash.clone');
var findWhere = require('lodash.findwhere');
var ua = require('app/util/ua/ua');

var popup = require('app/popup/popup');

var rowHeight = 44;

module.exports = BaseView.extend({
  template: require('app/templates').photo.form['photo-form'],
  
  props: {
    empty: ['boolean', true, true],
    totalPrice: ['number', true, 0],
    saveAsDefault: ['boolean', true, false],
    usableProducts: 'any',
    displayedVariantId: ['number', true, 0],
    instantPositioning: ['boolean', true, false],
  },
  
  derived: {
    displayedVariantPrefix: {
      deps: ['displayedVariantId'],
      fn: function () {
        return this.displayedVariantId > 0 
            ? app.colors.get(this.displayedVariantId).prefix
            : 'all-variants';
      }
    }
  },
  
  bindings: {
    cropOnly: {
      type: 'booleanClass',
      name: 'crop-only',
    },
    empty: {
      type: 'booleanClass',
      name: 'empty',
    },
    totalPrice: {
      type: function (el, val) {
        // this.empty = !val;
        // this.el.classList[val ? 'remove' : 'add']('empty');
        el.innerHTML = this.model.formatMoney(val);
      },
      selector: '.summary .total-price',
    },
    saveAsDefault: {
      type: 'booleanClass',
      name: 'save-as-default',
    },
    usableProducts: {
      type: 'booleanClass',
      name: 'usable-products',
    },
    instantPositioning: {
      type: 'booleanClass',
      name: 'instant-positioning',
    },
    displayedVariantPrefix: {
      type: 'class',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    
    // if (ops && ops.cropOnly) { 
    if (app.router.currentParameters.crop) { 
      this.cropOnly = true; 
    }
    
    if (app.storage.showAllVariants) {
      this.displayedVariantId = 0;
    }
    else {
      this.displayedVariantId = this.model.current_color_id;
    }
    
    var i;
    var products = [];
    var product;
    if (!this.productsPopupView) {
      for (i = 0; i < this.model.event.products.length; i++) {
        product = app.products.get(this.model.event.products[i]);
        if (product.includesPrints) products.push(product);
      }
    }
    if (products.length) this.usableProducts = new ProductsCollection(products);
    
    this.render();
    this.cacheElements({
      options: '.options',
    });
    
    
    var cart = this.model.event.cart;
    var cartPhotos = cart.photos.where({ pid: this.model.id });
    // var cartPhoto = cart.photos.get(
    //                     this.model.id +'-'+ this.model.current_color_id);
    // var data = (cartPhoto && cartPhoto.options);
    // if (data) this.prefill(cartPhotos);
    
    if (cartPhotos.length) this.prefill(cartPhotos);
    this.updatePrice();
    if (cartPhotos.length) {
      this.query('.button.save').innerHTML = this.lz('Save');
    }
    
    this.listenTo(cart, 'change:mtime', function () {
      this.updatePositions();
    });
  },
  
  events: {
    'click .make-default': 'toggleSaveAsDefault',
    'click .button.cancel': 'hide',
    'click .button.remove': 'removeFromCart',
    'click .button.save': 'save',
    'click .button.add': 'save',
    'click .button.add-product': 'showProductsPopup',
    
    'click .variant-name': 'selectVariant',
  },
  
  addProduct: function () {
    var popup = new ProductPopup({
      model: app.products.get(100187176),
      eventId: this.model.event.id,
    });
    
    document.body.insertBefore(popup.el, document.getElementById('popup'));
  },
  
  showProductsPopup: function () {
    if (!this.productsPopupView) {
      this.productsPopupView = new ProductsPopup({
        collection: this.usableProducts,
        model: this.model,
        parent: this,
      });
    }
    
    if (popup.displayedId && popup.displayedId === this.popupId) return;
    this.popupId = popup.show({
      targetEl: this.query('.button.add-product'),
      content: this.productsPopupView.el,
    });
  },
  
  render: function () {
    this.renderWithTemplate(this);
    
    this.productsView = this.renderCollection(this.model.event.cart.products,
        ProductView, this.query('.products-prints'), {
      filter: function (model) {
        // return model.product.includesPrints && !model.fulfilled;
        return model.product.includesPrints;
      },
      viewOptions: {
        photo: this.parent,
        buyForm: this,
        cropOnly: this.cropOnly,
      },
    });

    this.optionsView = this.renderCollection(
        this.model.buyOptions, OptionView, this.query('.options'), {
      viewOptions: {
        photo: this.parent,
        buyForm: this,
        // cropOnly: this.cropOnly,
      },
    });
  },
  
  prefill: function (cartPhotos) {
    var i, j, cartPhoto, option, addons, item;
    for (i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      // item = data[option.model.size.id];
      // cartPhoto = cartPhotos[option.model.color_id];
      cartPhoto = findWhere(cartPhotos, { color_id: option.model.color_id });
      if (!cartPhoto) continue;
      item = cartPhoto.options.get(option.model.size.id, 'size_id');
      if (!item) continue;
      
      option.active = true;
      option.qty = item.qty;
      
      if (option.crop && item.crop && typeof item.crop.top === 'number') {
        option.crop.setValues(item.crop);
      }
      
      if (option.addonsView) {
        addons = option.addonsView.views;
        var cartAddon;
        for (j = 0; j < addons.length; j++) {
          cartAddon = item.addons.get(addons[j].model.id);
          if (!cartAddon) continue;
          
          if (addons[j].model.options.length <= 1) {
            addons[j].active = true;
          }
          else {
            addons[j].selected_id = cartAddon.oid;
          }
          addons[j].updatePrice();
        }
      }
      option.updatePrice();
    }
    
    // for (i = 0; i < this.productsView.views.length; i++) {
    //   product = this.productsView.views[i];
      
    // }
    this.updatePrice();
    
    if (this.parent.inCart) {
      this.query('.button.save').innerHTML = this.lz('Save');
    }
  },
  
  selectVariant: function (e) {
    var id = e.target.getAttribute('data-color-id');
    if (this.displayedVariantId === +id) return;
    this.displayedVariantId = +id;
    
    // console.log(this.model.buyOptions.displayAll);
    
    // for all variants at once, group by size
    var displayAll = !this.displayedVariantId;
    if (this.model.buyOptions.displayAll !== displayAll) {
      this.model.buyOptions.displayAll = displayAll;
      if (displayAll) this.optionsView.views.sort(function (a, b) {
        return a.model.altOrder - b.model.altOrder;
      });
      else this.optionsView.views.sort(function (a, b) {
        return a.model.order - b.model.order;
      });
    }
    
    this.instantPositioning = true;
    this.after(100, function () {
      this.instantPositioning = false;
    });
    this.updatePositions();
  },
  
  toggleSaveAsDefault: function () {
    this.saveAsDefault = !this.saveAsDefault;
  },
  
  updatePositions: function () {
    var i, j;
    if (!this.optionsView) return;
    
    for (i = 0; i < this.productsView.views.length; i++) {
      for (j = 0; j < this.productsView.views[i].options.length; j++) {
        option = this.productsView.views[i].options[j];
        option.displayed = option.active || !this.displayedVariantId || 
                           this.displayedVariantId === option.colorId;
      }
    }
    
    // var height = 24;
    var height = 0;
    var itemHeight = rowHeight;
    var option;
    for (i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      option.displayed = option.active || !this.displayedVariantId || 
                         this.displayedVariantId === option.model.color_id;
                         
      if (!option.displayed) continue;
      
      option.el.style[ua.css.transform] = 'translateY('+ height +'px)';
      height += itemHeight + 
          (option.active && option.hasAddons ? option.addons.offsetHeight : 0);
    }
    this.options.style.height = height +'px';
  },
  
  updatePrice: function () {
    var option;
    var total = 0;
    for (var i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      total += option.price;
    }
    this.totalPrice = total;
    
    var formEmpty = !this.totalPrice && !this.query('.product-size.active');
    this.empty = !this.model.in_cart && formEmpty;
  },
  
  hide: function () {
    this.parent.hideBuyForm();
  },
  
  removeFromCart: function () {
    var cart = this.model.event.cart;
    var cartPhoto = cart.photos.get(this.model.id, 'pid');
    cart.photos.remove([cartPhoto]);
    for (var i = 0; i < cart.products.length; i++) {
      cart.products.at(i).photos.remove(this.model.id);
    }
    cart.save();
    // this.parent.inCart = false;
    this.hide();
  },
  
  save: function () {
    if (this.empty) return;
    // app.cart[this.model.id] = {
    //   photo: this.model.toJSON(),
    // };
    // var cart = app.cart[this.model.id];
    
    var i, j;
    var cart = this.model.event.cart;
    
    var cartPhotos = {};
    var cartOptions = {};
    var cartPhoto;
    var newPhotos = [];
    var option, addons;
    
    // make sure there's a cart photo for every color
    for (i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      if (!option.active) continue;
      if (!cartPhotos[option.model.color_id]) {
        cartPhoto = cart.photos.get(this.model.id+'-'+option.model.color_id);
        cartOptions[option.model.color_id] = [];
        if (!cartPhoto) {
          cartPhoto = new CartPhoto({ 
            eid: this.model.eid,
            cat_id: this.model.cat_id,
            pid: this.model.id,
            color_id: option.model.color_id,
          });
          newPhotos.push(cartPhoto);
        }
        cartPhotos[option.model.color_id] = cartPhoto;
      }
    }
    
    var cartOption;
    var cartAddon;
    for (i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      if (!option.active) continue;
      cartOption = {
        size_id: option.model.size.id,
        qty: option.qty,
        addons: [],
        crop: option.crop && option.crop.changed ? option.crop.getValues() : {},
      };
      
      // if (option.crop && option.crop.changed) {
      //   cartOption.crop = option.crop.getValues();
      // } 
      
      if (option.addonsView) {
        addons = option.addonsView.views;
        var cartAddons = [];
        var oid;
        for (j = 0; j < addons.length; j++) {
          if (addons[j].model.discount) continue;
          if (addons[j].model.bulkQty) continue;
          if (addons[j].model.bulkPrice) continue;
          if (addons[j].model.bulkType) continue;
          // if (addons[j].active || addons[j].model.options.length > 1) {
          if (addons[j].model.options && 
             (addons[j].active || addons[j].selected_id)) {
            // if (addons[j].price) {
            oid = addons[j].model.options.length ? addons[j].selected_id : true;
            if (addons[j].active && addons[j].model.options.length === 1) {
              oid = addons[j].model.options.at(0).id;
            }
            // cartAddons.push(new CartPhotoAddon({
            cartAddons.push({
              id: addons[j].model.id,
              oid: oid,
            });
            // }
          }
        }
        // cartOption.addons.set(cartAddons);
        cartOption.addons = cartAddons;
      }
      
      cartOptions[option.model.color_id].push(cartOption);
    }
    
    for (i in cartPhotos) {
      cartPhotos[i].options.set(cartOptions[i]);
    }
    // cartPhoto.options.set(cartOptions);
    
    if (newPhotos.length && this.totalPrice) {
      cart.photos.add(newPhotos);
    }
    
    
    var productView, product, productPhotos, empty;
    for (i = 0; i < this.productsView.views.length; i++) {
      productView = this.productsView.views[i];
      product = cart.products.get(productView.model);
      if (!product) continue;
      productPhotos = {};
      empty = true;
      for (j = 0; j < productView.options.length; j++) {
        option = productView.options[j];
        if (option.active) {
          empty = false;
          if (!productPhotos[option.colorId]) {
            productPhotos[option.colorId] = {
              pid: this.model.id,
              cat_id: this.model.cat_id,
              color_id: option.colorId,
              options: [],
            };
          }
          productPhotos[option.colorId].options.push({
            size_id: option.model.id,
            qty: option.qty,
            crop: option.crop && option.crop.changed 
                  ? option.crop.getValues() : {},
          });
        }
      }
      if (!empty) {
        var newPhotos = [];
        var photo;
        for (i in productPhotos) {
          photo = product.photos.get(productPhotos[i].pid +'-'+
                                     productPhotos[i].color_id);
          if (photo) photo.options.set(productPhotos[i].options);
          else newPhotos.push(productPhotos[i]);
        }
        product.photos.add(newPhotos, {
          merge: true,
        });
      }
    }
    
    if (this.saveAsDefault) {
      app.defaultOptions = clone(cartPhotos);
      app.storage.defaultOptions = JSON.stringify(cartPhotos);
    }
    
    this.parent.inCart = true;
    this.query('.button.save').innerHTML = 'Save';
    
    this.hide();
    // app.storage.cart = JSON.stringify(cart);
    // cart.store();
    cart.save();
  },
});
