// app/photo/form/products-popup-item.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var ProductPopup = require('app/product/product-popup');

var app = require('app');
var popup = require('app/popup/popup');

module.exports = BaseView.extend({
  viewType: 'products-popup-item',
  template:require('app/templates').photo.form['products-popup-item'],
  
  props: {
    selected: ['boolean', true, false],
  },
  
  derived: {
    event: {
      fn: function () {
        var parent = this.parent;
        while (!parent.model || !parent.model.event) {
          parent = parent.parent;
        }
        return parent.model.event;
      },
    },
    // size_id: {
    //   fn: function () {
    //     return this.parent.parent.parent.parent.parent.model.size_id;
    //   },
    // }
  },
  
  bindings: {
    selected: {
      type: 'booleanClass',
      name: 'selected',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    // this.singlePrice = this.model.price;
    // this.addonView = this.parent.parent;
    // this.selected = this.addonView.selected_id === this.model.id;
    // this.listenTo(this, 'change:active', this.updateFolding);
  },
  
  events: {
    click: 'onClick',
  },
  
  onClick: function () {
    var productPopup = new ProductPopup({
      model: app.products.get(this.model.id),
      eventId: this.event.id,
      // eventId: this.parent.parent.parent.parent.model.event.id,
    });
    
    document.body.insertBefore(productPopup.el, 
                               document.getElementById('popup'));
    popup.hide();
  },

});
