// app/misc-popup/cart-email-popup.js
/*==========================================================================*\
  A popup with a contact form
\*==========================================================================*/
"use strict";
var MiscPopup = require('./misc-popup');

var app = require('app');
var xhr = require('xhr');

module.exports = MiscPopup.extend({
  viewType: 'cart-email-popup',
  dictionary: 'popup.cart-email',
  template: require('app/templates')['misc-popup']['cart-email'],
  
  save: function () {
    if (this.xhr) return;
    
    var name = this.query('input.name').value;
    var email = this.query('input.email').value;
    var message = this.query('textarea.message').value;
    var toEmail = this.query('input.toEmail').value;
    if (!name) { this.query('input.name').focus(); return; }
    if (!email) { this.query('input.email').focus(); return; }
    if (!message) { this.query('textarea.message').focus(); return; }
    if (!toEmail) { this.query('input.toEmail').focus(); return; }
    
    var self = this;
    this.query('.button.save').classList.add('loading');
    
    var data = {
      sender_email: email,
      sender_name: name,
      message: message,
      to_email: toEmail,
      event_id: app.view.currentPage.eventId
    };
    
    // this.xhr = true;
    // setTimeout(function () {
    //   self.sent = true;
    // }, 500);
    // return;
    
    this.xhr = xhr({
      uri: app.apiUrl +'email/cart/' + app.view.currentPage.model.id,
      method: 'POST',
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }, function (err, resp, body) {
      body = JSON.parse(body);
      if (body.status === 200) {
        self.sent = true;
      }
      else {
        self.alert('$connection-fail');
      }
      self.xhr = null;
    });
  },
});
