// app/misc-popup/contact-popup.js
/*==========================================================================*\
  A popup with a contact form
\*==========================================================================*/
"use strict";
var MiscPopup = require('./misc-popup');

var app = require('app');
var xhr = require('xhr');

module.exports = MiscPopup.extend({
  viewType: 'contact-popup',
  dictionary: 'popup.contact',
  template: require('app/templates')['misc-popup'].contact,
  
  save: function () {
    if (this.xhr) return;
    
    var name = this.query('input.name').value;
    var email = this.query('input.email').value;
    var message = this.query('textarea.message').value;
    if (!name) { this.query('input.name').focus(); return; }
    if (!email) { this.query('input.email').focus(); return; }
    if (!message) { this.query('textarea.message').focus(); return; }
    
    var self = this;
    this.query('.button.save').classList.add('loading');
    
    var data = {
      sender_email: email,
      sender_name: name,
      message: message,
    };
    if (app.view.currentPage.eventId) {
      data.event_id = app.view.currentPage.eventId;
    }
    
    // this.xhr = true;
    // setTimeout(function () {
    //   self.sent = true;
    // }, 500);
    // return;
    
    this.xhr = xhr({
      uri: app.apiUrl +'email/contact?url='+ app.url,
      method: 'POST',
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }, function (err, resp, body) {
      body = JSON.parse(body);
      if (body.status === 200) {
        self.sent = true;
      }
      else {
        self.alert('$connection-fail');
      }
      self.xhr = null;
    });
  },
});
