// app/misc-popup/email-photo-popup.js
/*==========================================================================*\
  A popup with a contact form
\*==========================================================================*/
"use strict";
var MiscPopup = require('./misc-popup');

var app = require('app');
var axios = require('axios');
var assign = require('lodash.assign');

module.exports = MiscPopup.extend({
  viewType: 'email-photo-popup',
  dictionary: 'popup.emailPhoto',
  template: require('app/templates')['misc-popup']['email-photo'],
  
  save: function () {
    if (this.saving) return;

    var elements = this.query('form').elements;
    var i, len, name, value;
    var data = {};
    for (i = 0, len = elements.length; i < len; i++) {
      name = elements[i].name;
      value = elements[i].value;

      if (!value) {
        elements[i].focus();
        return;
      }

      data[name] = value;
    }

    this.query('.button.save').classList.add('loading');
    this.saving = true;

    var apiUrl = app.apiUrl + 'email/photo/' + this.model.id;
    var apiParams = assign({}, data, {
      event_id: this.model.eid,
      url: app.url
    });

    var self = this;
    axios.post(apiUrl, apiParams).then(function (response) {
      // Hide loading sign
      self.query('.button.save').classList.remove('loading');
      self.saving = false;

      var body = response.data;

      if (body.status !== 200) {
        self.alert('$connection-fail');
      } else {
        self.sent = true;
      }
    });
  }
});
