// app/misc-popup/email-request-popup.js
/*==========================================================================*\
  A popup requesting user's email
\*==========================================================================*/
"use strict";
var MiscPopup = require('./misc-popup');

var app = require('app');

module.exports = MiscPopup.extend({
  viewType: 'email-request-popup',
  dictionary: 'popup.email-request',
  template: require('app/templates')['misc-popup']['email-request-popup'],

  initialize: function (options) {
    MiscPopup.prototype.initialize.apply(this, arguments);
    this.success = options.success;
    this.after(340, function () {
      this.query('input').focus();
    });
  },
  
  save: function () {
    if (this.xhr) return;
    
    var email = this.query('input.email').value;
    if (!email) { this.query('input.email').focus(); return; }
    
    var self = this;
    this.query('.button.save').classList.add('loading');
    
    var data = {
      email: email,
    };
    if (app.view.currentPage.eventId) {
      data.event_id = app.view.currentPage.eventId;
    }

    if (typeof self.success === 'function') {
      self.success(data.email, function (successful) {
        if (successful) {
          self.hide();
          self.sent = true;
        }
        else {
          self.alert('$connection-fail');
        }
      });
    }
  },
});
