// app/misc-popup/misc-popup.js
/*==========================================================================*\
  A base view for popups with FAQ, Contact form etc.
\*==========================================================================*/
"use strict";
var BaseView = require('app/base/base-view');

module.exports = BaseView.extend({
  autoRender: true,
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
  },
  
  props: {
    animatedHeight: ['boolean', true, false],
    sent: ['boolean', true, false],
    data: 'object',
  },
  
  events: {
    click: 'onClick',
  },
  
  bindings: {
    animatedHeight: {
      type: 'booleanClass',
      name: 'animated-height',
    },
    sent: {
      type: 'booleanClass',
      name: 'sent',
    },
  },
  
  onClick: function (e) {
    if (!e.target.closest('.box') || e.target.closest('.close') ||
        e.target.closest('.cancel')) {
      this.hide();
    }
    if (e.target.closest('.button.save')) {
      this.save();
    }
  },
  
  hide: function () {
    this.el.classList.add('hidden');
    this.after(400, function () {
      this.stopListening();
      this.remove();
    });
  },
  
  render: function () {
    this.renderWithTemplate(this);
    // this.cacheElements({
    //   options: '.options',
    // });
    
    this.after(50, function () {
      this.updateHeight();
      var nameInput = this.query('input.name');
      if (nameInput) nameInput.focus();
    });
    
  },
  
  updateHeight: function () {
    var contentHeight = this.query('.misc-content').offsetHeight;
    this.query('.box').style.maxHeight = contentHeight + 74 + 'px';
  },
});
