// app/misc-popup/unsubscribe-popup.js
/*==========================================================================*\
 A popup with a contact form
 \*==========================================================================*/
"use strict";
var MiscPopup = require('./misc-popup');
var BaseView = require('app/base/base-view');
var UnsubscribeMixin = require('./unsubscribe-mixin');

var app = require('app');
var xhr = require('xhr');
var assign = require('lodash.assign');

module.exports = MiscPopup.extend(UnsubscribeMixin, {
  viewType: 'unsubscribe-popup',
  template: require('app/templates')['misc-popup'].unsubscribe,

  props: assign({}, MiscPopup.prototype.props, {
    saving: ['boolean', true, false]
  }),

  bindings: assign({}, MiscPopup.prototype.bindings, {
    'model.name': {
      type: function (el, val) {
        if(val !== "") {
          el.textContent = val;
        }
      },
      hook: 'unsubscribe-event-name',      
    },
    'data.bulk': {
      type: 'booleanAttribute',
      name: 'checked',
      selector: 'input[name="bulk"]'
    },
    'data.discount': {
      type: 'booleanAttribute',
      name: 'checked',
      selector: 'input[name="discount"]'
    },
    'data.event': {
      type: 'booleanAttribute',
      name: 'checked',
      selector: 'input[name="event"]'
    },
    'data.repeat': {
      type: 'booleanAttribute',
      name: 'checked',
      selector: 'input[name="repeat"]'
    },
    'data.update_all': {
      type: 'booleanAttribute',
      name: 'checked',
      selector: 'input[name="update_all"]'
    },
    'saving': {
      type: 'booleanClass',
      name: 'loading',
      selector: '.button.save',
    }
  }),

  events: assign({}, MiscPopup.prototype.events, {
    'click input[type="checkbox"]': 'onSettingChange'
  }),

  initialize: function (ops) {
    MiscPopup.prototype.initialize.apply(this, arguments);

    var self = this;

    this.eid = ops.eid;
    this.email = ops.email;
    this.prefId = ops.prefId;

    this.data = {
      update_all: false
    };

    if (!this.model.name || !this.model.complete) {
      this.model.fetchByOrigin();
    }

    this.listenTo(this.model, 'change:name', function () {
      this.getEmailSubscriptions(this.eid, this.prefId, this.email).then(function (response) {

        if (response.data.status === 200) {
          self.data = response.data.data;
        }
        else {
          console.log('Failed to get subscription settings.');
        }
      }, function (error) {
        console.log('There was an error in getting subscription settings. ' + error);
      });
    });

  },

  onSettingChange: function (event) {
    this.data[event.target.name] = event.target.checked;
  },

  save: function () {
    var self = this;

    if (this.saving) {
      return;
    }

    this.saving = true;
    this.updateEmailSubscriptions(this.eid, this.prefId, this.email, this.data).then(function (response) {
      if (response.data.status !== 200) {
        console.log('Failed to save subscription settings. ');
      }
      self.sent = true;
      self.saving = false;
    }, function (error) {
      console.log('There was an error in updating subscription settings. ' + error);
      self.saving = false;
    });

  }
  
});
