// app/models/base.js
/*==========================================================================*\
  Base model to be extended by all others
\*==========================================================================*/
"use strict";

var Model = require('ampersand-model');
var app = require('app');
var moment = require('moment');

module.exports = Model.extend({
  extraProperties: 'reject', // throw errors when setting unknown propery
  
  dataTypes: {
    
    // all dates are moments =================================================
    moment: {
      set: function (newVal) {
        var newType;
        if (newVal && typeof newVal.duration === 'function') {
          newType = 'moment';
          newVal = newVal.valueOf();
        }
        else {
          if (typeof newVal === 'number') {
            newVal = new Date(parseInt(newVal, 10));
            newVal = newVal.valueOf();
          }
          else {
            newVal = moment(newVal).valueOf();
          }
          newType = 'moment';
        }
        return {
          val: newVal,
          type: newType,
        };
      },
      get: function (val) {
        return moment(val);
      },
      default: function () {
        return moment();
      }
    },
  },
  
  
  fetch: function (options) {
    options = options || {};
    options.data = options.data || {};
    
    // request bootstrap data on first fetch since app was loaded ============
    if (!app.state.bootstrapped) {
      options.data.bootstrap = 1;
    }
    
    Model.prototype.fetch.call(this, options);
  },
  
  
  
  parse: function (res) {
    if (res.id && !res.data) return res; // not a server response 
    
    
    // parse boostrap data if it is present ==================================
    if (res.config) {
      app.state.bootstrapped = true;
      app.config.set(res.config);
    }
    if (res.photographer) app.photographer.set(res.photographer);
    
    
    // all model data is returned in a 'data' property
    return res.data;
  },
});
