// app/models/category-model.js
/*==========================================================================*\
  Event Model
\*==========================================================================*/
"use strict";

var PhotosCollection = require('./photos-collection');

module.exports = require('app/models/base-model').extend({
  urlRoot: '/api/categories',
  props: {
    id: 'number',
    name: ['string', true, ''],
    img_url: ['string', true, ''],
    img_count: ['number', true, 0],
    music: ['string', true, ''],
    // img_path: ['string', true, ''],
  },
  derived: {
    app_url: {
      fn: function () {
        return '#events/'+ this.collection.parent.id +'/'+ this.id;
      },
    },
    cover_url: {
      fn: function () {
        return this.collection.parent.img_path +'category'+ this.id +'.jpg';
      },
    },
    thumb_url: {
      deps: ['img_url'],
      fn: function () {
        return this.toThumbURL(this.img_url);
      },
    },
    
    page_id: {
      fn: function () {
        return this.collection.parent.id +'_'+ this.id;
      },
    },
    
    next: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index+1) || null;
      },
    },
    prev: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index-1) || null;
      },
    },
    
    nextAround: {
      fn: function () {
        return this.next || this.collection.at(0);
      },
    },
    prevAround: {
      fn: function () {
        return this.prev || this.collection.last();
      },
    },
  },
  collections: {
    photos: PhotosCollection,
  },
});
