// app/models/hides-model.js
/*==========================================================================*\
  A model farvorite photos for an event
\*==========================================================================*/
"use strict";

var app = require('app');
var _ =  require('underscore');

var Photo = require('./base-model').extend({
  idAttribute: 'pid',
  props: {
    pid: 'number',
    cat_id: 'number',
  },
});

var PhotosCollection = require('./base-collection').extend({
  mainIndex: 'pid',
  model: Photo,
});

module.exports = require('app/models/base-model').extend({
  // extraProperties: 'ignore',
  props: {
    id: 'number',
    eid: 'number',
    mtime: ['number', true, 0], // last modification time
  },
  
  collections: {
    photos: PhotosCollection,
  },
  
  session: {
    isHidesModel: ['boolean', true, true],
    event: 'state',
  },
  
  derived: {},
  
  initialize: function (data) {
    // this.store = _.throttle(this.store, 1000);
    this.save = _.throttle(this.save, 1000, { leading: false });
    this.event = data.event;
    
    this.listenTo(this.photos, 'add', function (model) {
      this.save();
      this.trigger('add:'+ model.pid);
    });
    this.listenTo(this.photos, 'remove', function (model) {
      this.save();
      console.log('hides remove');
      this.trigger('remove:'+ model.pid);
    });
    
    this.listenTo(app, 'storage:hides'+this.eid, this.storageCallback);
    
    if (this.event.complete) this.readStorage();
    else this.listenTo(this.event, 'change:complete', this.readStorage);
    
  },
  
  readStorage: function (){
    var storedData = app.storage['hides'+this.eid];
    if (storedData) this.set(JSON.parse(storedData), { silent: true });
  },
  
  storageCallback: function (newData) {
    newData = JSON.parse(newData);
    if (newData.mtime > this.mtime) {
      this.set(newData);
    }
  },
  
  store: function () {
    // console.log('storing cart', 
    //     JSON.stringify(this.toJSON()).length * 2 + ' bytes');
    // console.log('storing cart', this.toJSON());
    
    this.mtime = Date.now();
    app.storage['hides'+this.eid] = JSON.stringify(this.toJSON());
  },
  
  url: function () {
    var keychain = app.keychain.events.get(this.eid);
    
    var query = '?url=http://demo.instaproofs.com'+
        // '&eid='+ this.eid +
        // '&email=johnny3k@gmail.com'+
        (keychain && keychain.email ? 
        '&email='+ encodeURIComponent(keychain.email) : '') +
        (keychain && keychain.password ? 
        '&password='+ encodeURIComponent(keychain.password) : '');
        
    return 'http://www.instaproofs.com/api/rest2/images/hide/'+ this.event.id 
           + query;
  },
  
  save: function () {
    // return this.store();
    console.trace();
    var self = this;
    
    // var data = this.toJSON()
    // Parent.prototype.save.apply(this, arguments);
    // this.sync(this.isNew() ? 'create' : 'update', this, {
    this.sync(this.isNew() ? 'create' : 'update', this, {
      url: this.url(),
      // data: this.toJSON(),
      success: function (resp) {
        if (+resp.status !== 200) return;
        // self.id = resp.data.id;
        self.id = 1;
        
      },
    });
    
    this.store();
  },
  
});
