// app/models/hides-model.js
/*==========================================================================*\
  A model farvorite photos for an event
\*==========================================================================*/
"use strict";

var app = require('app');
var _ =  require('underscore');

var Photo = require('./photo-model').extend({
  idAttribute: 'pid',
  props: {
    pid: 'number',
    cat_id: 'number',
  },
});

var PhotosCollection = require('./base-collection').extend({
  mainIndex: 'pid',
  model: Photo,
});

module.exports = require('app/models/base-model').extend({
  // extraProperties: 'ignore',
  props: {
    id: 'number',
    eid: 'number',
    mtime: ['number', true, 0], // last modification time
  },
  
  collections: {
    photos: PhotosCollection,
  },
  
  session: {
    isHidesModel: ['boolean', true, true],
    event: 'state',
  },
  
  derived: {},
  
  initialize: function (data) {
    // this.store = _.throttle(this.store, 1000);
    this.save = _.throttle(this.save, 1000);
    this.event = data.event;
    
    this.listenTo(this.photos, 'add', function (model) {
      this.save();
      this.trigger('add:'+ model.pid);
    });
    this.listenTo(this.photos, 'remove', function (model) {
      this.save();
      console.log('hides remove');
      this.trigger('remove:'+ model.pid);
    });
    
    this.listenTo(app, 'storage:hides'+this.eid, this.storageCallback);
    
    if (this.event.complete) this.readStorage();
    else this.listenTo(this.event, 'change:complete', this.readStorage);
    
  },
  
  readStorage: function (){
    var storedData = app.storage['hides'+this.eid];
    if (storedData) this.set(JSON.parse(storedData));
  },
  
  storageCallback: function (newData) {
    newData = JSON.parse(newData);
    if (newData.mtime > this.mtime) {
      this.set(newData);
    }
  },
  
  store: function () {
    // console.log('storing cart', 
    //     JSON.stringify(this.toJSON()).length * 2 + ' bytes');
    // console.log('storing cart', this.toJSON());
    
    this.mtime = Date.now();
    app.storage['hides'+this.eid] = JSON.stringify(this.toJSON());
  },
  
  save: function () {
    return this.store();
    
    var self = this;
    
    // var data = this.toJSON()
    // Parent.prototype.save.apply(this, arguments);
    this.sync(this.isNew() ? 'create' : 'update', this, {
      url: this.url(),
      // data: this.toJSON(),
      success: function (resp) {
          if (+resp.status !== 200) return;
        var i, j, product, model, cartProduct, newProduct;
        self.id = resp.data.id;
        for (i = 0; i < resp.data.products.length; i++) {
          product = resp.data.products[i];
          cartProduct = self.products.get(product.cart_item_id);
          if (cartProduct) continue;
          
          newProduct = [];
          for (j = 0; j < self.products.length; j++) {
            model = self.products.at(j);
            if (model.pid === product.pid && model.oid === product.oid &&
                model.qty === product.qty && model.notes === product.notes &&
                model.photos.length === product.photos.length) {
              newProduct.push(model);
            }
          }
          
          // if we don't have a perfect fit - just reset all products 
          if (newProduct.length !== 1) {
            console.log('reset products');
            self.products.set(resp.data.products);
            break;
          }
          newProduct[0].set(product);
        }
      },
    });
    
    this.store();
  },
  
});
