// app/models/keychain-model.js
/*==========================================================================*\
  Keychain model storing all the user's login data 
\*==========================================================================*/
"use strict";

var app = require('app');
var _ =  require('underscore');

var EventKeysModel = require('./base-model').extend({
  props: {
    id: 'number',
    email: 'any',
    password: 'any',
    verified: ['boolean', true, false],
  },
});

var EventsCollection = require('./base-collection').extend({
  model: EventKeysModel,
});

module.exports = require('app/models/base-model').extend({
  // extraProperties: 'ignore',
  props: {
    mtime: 'number',
  },
  
  collections: {
    events: EventsCollection,
  },
  
  derived: {
    
  },
  
  initialize: function () {
    this.store = _.throttle(this.store, 300);
    var stored = app.storage.keychain;
    if (stored) this.set(JSON.parse(stored));
    this.events.reset();
    this.listenTo(app, 'storage:keychain', this.storageCallback);
  },
  
  storageCallback: function (newData) {
    newData = JSON.parse(newData);
    if (!newData.mtime) return;
    console.log('storage callback', newData.mtime, this.mtime);
    if (newData.mtime > this.mtime) {
      this.set(newData);
    }
  },
  
  confirmKeys: function (id) {
    var model = this.events.get(id);
    if (!model) return;
    var changed = !model.verified;
    model.verified = true;
    if (changed) {
      this.mtime = Date.now();
      this.store();
    }
  },
  saveEventKeys: function (event, email, password, verified) {
    var model = this.events.get(event.id);
    verified = !!verified;
    var changed = false;
    if (!model) {
      changed = true;
      this.events.add({
        id: event.id,
        email: email,
        password: password,
        verified: verified,
      });
    }
    else {
      changed = model.email !== email || model.password !== password ||
                model.verified !== verified;
      model.set({
        email: email,
        password: password,
        verified: verified,
      });
    }
    if (changed) {
      this.mtime = Date.now();
      this.store();
    }
  },
  removeEventKeys: function (eid) {
    if (!this.events.get(eid)) return;
    this.events.remove(eid);
    this.mtime = Date.now();
    this.store();
  },
  
  store: function () {
    console.log('storing keychaing');
    app.storage.keychain = JSON.stringify(this.serialize());
  },
});
