// app/models/photo.js
/*==========================================================================*\
  A Photo
\*==========================================================================*/
"use strict";

module.exports = require('app/models/base').extend({
  props: {
    id: 'number',
    name: ['string', true, ''],
    desc: ['string', true, ''],
    fav_id: ['number', true, 0],
    can_hide: ['boolean', true, false],
    hidden: ['boolean', true, false],
    download: ['boolean', true, false], // can be downloaded
    product: ['boolean', true, false],
    orig_width: ['number', true, 0],
    orig_height: ['number', true, 0],
    file_name: ['string', true, ''],
    
    // featured: ['boolean', true, false],
  },
  
  derived: {
    app_url: {
      deps: ['file_name'],
      fn: function () {
        return '#events/'+ this.collection.parent.collection.parent.id +'/'+
               this.collection.parent.id +'/'+ this.id;
      },
    },
    url: {
      deps: ['file_name'],
      fn: function () {
        return this.collection.parent.collection.parent.img_path + this.file_name;
      },
    },
    thumb_url: {
      deps: ['file_name'],
      fn: function () {
        return this.collection.parent.collection.parent.img_path 
               +'thumbs/'+ this.file_name;
      },
    },
    
    ratio: {
      deps: ['orig_width', 'orig_height'],
      fn: function () {
        return this.orig_width / this.orig_height;
      },
    },
    
    next: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index+1) || null;
      },
    },
    prev: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index-1) || null;
      },
    },
    
    nextAround: {
      fn: function () {
        var category = this.collection.parent;
        if (!this.next && category.collection && category.collection.length > 1) {
          return category.nextAround.prod_list.at(0);
        }
        return this.next || this.collection.at(0);
      },
    },
    prevAround: {
      fn: function () {
        var category = this.collection.parent;
        if (!this.prev && category.collection && category.collection.length > 1) {
          return category.prevAround.prod_list.last();
        }
        return this.prev || this.collection.last();
      },
    },
  },
  
  // session: {
  //   xPos: ['number', true, 0],
  //   yPos: ['number', true, 0],
  //   width: ['number', true, 0],
  //   height: ['number', true, 0],
  // },
  
  initialize: function (ops) {
    // console.log(this.collection.parent.img_path);
    // console.trace();
    // console.log(this.collection.parent.toJSON());
  },
});
