// app/models/photographer.js
/*==========================================================================*\
  Photographer's info
\*==========================================================================*/
"use strict";

module.exports = require('app/models/base').extend({
  props: {
    company_name: ['string', true, ''], // The name of the company
    
    // contact person (generally the photographer)
    first_name: ['string', true, ''],
    last_name: ['string', true, ''],
    
    email: ['string', true, ''], // The photographer's email address
    phone: ['string', true, ''], // The photographer's phone number
    storefront_url: ['string', true, ''], // URL to photographer's storefront
    facebook_url: ['string', true, ''], // The company's facebook URL
    twitter_url: ['string', true, ''], // The company's twitter URL
    linkedin_url: ['string', true, ''], // The company's linkedin URL
    instagram_url: ['string', true, ''], // The company's instagram URL
    website_url: ['string', true, ''], // The company's website URL
    address: ['string', true, ''], // The address the company is located in
    city: ['string', true, ''], // The city the company is located in
    state: ['string', true, ''], // The state the company is located in
    zip: ['string', true, ''], // company's ZIP code
    country: ['string', true, ''], // The country the company is located in
  },
  
  derived: {
    name: {
      deps: ['company_name', 'first_name', 'last_name'],
      fn: function () {
        return this.company_name || this.first_name + ' ' + this.last_name;
      },
    },
    
    combinedAddress: {
      deps: ['address', 'city', 'state', 'zip', 'country'],
      fn: function () {
        var address = [];
        if (this.address) address.push(this.address);
        if (this.city) address.push(this.city);
        if (this.state) address.push(this.state);
        if (this.zip) address.push(this.zip);
        if (this.country) address.push(this.country);
        return address.length ? address.join(', ') : '';
      },
    }
  },
});
