// app/models/product-model.js
/*==========================================================================*\
  Product Model
\*==========================================================================*/
"use strict";

module.exports = require('app/models/base-model').extend({
  urlRoot: '/api/products',
  props: {
    id: 'number',
    name: ['string', true, ''],
    descripion: ['string', true, ''],
    type: ['string', true, ''],
    option: {
      name: ['string', true, ''],
      description: ['string', true, ''],
      original_price: ['string', true, ''],
      current_price: ['string', true, ''],
      image: ['string', true, ''],
      note: ['string', true, ''],
      limited_quantity: ['boolean', true, false],
    },
    photo_list: {
      name: ['string', true, ''],
      description: ['string', true, ''],
      image: ['string', true, ''],
      photo_count: ['number', true, 0],
      aspect_ratio: ['string', true, ''],
    }

    // img_path: ['string', true, ''],
  },
  derived: {
    app_url: {
      fn: function () {
        return '#events/'+ this.collection.parent.id +'/'+ this.id;
      },
    },
    cover_url: {
      fn: function () {
        return this.collection.parent.img_path +'product'+ this.id +'.jpg';
      },
    },
    
    page_id: {
      fn: function () {
        return this.collection.parent.id +'_'+ this.id;
      },
    },
    
    next: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index+1) || null;
      },
    },
    prev: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index-1) || null;
      },
    },
    
    nextAround: {
      fn: function () {
        return this.next || this.collection.at(0);
      },
    },
    prevAround: {
      fn: function () {
        return this.prev || this.collection.last();
      },
    },
  },
});
