// app/models/product.js
/*==========================================================================*\
  Product Model
\*==========================================================================*/
"use strict";

var PhotosCollection = require('app/models/photos');

module.exports = require('app/models/base').extend({
  urlRoot: '/api/products',
  props: {
    id: 'number',
    name: ['string', true, ''],
    img_count: ['number', true, 0],
    music: ['string', true, ''],
    // img_path: ['string', true, ''],
  },
  derived: {
    app_url: {
      fn: function () {
        return '#events/'+ this.collection.parent.id +'/'+ this.id;
      },
    },
    cover_url: {
      fn: function () {
        return this.collection.parent.img_path +'product'+ this.id +'.jpg';
      },
    },
    
    page_id: {
      fn: function () {
        return this.collection.parent.id +'_'+ this.id;
      },
    },
    
    next: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index+1) || null;
      },
    },
    prev: {
      fn: function () {
        var index = this.collection.indexOf(this);
        return this.collection.at(index-1) || null;
      },
    },
    
    nextAround: {
      fn: function () {
        return this.next || this.collection.at(0);
      },
    },
    prevAround: {
      fn: function () {
        return this.prev || this.collection.last();
      },
    },
  },
  collections: {
    prod_list: PhotosCollection,
  },
});
