// app/sale/event-sales-model.js
/*==========================================================================*\
  Event's Discounts
\*==========================================================================*/
"use strict";

var BaseModel = require('app/base/base-model');
var BaseCollection = require('app/base/base-collection');

var app = require('app');
var debounce = require('lodash.debounce');
var groupBy = require('lodash.groupby');

var Discount = BaseModel.extend({
  idAttribute: 'id',
  extraProperties: 'ignore',
  props: {
    discount: 'number',
    expires: 'moment',
    description: 'string',
    id: ['number', true, 0],
  },
  
  session: {
    
    eventDiscounts: 'any', // reference to Event's Discounts model
    expiration: 'any', // moment or number of ms left if less than 48 hrs
    type: 'string',
    _discounts: 'any', // for grouped discounts listings
    _currentData: 'any', // latest raw data from the API
  },
  
  derived: {
  },
  
  initialize: function () {
    BaseModel.prototype.initialize.apply(this, arguments);
    var diff = this.expires - app.clock;
    if (diff < 50 * 3600 * 1000) {
      this.listenTo(app.clock, 'change:seconds', this.updateExpiration);
    }
    else {
      this.listenTo(app.clock, 'change:hours', this.onHourChange);
    }
    
    this.updateExpiration();
  },
  
  onHourChange: function () {
    var diff = this.expires - app.clock;
    if (diff < 50 * 3600 * 1000) {
      this.stopListening(app.clock, 'change:hours', this.onHourChange);
      this.listenTo(app.clock, 'change:seconds', this.updateExpiration);
    }
  },
  updateExpiration: function () {
    var diff = this.expires - app.clock;
    if (diff <= 0) {
      this.discount = 0;
      this.expiration = 0;
      this.stopListening();
      this.eventDiscounts.reSetData();
    }
    else if (diff < 48 * 3600 * 1000) {
      this.expiration = diff;
    }
    else {
      this.expiration = this.expires;
    }
  },
});

var Discounts = BaseCollection.extend({
  mainIndex: 'id',
  model: Discount,
});

module.exports = BaseModel.extend({
  extraProperties: 'ignore',
  props: {
    eid: 'number',
    mtime: ['number', true, 0], // last modification time
    
    allProducts: 'state',
    allSizes: 'state',
    
    general: 'array',
    size_override: 'array',
    isEmpty: ['boolean', true, true],
  },
  
  collections: {
    sizes: Discounts,
    groups: Discounts,
  },
  
  session: {
    isHidesModel: ['boolean', true, true],
    event: 'state',
  },
  
  derived: {
    // allProducts: {
    //   deps: ['general', 'size_override'],
    //   fn: function () {
    //     return this.setData({
    //       general: this.general,
    //       size_override: this.size_override,
    //     });
    //   },
    // },
  },
  
  initialize: function (data) {
    // this.allProducts = new Discount({ expires: 0, parent: this });
    // this.allSizes = new Discount({ expires: 0, parent: this });
    // console.log('init discounts', data);
    this.reSetData = debounce(this.reSetData, 1);
    // this.setData(data);
  },
  
  reSetData: function () {
    this.setData(this._currentData);
  },
  
  setData: function (data) {
    // console.log('set data', data);
    this._currentData = data;
    var i;
    var maxProductDiscount = 0;
    var maxSizeDiscount = 0;
    var productsDiscount;
    var sizesDiscount;
    var maxSizeDiscounts = {};
    var sizeDiscounts = {};
    var now = Date.now();
    var discount;
    // console.log('discounts data', data);
    if (Array.isArray(data.general)) {
      for (i = 0; i < data.general.length; i++) {
        if (Date.parse(data.general[i].expires) <= now) continue;
        if (data.general[i].product_discount > maxProductDiscount) {
          productsDiscount = data.general[i];
          maxProductDiscount = data.general[i].product_discount;
        }
        if (data.general[i].size_discount > maxSizeDiscount) {
          sizesDiscount = data.general[i];
          maxSizeDiscount = data.general[i].size_discount;
        } 
      }
    }
    
    this.allProducts = !productsDiscount ? undefined : new Discount({
      discount: productsDiscount.product_discount,
      expires: productsDiscount.expires,
      description: productsDiscount.description,
      type: 'allProducts',
      eventDiscounts: this,
    });
    // console.log('allProducts', this.allProducts);
    this.allSizes = !sizesDiscount ? undefined : new Discount({
      discount: sizesDiscount.size_discount,
      expires: sizesDiscount.expires,
      description: sizesDiscount.description,
      type: 'allSizes',
      eventDiscounts: this,
    });
    
    if (Array.isArray(data.size_override)) {
      for (i = 0; i < data.size_override.length; i++) {
        discount = data.size_override[i];
        discount.type = 'size';
        discount.eventDiscounts = this;
        if (Date.parse(discount.expires) <= now) continue;
        if (!maxSizeDiscounts[discount.id]) {
          maxSizeDiscounts[discount.id] = 0;
        }
        if (discount.discount > maxSizeDiscounts[discount.id]) {
          sizeDiscounts[discount.id] = discount;
        }
      }
    }
    
    var sizeDiscountsArray = [];
    for (var id in sizeDiscounts) {
      sizeDiscountsArray.push(sizeDiscounts[id]);
    }
    this.sizes.set(sizeDiscountsArray);
    
    // this.allProducts.set(allProducts);
    // this.allSizes.set(allSizes);
    
    this.groups.set(this.getGroups());
    
    // console.log('is empty', this.groups.length);
    this.isEmpty = !this.groups.length;
    
    // return allProducts;
  },
  
  getGroups: function () {
    var i;
    var id = 0;
    var byDate = {};
    var discounts = [];
    for (i = 0; i < this.sizes.length; i++) discounts.push(this.sizes.at(i));
    // var collection = new Discounts(this.sizes);
    if (this.allSizes) discounts.push(this.allSizes);
    if (this.allProducts) discounts.push(this.allProducts);
    var grouped = groupBy(discounts, function (discount) {
      return +discount.expires;
    });
    var groups = [];
    for (var expires in grouped) {
      groups.push({
        id: id++,
        discount: 0,
        expires: +expires,
        _discounts: grouped[expires],
        description: '',
        eventDiscounts: this,
      });
    }
    
    return groups;
  },
  
});
