// app/cart/checkout/confirmation-page.js
"use strict";

var BasePage = require('app/base/base-page');

var app = require('app');
var assign = require('lodash.assign');
var templates = require('app/templates');

module.exports = BasePage.extend({
  viewType: 'confirmation-page',
  dictionary: 'orderConfirmation', 
  pageTitle: 'Confirmation',
  pageType: 'event-confirmation',
  
  template: templates.cart.checkout['confirmation-page'],
  
  props: {
    eventId: ['number', true, 0],
  },
  
  events: {
  },
  
  bindings: {
    'model.event.name': '[data-hook=event-name]',
  },
  
  subviews: assign({}, BasePage.prototype.subviews),
  
  initialize: function (ops) {
    BasePage.prototype.initialize.apply(this, arguments);
    ops = ops || {};
    var event = this.model.event;
    this.eventId = event.id;
    if (event.complete) {
      this.triggerReady();
    }
    else {
      this.listenToOnce(event, 'change:complete', function () {
        this.triggerReady();
      });
    }
  },
  
  // render: function () {
  //   Page.prototype.render.apply(this, arguments);
  // },
  
});
