// app/cart/checkout/order-model.js
"use strict";

var BaseModel = require('app/base/base-model');

var app = require('app');
var debounce = require('lodash.debounce');

module.exports = BaseModel.extend({
  idAttribute: 'order_id',
  extraProperties: 'ignore',
  props: {
    order_id: 'number',
    confirmation_num: 'string',
    transaction_id: 'any',
    authorization_code: 'any',
    event_id: 'number',
    cart_id: 'number',
    error: 'any',
    contracts: 'any',
    digital_downloads: 'any',
    gift_certificates: 'any',
  },
  
  derived: {
    event: {
      fn: function () {
        return app.eventModels[this.event_id];
      },
    },
    receipt_url: {
      deps: ['order_id'],
      fn: function () {
        console.log('photographer', app.config.photographer.toJSON());
        return '/store/pages/orderDetails.php?oid='+ this.order_id +
               '&pid='+ app.config.photographer.user_id;
      },
    },
  },
  
  initialize: function () {
    BaseModel.prototype.initialize.apply(this, arguments);
    this.store = debounce(this.store, 100);
    
    this.listenTo(this, 'change', this.store);
    
    var stored = app.storage['order'+ this.order_id];
    if (stored) this.set(JSON.parse(stored));
    
    app.orders[this.order_id] = this;
  },
  
  store: function () {
    app.storage['order'+ this.order_id] = JSON.stringify(this.toJSON());
  },
});
