// app/cart/checkout/paypal-mixin.js
/*==========================================================================*\
 Paypal Checkout Functionalities Mixin
 \*==========================================================================*/
"use strict";
var app = require('app');
var axios = require('axios');
var assign = require('lodash.assign');
var appUrl = 'http://demo.instaproofs.com'; // app.url;

module.exports = {
  /**
   * Verifies cart totals
   *
   * @param {string} cid - Cart ID
   * @param {Object} params - Parameters
   * @param {number} params.event_id - Event ID
   * @param {number} params.shipping_state_id - Shipping State ID, optional
   * @param {string} params.shipping_city - Shipping City, optional
   * @param {string} params.shipping_zipcode - Shipping Zipcode, optional
   * @param {string} params.shipping_type - Shipping type, optional
   * @param {string} params.discount_code - Discount code, optional
   * @param {string} params.billing_email - Billing Email, optional
   *
   * @returns {Object} promise
   */
  verifyCartTotal: function (cid, params) {
    var apiUrl = app.apiUrl + 'cart/totals/' + cid;
    var apiParams = assign({}, params, {
      url: appUrl
    });

    return axios.get(apiUrl, {
      params: apiParams
    });
  },

  /**
   * Verifies that transaction is ready to start
   *
   * @param {Object} params - params
   * @param {number} params.event_id - Event ID
   * @param {number} params.amount - Amount of transaction (no decimal mark)
   * @param {string} params.discount_code - Discount code
   *                (can be blank, but must be present)
   * @param {number} params.shipping_type - Shipping Type
   *                (0=pickup, 1=shipping, 2=shipping2, 3=shipping3,
   *                4=international)
   * @param {string} params.shipping_city - Shipping City
   * @param {number} params.shipping_state_id - Shipping State ID
   * @param {string} params.shipping_zip - Shipping Zip
   * @param {string} params.email - Email
   *
   * @returns {Object} promise
   */
  verifyTransaction: function (params) {

  },

  /**
   * Starts the transaction
   *
   * @param {string} cid - Cart ID
   * @param {Object} params - params
   * @param {number} params.event_id - Event ID
   * @param {number} params.amount - Amount of transaction (no decimal mark)
   * @param {string} params.discount_code - Discount code
   *                (can be blank, but must be present)
   * @param {number} params.shipping_type - Shipping Type
   *                (0=pickup, 1=shipping, 2=shipping2, 3=shipping3,
   *                4=international)
   * @param {string} params.shipping_city - Shipping City
   * @param {number} params.shipping_state_id - Shipping State ID
   * @param {string} params.shipping_zip - Shipping Zip
   * @param {string} params.email - Email
   *
   * @returns {Object} promise
   */
  startTransaction: function (cid, params) {
    var apiUrl = app.apiUrl + 'checkout/paypal/' + cid;
    var apiParams = assign({}, params, {
      url: appUrl
    });

    return axios.post(apiUrl, apiParams);
  },

  /**
   * Finalizes the transaction
   *
   * @param {string} cid - Cart ID
   * @param {Object} params - params
   * @param {number} params.event_id
   * @param {string} params.token - this value is sent by PayPal as a URL parameter when they send the user back.
   *                It should match the token retrieved from the previous API call.
   * @param {number} params.amount - no decimal mark
   * @param {string} params.order_note - optionally provided by the user at checkout, the key must be present in the API call
   * @param {string} params.discount_code - value can be blank, but the key must be present
   * @param {number} params.shipping_type (0=pickup, 1=shipping, 2=shipping2, 3=shipping3, 4=international)
   * @param {string} params.shipping_first_name
   * @param {string} params.shipping_last_name
   * @param {string} params.shipping_address1
   * @param {string} params.shipping_address2
   * @param {string} params.shipping_city
   * @param {number} params.shipping_state_id
   * @param {string} params.shipping_zip
   * @param {string} params.email
   * @param {string} params.phone
   *
   * @returns {Object} promise
   */
  finalizeTransaction: function (cid, params) {
    var apiUrl = app.apiUrl + 'checkout/paypal/' + cid;
    var apiParams = assign({}, params, {
      url: appUrl
    });

    return axios.put(apiUrl, apiParams);
  }

  
};
