// app/views/lightbox/buy-form/addon-option.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var Parent = require('app/views/base-view');
var popup = require('app/views/global/popup');

module.exports = Parent.extend({
  template: require('app/templates').lightbox['buy-form']['addon-option'],
  
  props: {
    selected: ['boolean', true, false],
  },
  
  derived: {
    size_id: {
      fn: function () {
        return this.parent.parent.parent.parent.parent.model.size_id;
      },
    }
  },
  
  bindings: {
    selected: {
      type: 'booleanClass',
      name: 'selected',
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.singlePrice = this.model.price;
    this.addonView = this.parent.parent;
    this.selected = this.addonView.selected_id === this.model.id;
    this.listenTo(this, 'change:active', this.updateFolding);
  },
  
  events: {
    click: 'onClick',
  },
  
  onClick: function () {
    for (var i = 0; i < this.parent.views.length; i++) {
      this.parent.views[i].selected = this.parent.views[i] === this;
    }
    this.addonView.selected_id = this.model.id;
    this.addonView.updatePrice();
    popup.hide();
  },

});
