// app/views/lightbox/buy-form/addon.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var popup = require('app/views/global/popup');
var Parent = require('app/views/base-view');
var OptionsView = require('./addon-options');

module.exports = Parent.extend({
  template: require('app/templates').lightbox['buy-form'].addon,
  
  popupId: null, // ID of this view's currently displayed popup
  
  props: {
    active: ['boolean', true, false],
    price: ['number', true, 0],
    selected_id: 'number',
  },
  
  bindings: {
    active: {
      type: 'booleanClass',
      name: 'active',
    },
    price: {
      type: function (el, val) {
        el.innerHTML = '+' + this.model.formatMoney(val);
      },
      selector: '.price',
    },
    selected_id: {
      type: function (el, id) {
        // if (!id || !this.model.options.length) return;
        if (!this.model.options.length || id === undefined) return;
        var option = id === 0 ? this.model.options.at(0) 
                              : this.model.options.get(id);
        this.nameEl.innerHTML = id ? option.name : this.model.name;
        this.price = option.price;
      },
    }
  },
  
  // initialize: function () {
  //   Parent.prototype.initialize.apply(this, arguments);
  // },
  
  events: {
    'click .main': 'onClick',
    'click .info': 'showDescription',
  },
  
  onClick: function () {
    if (!this.model.options.length) {
      this.active = !this.active;
      this.updatePrice();
    }
    else {
      this.showOptions();
    }
  },
  
  render: function () {
    // console.trace();
    // if (!this.parent.collection) return;
    this.renderWithTemplate();
    this.cacheElements({
      nameEl: '.name',
      priceEl: '.price',
      info: '.info',
    });
    
    if (this.model.options.length) {
      var selected_option = this.model.options.where({ selected: true })[0];
      if (selected_option) {
        this.selected_id = selected_option.id;
      }
    }
    if (this.model) this.updatePrice(true);
  },
  
  showDescription: function (e) {
    if (popup.displayedId && popup.displayedId === this.popupId) return;
    this.popupId = popup.show({
      title: this.model.full_name || this.model.name,
      content: this.model.description,
      image: this.model.img_url ? {
        src: this.model.img_url,
        width: this.model.img_width,
        height: this.model.img_height,
      } : null,
      targetEl: this.info,
    });
  },
  
  showOptions: function () {
    if (!this.optionsView) {
      this.optionsView = new OptionsView({
        collection: this.model.options,
        model: this.model,
        parent: this,
      });
    }
    
    if (popup.displayedId && popup.displayedId === this.popupId) return;
    this.popupId = popup.show({
      targetEl: this.el,
      content: this.optionsView.el,
    });
  },
  
  updatePrice: function (noBubbling) {
    if (!this.model.options.length) {
      this.price = this.model.price;
    }
    else {
      this.price = this.selected_id ? 
          this.model.options.get(this.selected_id).price : 0;
    }
    if (!noBubbling) this.parent.parent.updatePrice();
  },

});
