// app/views/lightbox/buy-form/buy-form.js
/*==========================================================================*\
  Buy Image Form
\*==========================================================================*/
"use strict";

var app = require('app');
var ua = require('app/utils/ua');
var Parent = require('app/views/base-view');
var CartPhoto = require('app/models/cart/cart-photo-model');
var CartPhotoOption = require('app/models/cart/cart-option-model');
var CartPhotoAddon = require('app/models/cart/cart-addon-model');
var OptionView = require('./option');
var _ = require('underscore');

var rowHeight = 44;

module.exports = Parent.extend({
  template: require('app/templates').lightbox['buy-form']['buy-form'],
  
  props: {
    totalPrice: ['number', true, 0],
    saveAsDefault: ['boolean', true, false],
  },
  
  bindings: {
    totalPrice: {
      type: function (el, val) {
        this.el.classList[val ? 'remove' : 'add']('empty');
        el.innerHTML = this.model.formatMoney(val);
      },
      selector: '.total-price',
    },
    saveAsDefault: {
      type: 'booleanClass',
      name: 'save-as-default',
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.render();
    this.cacheElements({
      options: '.options',
    });
    
    var cart = this.model.event.cart;
    var cartPhoto = cart.photos.get(this.model.id, 'pid');
    // var cartPhoto = app.cart.photos.get(this.model.id, 'pid');
    // var data = (cartPhoto && cartPhoto.options) || app.defaultOptions;
    var data = (cartPhoto && cartPhoto.options);
    if (data) this.prefill(data);
    if (cartPhoto) {
      this.query('.button.save').innerHTML = 'Save';
    }
    
  },
  
  events: {
    'click .make-default': 'toggleSaveAsDefault',
    'click .button.cancel': 'hide',
    'click .button.remove': 'removeFromCart',
    'click .button.save': 'save',
    'click .button.add': 'save',
  },
  
  render: function () {
    this.renderWithTemplate(this);
    // this.optionsView = this.renderCollection(app.purchaseOptions, OptionView,
    //                                          this.query('.options'));
    this.optionsView = this.renderCollection(
        this.model.buyOptions, OptionView, this.query('.options'));
  },
  
  prefill: function (data) {
    var option, addons, item;
    for (var i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      // item = data[option.model.size.id];
      item = data.get(option.model.size.id, 'size_id');
      if (!item) continue;
      
      option.active = true;
      option.qty = item.qty;
      
      if (option.crop && item.crop && item.crop.changed) {
        option.crop.setValues(item.crop);
      }
      
      if (option.addonsView) {
        addons = option.addonsView.views;
        var cartAddon;
        for (var j = 0; j < addons.length; j++) {
          cartAddon = item.addons.get(addons[j].model.id);
          if (!cartAddon) continue;
          
          if (addons[j].model.options.length <= 1) {
            addons[j].active = true;
          }
          else {
            addons[j].selected_id = cartAddon.oid;
          }
          addons[j].updatePrice();
        }
      }
      option.updatePrice();
    }
    
    if (this.parent.inCart) {
      this.query('.button.save').innerHTML = this.lz('Save');
    }
  },
  
  toggleSaveAsDefault: function () {
    this.saveAsDefault = !this.saveAsDefault;
  },
  
  updatePositions: function () {
    var height = 24;
    // var itemHeight = this.parent.buyFormFullscreen ? rowHeight * 2 : rowHeight;
    var itemHeight = rowHeight;
    var option;
    for (var i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      option.el.style[ua.css.transform] = 'translateY('+ height +'px)';
      height += itemHeight + 
          (option.active && option.hasAddons ? option.addons.offsetHeight : 0);
    }
    this.options.style.height = height +'px';
  },
  
  updatePrice: function () {
    var option;
    var total = 0;
    for (var i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      total += option.price;
    }
    this.totalPrice = total;
  },
  
  hide: function () {
    this.parent.hideBuyForm();
  },
  
  removeFromCart: function () {
    var cartPhoto = this.model.event.cart.photos.get(this.model.id, 'pid');
    var rem = this.model.event.cart.photos.remove([cartPhoto]);
    this.model.event.cart.save();
    this.parent.inCart = false;
    this.hide();
  },
  
  save: function () {
    if (!this.totalPrice) return;
    // app.cart[this.model.id] = {
    //   photo: this.model.toJSON(),
    // };
    // var cart = app.cart[this.model.id];
    
    var cart = this.model.event.cart;
    
    var cartPhoto = cart.photos.get(this.model.id, 'pid');
    var cartOption;
    var cartAddon;
    var newPhoto = false;
    if (!cartPhoto) {
      newPhoto = true;
      cartPhoto = new CartPhoto({ 
        eid: this.model.eid,
        cat_id: this.model.cat_id,
        pid: this.model.id,
        color_id: this.model.current_color_id,
      });
    }
    var option, addons;
    var cartOptions = [];
    for (var i = 0; i < this.optionsView.views.length; i++) {
      option = this.optionsView.views[i];
      if (!option.active) continue;
      cartOption = {
        size_id: option.model.size.id,
        qty: option.qty,
        addons: [],
        crop: option.crop && option.crop.changed ? option.crop.getValues() : {},
      };
      
      // if (option.crop && option.crop.changed) {
      //   cartOption.crop = option.crop.getValues();
      // } 
      
      if (option.addonsView) {
        addons = option.addonsView.views;
        var cartAddons = [];
        var oid;
        for (var j = 0; j < addons.length; j++) {
          // if (addons[j].active || addons[j].model.options.length > 1) {
          if (addons[j].active || addons[j].selected_id) {
            // if (addons[j].price) {
            oid = addons[j].model.options.length ? addons[j].selected_id : true;
            if (addons[j].active && addons[j].model.options.length === 1) {
              oid = addons[j].model.options.at(0).id;
            }
            // cartAddons.push(new CartPhotoAddon({
            cartAddons.push({
              id: addons[j].model.id,
              oid: oid,
            });
            // }
          }
        }
        // cartOption.addons.set(cartAddons);
        cartOption.addons = cartAddons;
      }
      
      cartOptions.push(cartOption);
    }
    
    cartPhoto.options.set(cartOptions);
    
    if (newPhoto) {
      cart.photos.add(cartPhoto);
    }
    
    if (this.saveAsDefault) {
      app.defaultOptions = _.clone(cartPhoto.options);
      app.storage.defaultOptions = JSON.stringify(cartPhoto.options);
    }
    
    this.parent.inCart = true;
    this.query('.button.save').innerHTML = 'Save';
    
    this.hide();
    // app.storage.cart = JSON.stringify(cart);
    // cart.store();
    cart.save();
  },
});
