// app/views/lightbox/buy-form/product-sizes.js
/*==========================================================================*\
  A product's size options for the buy form
\*==========================================================================*/
"use strict";

var Parent = require('app/views/base-view');
var OptionView = require('./product-option');

module.exports = Parent.extend({
  template: require('app/templates').lightbox['buy-form']['product-sizes'],
  
  props: {
  },
  
  bindings: {
  },
  
  initialize: function (options) {
    Parent.prototype.initialize.apply(this, arguments);
    this.photo = options.photo;
    this.buyForm = options.buyForm;
    this.cropOnly = false;
    if (Array.isArray(options.cropOnly)) {
      if (options.cropOnly[0] === this.model.pid) {
        this.cropOnly = options.cropOnly[1];
      }
    }
  },
  
  events: {
  },
  
  
  render: function () {
    this.renderWithTemplate(this);
    this.options = this.renderCollection(this.model.option.required_size_list,
        OptionView, this.query('.sizes'), {
      viewOptions: {
        cartProduct: this.model,
        photo: this.photo,
        buyForm: this.buyForm,
        cropOnly: this.cropOnly,
      },     
    });
  },

});
