// app/nav/menu.js
/*==========================================================================*\
  Dropdown menu for the header navbar
\*==========================================================================*/
"use strict";

var app = require('app');
var ua = require('app/util/ua/ua');
var BaseView = require('app/base/base-view');
var FaqPopup = require('app/misc-popup/faq-popup');
var ContactPopup = require('app/misc-popup/contact-popup');
var OrderStatusPopup = require('app/misc-popup/order-status-popup');

var popup = require('app/popup/popup');
var LanguagesView = require('app/footer/languages');

window.toggleScheme = function () {
  app.state.darkMode = !app.state.darkMode;
  app.storage.darkMode = app.state.darkMode ? '1' : '';
  return 'Switched to '+ (app.state.darkMode ? 'dark' : 'light') + 
         ' color scheme';
};
window.toggleTransitions = function () {
  app.state.noTransitions = !app.state.noTransitions;
  app.storage.noTransitions = app.state.noTransitions ? '1' : '';
  window.location.reload();
  return 'Page transitions are '+ (app.state.noTransitions ? 'off' : 'on');
};

module.exports = BaseView.extend({
  viewType: 'menu',
  dictionary: 'nav.menu',
  // autoRender: true,
  template: require('app/templates').nav.menu,

  props: {
    isHome: ['boolean', true, false],
    i18n: 'state',
  },

  derived: {
    displayHome: {
      deps: ['isHome'],
      fn: function () {
        return app.config.display_home && !this.isHome;
      },
    }
  },
  
  events: {
    // 'click .color-scheme': 'toggleScheme',
    // 'click .transitions': 'toggleTransitions',
    'click .home': 'showHome',
    'click .faq': 'showFaq',
    'click .contact': 'showContacts',
    'click .order-status': 'showOrderStatus',
    'click .language': 'showLanguages',
  },
  
  bindings: {
    // 'model.darkMode': {
    //   type: function (el, val) {
    //     el.textContent = val ? 'Switch to light color scheme'
    //                          : 'Switch to dark color scheme';
    //   },
    //   selector: '.color-scheme',
    // },
    // 'model.noTransitions': {
    //   type: function (el, val) {
    //     el.textContent = val ? 'Turn on page transitions'
    //                          : 'Turn off page transitions';
    //   },
    //   selector: '.transitions',
    // },
    'i18n.id': {
      hook: 'language',
      type: function () {
        this.updateLanguage();
      },
    },
    displayHome: {
      type: function (el, val) {
        el.classList.toggle('hidden', !val);
      },
      selector: '.home',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);

    var self = this;
    
    this.i18n = app.i18n;
    this.listenTo(app, 'language-change', this.updateLanguage);
    
    if (app.state.bootstrapped) {
      this.render();
    }
    else {
      this.renderWithTemplate('');
      app.state.on('change:bootstrapped', function () {
        self.render();
      });
    }

    self.isHome = app.state.isHome;
    app.state.on('change:isHome', function (state, val) {
      self.isHome = val;
    });
  },

  showHome: function (e) {
    this.parent.menuDisplayed = false;
    app.router.redirectTo('events');
  },

  showFaq: function (e) {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    this.popup(new FaqPopup());

    return false;
  },

  showContacts: function (e) {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    this.popup(new ContactPopup());

    return false;
  },

  showOrderStatus: function (e) {
    if (e && e.preventDefault) {
      e.preventDefault();
    }

    this.popup(new OrderStatusPopup());

    return false;
  },
  
  updateLanguage: function () {
    this.queryByHook('language').textContent = this.i18n.locale.name;
    this.query('.language').className = 'li language '+ this.i18n.id;
  },
  
  showLanguages: function (e) {
    if (e && e.preventDefault) e.preventDefault();
    
    if (!this.languagesView || ua.ie) {
      this.languagesView = new LanguagesView({
        parent: this,
      });
    }
    
    if (popup.displayedId && popup.displayedId === this.popupId) return;
    this.popupId = popup.show({
      targetEl: this.query('.language'),
      content: this.languagesView.el,
      minWidth: 1,
    });
  },
});
