// app/cart/summary/code/discount-code.js
"use strict";

var BaseView = require('app/base/base-view');
var TierView = require('./discount-code-tier');

module.exports = BaseView.extend({
  viewType: 'discount-code',
  
  template: require('app/templates').cart.summary.code['discount-code'],
  
  props: {
    loading: ['boolean', true, false],
  },
  
  bindings: {
    'model.amount': {
      type: function (el, value) {
        el.classList.toggle('active', !!value);
      },
    },
    'model.include_products': {
      type: 'booleanClass',
      name: 'includes-products',
    },
    'model.include_tax': {
      type: 'booleanClass',
      name: 'includes-tax',
    },
    'model.discount_code': {
      type: 'value',
      selector: 'input',
    },
    'model.loading': {
      type: function () {
        this.loading = this.model.loading;
      },
    },
    'model.type': {
      type: function (el, val) {
        el.classList.toggle('shipping-code', val === 'shipping');
      },
    },
    'model.valid': {
      type: function (el, val) {
        el.classList.toggle('valid', val);
        this.query('input').readOnly = val;
      },
    },
    loading: {
      type: function (el, val) {
        this.el.classList.toggle('loading', val);
        this.query('.button.apply-code').classList.toggle('loading', val);
      },
    },
  },
  
  events: {
    'click .button.apply-code': 'applyCode',
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    
    this.listenTo(this.model, 'calc', this.updateTiers);
    this.updateTiers();
  },
  
  render: function () {
    BaseView.prototype.render.apply(this, arguments);
    this.renderCollection(this.model.tiers, TierView, '.tiers');
    this.cacheElements({
      inputEl: 'input',
    });
  },
  
  applyCode: function () {
    if (this.loading) return;
    if (this.model.valid) return this.resetCode();
    
    var code = this.inputEl.value.trim();
    if (!code) {
      this.inputEl.focus();
      return;
    }
    var self = this;
    
    console.log('apply code', code);
    if (this.model.discount_code === code) return;
    this.model.set('discount_code', code, { silent: true });
    
    this.model.fetch({
      success: function (model, res) {
        // res = { // Single Dollar Value: $12 off any size order
        //   "status": 200,
        //   "data": {
        //     "discount_code": "d4e2e4a423",
        //     "valid": true,
        //     "message": null,
        //     "type": "sDollar",
        //     "remaining": 1200,
        //     "tiers": [{ "min": 0, "value": 1200 }],
        //     "include_products": false,
        //     "pay_for_tax": false,
        //     "apply_pre_tax_calc": false
        //   },
        //   "timestamp": 1431543048
        // };
        
        
        if (res.status === 404) {
          self.alert("No matching discount code found.");
          self.inputEl.value = '';
          self.model.clear();
          return;
        }
        if (res.status === 410) {
          self.alert("This discount code has no value remaining.", {
            width: 400,
          });
          self.model.clear();
          self.inputEl.value = '';
          return;
        }
        console.log('success', res);
      },
      error: function () {
        self.alert("Failed to verify the checkout code.<br>"+
            "Please check your internet connection and try again.");
      },
    });
  },
  
  resetCode: function () {
    this.inputEl.value = '';
    this.model.clear();
    this.model.store();
    this.model.parent.updateSummary();
  },
  
  updateTiers: function () {
    if (!this.el) return;
    var firstTier = this.model.tiers.at(0);
    
    
    this.el.classList.toggle('with-header', this.model.type !== 'shipping' && 
        (this.model.tiers.length >= 2 || firstTier && firstTier.min));
  },
});
