// app/cart/photo/cart-photo.js
/*==========================================================================*\
  A single photo in the cart 
\*==========================================================================*/
"use strict";

var BaseView = require('app/base/base-view');
var OptionView = require('./option/cart-photo-option');

module.exports = BaseView.extend({
  viewType: 'cart-photo',
  dictionary: 'cartPhoto',
  template: require('app/templates').cart.photo['cart-photo'],
  
  props: {
    imageWidth: 'number', // dimensions of an image as currently displayed 
    imageHeight: 'number',
    
    currentCrop: 'object',
  },
  
  bindings: {
    imageWidth: {
      type: function (el, val) {
        el.style.width = val + 'px';
      },
      selector: '.img',
    },
    imageHeight: {
      type: function (el, val) {
      el.classList.toggle('small', val <= 90);
        el.style.height = val + 'px';
      },
      selector: '.img',
    },
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.render();
    this.imageWidth = 100;
    this.imageHeight = 100 / this.model.photo.ratio;
    // this.cacheElements({
    // });
  },
  
  events: {
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.renderCollection(this.model.options, OptionView, this.query('.items'));
  },
});
