// app/event/maintenance/event-maintenance.js
/*==========================================================================*\
  An overlay view covering event's pages when it's under maintenance
\*==========================================================================*/
"use strict";

var app = require('app');
var BaseView = require('app/base/base-view');


module.exports = BaseView.extend({
  viewType: 'event-maintenance',
  dictionary: 'event', // localization dictionary 
  
  template: require('app/templates').event.maintenance['event-maintenance'],
  
  props: {
    eventId: ['number', true, 0],
  },
  
  events: {
  },
  
  bindings: {
    'model.name': '[data-hook=event-name]',
  },
  
  initialize: function (ops) {
    var event = this.model;
    BaseView.prototype.initialize.apply(this, arguments);
    this.render();

    if (event && event.display_image) {
      var lrImageEl = this.query('.event-meintenance-lowres-image');
      var hrImageEl = this.query('.event-maintenance-highres-image');

      lrImageEl.style.backgroundImage = 'url('+ event.thumb_url +')';
      
      var pos = this.model.img_position || 'center center';
      // thumbImage.style.backgroundPosition = pos;
      
      hrImageEl.style.backgroundImage = 'url('+
          event.img_url.replace(/&w=\d+&h=\d+/g, '') + 
          '&w='+ app.state.width +
          '&h='+ app.state.height +
          '&fill=1' +
          '&position='+ encodeURIComponent(pos) +
          ')';
    }
  },
  
});
