// app/views/event/categories.js
/*==========================================================================*\
  List of categories on event's page
\*==========================================================================*/
"use strict";

var Parent = require('app/views/base-view');
var CatView = require('./category-item');

module.exports = Parent.extend({
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
  },
  
  events: {
    
  },
  
  render: function () {
    this.updateLayout();
    
    this.renderCollection(this.collection, CatView, this.query('.categories'), {
      viewOptions: {
        preview: true,
        parent: this,
      },
    });
  },
  
  updateLayout: function () {
    this.scheduleWrite(function () {
      var len = this.collection.length;
      this.el.classList.toggle('single', len === 1);
      // this.el.classList.toggle('odd', len <= 5 && len > 1 && len % 2 > 0);
      this.el.classList.toggle('odd', len > 1 && len % 2 > 0);
      this.el.classList.toggle('trios', len > 2);
      this.el.classList.toggle('odd-1', len % 3 === 1);
      this.el.classList.toggle('odd-2', len % 3 === 2);
    });
  },
 
});
