// app/views/event/event-page.js
/*==========================================================================*\
  Event page with a list of categories 
\*==========================================================================*/
"use strict";

var _ = require('underscore');
var Page = require('app/views/base-page');
var CategoriesView = require('./categories');
var ProductsView = require('./products');

module.exports = Page.extend({
  dictionary: 'event', // localization dictionary 
  pageTitle: '',
  pageType: 'event',
  
  template: require('app/templates').event['event-page'],
  
  props: {
    eventId: ['number', true, 0],
  },
  
  events: {
    // 'click .sorting a': 'sort',
    'click .load-more > .label': 'loadMore',
  },
  
  subviews: _.extend({}, Page.prototype.subviews, {
    categories: {
      container: '.categories',
      waitFor: 'model.name',
      prepareView: function (el) {
        this.triggerReady();
        this.categories = new CategoriesView({
          el: el,
          parent: this,
          collection: this.model.categories,
        });
        return this.categories;
      },
    },
    products: {
      container: '.products',
      waitFor: 'model.name',
      prepareView: function (el) {
        this.triggerReady();
        this.products = new ProductsView({
          el: el,
          parent: this,
          collection: this.model.products, 
        });
        return this.products;
      },
    },
  }, Page.prototype.subviews),
  
  bindings: {
    'model.name': '[data-hook=event-name]',
  },
  
  initialize: function (ops) {
    Page.prototype.initialize.apply(this, arguments);
    ops = ops || {};
    
    this.eventId = this.model.id;
    
    if (!this.model.name || !this.model.complete) {
      this.model.fetch();
    } 
    else {
      this.triggerReady();
    }
    
    this.setTitle(this.model.name);
    this.listenTo(this.model, 'change:name', function () {
      this.setTitle(this.model.name);
    });
    
    this.render();
  },
  
});
