// app/views/event/event-nav.js
/*==========================================================================*\
  Fixed header navbar for an event 
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/global/nav-contents');

module.exports = Parent.extend({
  template: require('app/templates').event['event-nav'],
  
  props: {
    cartEmpty: ['boolean', true, true],
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    
    this.listenTo(this.model.cart, 'change:is_empty', this.updateIcons);
    this.listenTo(app.router, 'page', this.updateIcons);
  },
  
  switchNav: function (newView) { // show a nav view 
    this.viewSwitcher.set(newView);
  },
  
  onLinkClick: function (e) {
    e.stopPropagation();
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.cacheElements({
      favsEl: '.favs',
      cartEl: '.cart',
      menuEl: '.menu',
    });
    this.updateIcons();
  },
  
  updateIcons: function () {
    this.cartEl.classList.toggle('filled', !this.model.cart.is_empty);
    
    if (app.router.currentPage) {
      this.cartEl.classList.toggle('current', 
          app.router.currentPage === 'cart');
    }
  },
});
