// app/views/event/photo.js
/*==========================================================================*\
  A single photo on the grid
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base');
var bindTransforms = require('bind-transforms');

module.exports = Parent.extend(bindTransforms).extend({
  template: require('app/templates').events.photo,
  displayed: false, // whether the photo has been or just about to be displayed 

  initialize: function () {
    this.drawImage = this.drawImage.bind(this);
    this.onLoad = this.onLoad.bind(this);
    Parent.prototype.initialize.apply(this, arguments);
    this.render();
    this.cacheElements({
      imageContainer: '.image',
      // img: 'img',
    });
  },
  
  events: {
    
  },
  
  loadImage: function () {
    if (this.loading || this.loaded) return;
    this.loading = true;
    
    // if (!this.image) this.image = new Image();
    if (!this.image) this.image = document.createElement('img');
    this.image.addEventListener('load', this, false);
    this.image.addEventListener('error', this, false);
    this.image.src = this.model.thumb_url;
    
    this.imageContainer.appendChild(this.image);
    // this.scheduleWrite(function () {
    //   // this.img.style.display = 'block';
    // }, 'high');
    
    // this.drawImage();
    
    // var rq = this.rq = new app.XHR();
    // // rq.addEventListener("progress", this, false);
    // rq.addEventListener("load", this, false);
    // rq.addEventListener("error", this, false);
    // rq.open('get', this.model.thumb_url, true);
    // rq.send();
    
    // if (this.model.id === 7128) {
    //   console.log('loading the image!!!');
    //   console.trace();
    // }
  },
  
  handleEvent: function (e) {
    if (!this.loading) return;
    switch (e.type) {
      case 'progress':
        // console.log('progress', Math.round(e.loaded / e.total * 100));
      break;
      case 'load':
        this.loaded = true;
        app.loadedImages[this.model.thumb_url] = true;
        this.trigger('load');
        this.drawImage();
        // this.el.classList.add('loaded');
        // console.log('loaded image');
        // this.parent.preloading = false;
        // this.parent.preloadMore();
        // this.trigger('complete');
        
        // this.rq = null;
        this.image.removeEventListener('load', this, false);
        this.image.removeEventListener('error', this, false);
        // this.image.removeAttribute('src');
        // this.image = null;
        break;
      case 'error':
        console.log('failed to load photo: '+ this.model.url, arguments);
        if (this.failures < 5) {
          this.load();
          // this.image.src = this.model.url + 
          //     (this.model.url.indexOf('?') > -1 ? '&x=' : '?') + Date.now();
        }
        else {
          this.failed = true;
          this.trigger('error');
          // this.rq = null;
          this.image.removeEventListener('load', this, false);
          this.image.removeEventListener('error', this, false);
          this.image = null;
        }
        this.failures++;
        break;
    }
    
    this.parent.loadingCount--;
    this.parent.loadPhotos();
  },
  
  onLoad: function () {
    app.loadedImages[this.model.thumb_url] = true;
  },
  
  drawImage: function () {
    // this.image = null;
    // this.rq = null;
    // console.log('draw', this.model.id);
    // this.after(200 + Math.random() * 300, function () {
      this.scheduleWrite(function () {
        this.el.classList.add('loaded');
        this.el.firstChild.style.backgroundImage = 'url('+this.model.thumb_url+')';
      }, 'high');
    // });
  },
  
  // render: function () {
  //   this.renderWithTemplate(this);
  //   this.bindTransforms({
  //     top: 'yPos',
  //     left: 'xPos',
  //     width: 'width',
  //     height: 'height',
  //   }, this.el, true);
  // }
});
