// app/views/event/product-footer.js
/*==========================================================================*\
  Fixed footer for the event product page
\*==========================================================================*/
"use strict";
var Parent = require('app/views/base');

module.exports = Parent.extend({
  
  template: require('app/templates').events['product-footer'],
  
  events: {
    'click a': 'onLinkClick'
  },
  
  'model.next': {
    hook: 'next-product',
    type: function (el, model) {
      el.style.display = model ? 'flex' : 'none';
      if (model) {
        el.href = '#events/'+ model.collection.parent.id +'/'+ model.id;
        el.innerHTML = model.name;
      }
    },
  },
  'model.prev': {
    hook: 'prev-product',
    type: function (el, model) {
      el.style.display = model ? 'flex' : 'none';
      if (model) {
        el.href = '#events/'+ model.collection.parent.id +'/'+ model.id;
        el.innerHTML = model.name;
      }
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.renderWithTemplate();
  },
  
  onLinkClick: function (e) {
    e.stopPropagation();
  },
});
