// app/views/event/products.js
/*==========================================================================*\
  List of products on event's page
\*==========================================================================*/
"use strict";

var Parent = require('app/views/base');
// var CatView = require('app/views/event/product');
var ProdView = require('app/views/event/product-item');
var $ = require('jquery');

module.exports = Parent.extend({
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
  },
  
  events: {
    
  },
  
  render: function () {
    this.updateLayout();
    
    this.renderCollection(this.collection, ProdView, this.query('.products'), {
      viewOptions: {
        preview: true,
      },
    });
  },
  
  updateLayout: function () {
    this.scheduleWrite(function () {
      var $el = $(this.el);
      var len = this.collection.length;
      $el.toggleClass('single', len === 1);
      // $el.toggleClass('odd', len <= 5 && len > 1 && len % 2 > 0);
      $el.toggleClass('odd', len > 1 && len % 2 > 0);
      $el.toggleClass('trios', len > 2);
      $el.toggleClass('odd-1', len % 3 === 1);
      $el.toggleClass('odd-2', len % 3 === 2);
    });
  },
 
});
