// app/views/events/events-group.js
/*==========================================================================*\
  A group of events displayed together (e.g. from the same month)
\*==========================================================================*/
"use strict";

var Parent = require('app/views/base-view');
var moment = require('moment');
var ItemView = require('./event-item');

module.exports = Parent.extend({
  sortedBy: 'date',
  template: require('app/templates').events['events-group'],
  
  initialize: function (options) {
    Parent.prototype.initialize.apply(this, arguments);
    
    this.sortedBy = options.sortedBy;
    this.key = options.key;
    this.noHeader = !!options.noHeader;
  },
  
  events: {
    // 'click a[href]': 'handleLinkClick'
  },
  
  render: function () {
    var event = this.collection.at(0);
    var label;
    switch (this.sortedBy) {
      // case 'date': label = event.date.format('YYYY<i>MMM</i>'); break;
      // case 'date': label = event.date.format('MMMM<i>YYYY</i>'); break;
      case 'date': label = this.lz('MMMM<i>YYYY</i>', event.date); break;
      case 'name': label = event.name.substr(0, 1); break;
      case 'type': label = event.type; break;
    }
    
    var classes = [];
    // if (this.collection.length % 3 === 0) classes.push('trios');
    if (this.noHeader) classes.push('no-header');
    if (this.collection.length === 1) classes.push('single');
    if (this.collection.length > 2) {
      classes.push('trios');
      if (this.collection.length % 3 === 1) classes.push('odd-1');
      else if (this.collection.length % 3 === 2) classes.push('odd-2');
    }
    if (this.collection.length > 1 && this.collection.length % 2 > 0) {
      classes.push('odd');
    }
    if (this.sortedBy === 'date' 
        && event.date.format('MM-YYYY') === moment().format('MM-YYYY')) {
      classes.push('no-header');     
    }
    this.renderWithTemplate({
      label: label,
      classes: classes,
    });
    this.renderCollection(this.collection, ItemView, 
                          this.query('.items'));
  },
  
  updateLayout: function () {
    var len = this.collection.length;
    this.el.classList.toggle('single', len === 1);
    // this.el.classList.toggle('odd', len <= 5 && len > 1 && len % 2 > 0);
    this.el.classList.toggle('odd', len > 1 && len % 2 > 0);
    this.el.classList.toggle('trios', len > 2);
    this.el.classList.toggle('odd-1', len % 3 === 1);
    this.el.classList.toggle('odd-2', len % 3 === 2);
  },
});
