// app/cart/product/cart-product.js
/*==========================================================================*\
  A single product in the cart 
\*==========================================================================*/
"use strict";

var app = require('app');

var BaseView = require('app/base/base-view');
var OptionView = require('./cart-product-option');

module.exports = BaseView.extend({
  template: require('app/templates').cart.product['cart-product'],
  
  props: {
    photoWidth: 'number',
    columnsNum: 'number',
    removed: ['boolean', true, false],
  },
  
  derived: {
    cartPage: {
      fn: function () {
        var parent = this.parent;
        while (!parent.pageType) parent = parent.parent;
        return parent;
      },
    }
  },
  
  bindings: {
    'model.qty': {
      type: function (el, val) {
        el.value = val;
        this.el.classList.toggle('qty-is-one', val === 1);
      },
      // type: 'value',
      selector: '.qty input',
    },
    'model.basePrice': {
      type: function (el, val) {
        el.textContent = this.model.formatMoney(val);
      },
      selector: '.original-price',
    },
    'model.price': {
      type: function (el, val) {
        el.textContent = this.model.formatMoney(val);
      },
      selector: '.discounted-price',
    },
    'removed': {
      type: 'booleanClass',
      name: 'removed',
    },
  },
  
  events: {
    'click .head .plus': 'plus',
    'click .head .minus': 'minus',
    
    'click .head .action.remove': 'removeFromCart',
    'click .head .button.restore': 'restore',
    'click .head .button.hide': 'hide',
    
    'dblclick .plus, .minus, .button': 'preventDefault',
  },
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    // this.render();
    
    // keep an up-to-date calculation of product's content width
    this.listenTo(app.state, 'change:width', this.updateWidth);
    this.updateWidth();
  },
  
  updateWidth: function () {
    // var i;
    var width = app.state.width >= 1304 ? 1200 : app.state.width * 0.92;
    if (app.state.width > 950) { // substruct the summary box
      width -= 220;
    }
    width -= 90; // paddings and borders
    width = Math.floor(width);
    
    this.columnsNum = Math.floor(width / 180);
    this.photoWidth = Math.floor(width / this.columnsNum);
  },
  
  plus: function (e) {
    if (e && e.preventDefault) e.preventDefault();
    var multi = 1;
    if (e && e.altKey) multi *= 10;
    if (e && e.shiftKey) multi *= 5;
    this.model.qty += multi;
    
    this.model.cart.updateSummary();
    this.model.cart.save();
  },
  minus: function (e) {
    if (e && e.preventDefault) e.preventDefault();
    var multi = 1;
    if (e && e.altKey) multi *= 10;
    if (e && e.shiftKey) multi *= 5;
    
    this.model.qty -= this.model.qty - multi < 1 ? this.model.qty - 1 : multi;
    
    this.model.cart.updateSummary();
    this.model.cart.save();
  },
  
  removeFromCart: function () {
    // this.model.qty = 0;
    this.cartModel = this.model.cart;
    this.collection = this.model.collection;
    this.model.collection.remove(this.model, { silent: true });
    this.removed = true;
    
    this.cartModel.save();
    // this.el.classList.add('removed');
  },
  restore: function () {
    var items = this.parent.el.querySelectorAll('.cart-product');
    var i, index = 0;
    for (i = 0; i < items.length; i++) {
      if (items[i] === this.el) {
        index = i;
        break;
      }
    }
    this.collection.add(this.model, { at: index, silent: true });
    this.removed = false;
    
    this.cartModel.save();
  },
  hide: function () {
    this.remove();
    
    this.cartPage.checkForEmptiness();
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.cacheElements({
      sizesEl: '.sizes',
    });
    var i, view;
    
    this.optionViews = [];
    for (i = 0; i < this.model.option.required_size_list.length; i++) {
      view = new OptionView({
        collection: this.model.photos,
        model: this.model.option.required_size_list.at(i),
        parent: this,
      });
      this.optionViews.push(view);
      this.sizesEl.appendChild(view.el);
    }
    
    var e = this.model.cart.event;
    var discount = e.sales.allProducts;
    this.el.classList.toggle('discounted', !!discount);
    if (discount) this.query('.discount').innerHTML = 
                      '-'+ Math.round(discount.discount * 100) + '%';
    
    this.updateWidth();
  },
});
