// app/views/lightbox/lightbox.js
/*==========================================================================*\
  Main view for the lightbox photo viewer
\*==========================================================================*/
"use strict";

var app = require('app');
var fullscreen = require('app/utils/fullscreen');
var keyboard = require('app/utils/keyboard');
var templates = require('app/templates');
var ua = require('app/utils/ua');

var Parent = require('app/views/base-view');
var PhotoView = require('./photo');


module.exports = Parent.extend(require('./slideshow-mixin'),
                               require('./touch-mixin'),
{
  props: {
    audioReady: ['boolean', true, false], // audio is ready and playing
    slideshow: ['boolean', true, false], // whether slideshow is rolling 
    hasMusic: ['boolean', true, false], // whether current category has music
    music: ['boolean', true, false], // whether music is currently playing
    mouseRest: 'any', // no mouse or touch activity for a while
    mouseOverSlideshow: ['boolean', true, false], // mouse over slidshow controls
    mouseOverTopRight: ['boolean', true, false], // mouse over top right block
    
    buyFormDisplayed: ['boolean', true, false], // whether buy form is displayed
    
    // whether the UI should currently be hidden if in fullscreen
    noUI: ['boolean', true, false], 
    
    // whether we're currently preloading a photo
    preloading: ['boolean', true, false],
  },
  backTo: null, // page to go back to when closing 
  displayed: false, // whether the lightbox is currently displayed
  mouseLastPosition: null, // last registered cursor position
  mouseRestPosition: null, // position at which we entered 'no-input' mode
  mouseRestTimer: null,
  mouseRestSpeed: 1500, // time after which the mouse is considered inactive
  // noUI: false, // whether the UI is currently hidden
  noUITimer: null, // timer for hiding UI in fullscreen 
  
  initialize: function () {
    var self = this;
    Parent.prototype.initialize.apply(this, arguments);
    this.back = this.back.bind(this);
    this.forward = this.forward.bind(this);
    this.close = this.close.bind(this);
    this.slideshowCallback = this.slideshowCallback.bind(this);
    this.onMouseRest = this.onMouseRest.bind(this);
    
    if (ua.mobile) this.mouseRestSpeed = 5000;
    
    // this.render();
    this.cacheElements({
      audio: 'audio',
      categoryEl: '.category', // current category title
      categoriesDropdown: '.categories', // dropdown with event's categories
      count: '.count', // current position in the list
      slides: '.slides', // container for the slides 
      prevButton: '.prev',
      nextButton: '.next',
      musicToggle: '.toggle-music',
      slideshowControlsEl: '.slideshow-controls',
      slideshowSpeedEl: '.slideshow-speed',
      slideshowSpeedFill: '.slideshow-speed .fill',
      slideshowSpeedNob: '.slideshow-speed .nob',
      slideshowSpeedText: '.slideshow-speed .text',
      topRightEl: '.top-right',
    });
    
    this.mouseLastPosition = {
      x: null,
      y: null,
    };
    
    // number of previous and next photos to preload
    this.forwardPreloads = ua.desktop ? 10 : 3;
    this.backPreloads = ua.desktop ? 3 : 1;
    
    keyboard.bind('left', this.back);
    keyboard.bind('right', this.forward);
    keyboard.bind('esc', this.close);
    
    this.listenTo(app.state, 'resize', this.resize);
    this.listenTo(app.state, 'change:fullscreen', this.resize);
    
    this.setSlideshowSpeed(ua.desktop ? 5 : 4);
    
    this.audio.addEventListener('loadstart', function () {
      self.audioReady = false;
    }, false);
    this.audio.addEventListener('canplay', function () {
      self.audioReady = true;
    }, false);
    
    this.prev = []; // queue of photo views to the left 
    this.next = []; // queue of photo views to the right
    
    if (ua.mobile) this.initTouch();
    
    this.el.classList.add('cross-fade');
  },
  
  events: {
    'click .close': 'close',
    'click .next': 'forward',
    'click .prev': 'back',
    'click .toggle-music': 'toggleMusic',
    'click .toggle-slideshow': 'toggleSlideshow',
    'click .toggle-fullscreen': 'toggleFullscreen',
    'click .category': 'toggleCategories',
    
    // 'wheel': 'onWheel',
    // 'mousewheel': 'onWheel',
    
    'mousemove': 'onMouseMove',
    'mouseup': 'onMouseUp',
    
    'mousedown .slideshow-speed': 'speedDown',
    
    'touchstart': 'onTouchStart',
    'touchmove': 'onTouchMove',
    'touchend': 'onTouchEnd',
    'touchcancel': 'onTouchEnd',
  },
  
  bindings: {
    audioReady: {
      type: 'booleanClass',
      name: 'audio-ready',
    },
    slideshow: {
      type: 'booleanClass',
      name: 'slideshow',
    },
    music: {
      type: 'booleanClass',
      name: 'music-on',
    },
    hasMusic: {
      type: 'booleanClass',
      name: 'has-music',
    },
    mouseRest: {
      type: 'booleanClass',
      name: 'mouse-rest',
    },
    mouseOverSlideshow: {
      type: 'booleanClass',
      name: 'mouse-over-slideshow',
    },
    mouseOverTopRight: {
      type: 'booleanClass',
      name: 'mouse-over-top-right',
    },
    mouseOverActions: {
      type: 'booleanClass',
      name: 'mouse-over-actions',
    },
    noUI: {
      type: 'booleanClass',
      name: 'no-ui',
    },
    buyFormDisplayed: {
      type: 'booleanClass',
      name: 'buy-form-displayed',
    },
  },
  
  cleanup: function () { // clear the lightbox after hiding
    var i;
    for (i = 0; i < this.prev.length; i++) {
      this.prev[i].remove('now');
      this.prev[i].stopListening();
    }
    for (i = 0; i < this.next.length; i++) {
      this.next[i].remove('now');
      this.next[i].stopListening();
    }
    if (this.current) {
      this.current.remove('now');
      this.current.stopListening();
    }
    this.prev = []; // queue of photo views to the left 
    this.next = []; // queue of photo views to the right
    this.current = null; // view of the current photo
  },
  
  
  /*==========================================================================*\
    Reset the lightbox to a different list
  \*==========================================================================*/
  reset: function (ops) {
    if (ops.backTo) this.backTo = ops.backTo;
    
    this.cleanup();
    
    this.collection = ops.collection;
    
    this.current = new PhotoView({ 
      model: ops.photo || this.collection.at(0),
      parent: this,
    });
    this.preloading = true;
    
    // preload current, prev and next slides
    this.preloadMore({ inBackground: true });
    this.preloadMore({ inBackground: true });
    this.preloadMore({ inBackground: true });
    
    this.updateCount();
    this.updateDirectionalAreas();
    
    this.current.el.classList.add('current');
    this.updateSlides();
    
    var category = this.collection.parent;
    this.hasMusic = !!category.music;
    if (category.music) {
      if (category.music !== this.currentTrack) {
        this.audio.setAttribute('src', category.music);
        // this.audioReady = false;
        if (this.music) this.audio.play();
      }
      this.currentTrack = category.music;
    }
    else {
      this.currentTrack = null;
      if (this.music) this.toggleMusic();
    }
    
    this.categoryEl.innerHTML = category.name;
    
    this.categoriesDropdown.innerHTML = templates.lightbox.categories({ 
      categories: category.collection.models,
      active: category,
    });
    
    this.el.classList.remove('show-categories');
    
    if (ops.slideshow && !this.slideshow) {
      this.toggleSlideshow();
    }
    if (ops.slideshow && !app.state.fullscreen) {
      this.toggleFullscreen();
    }
  },
  
  
  /*==========================================================================*\
    Handle browser resize
  \*==========================================================================*/
  resize: function () {
    if (!this.displayed) return;
    var i;
    
    if (this.prev) { // update all currently handled slides
      for (i = 0; i < this.prev.length; i++) {
        this.prev[i].resize();
      }
      for (i = 0; i < this.next.length; i++) {
        this.next[i].resize();
      }
      this.current.resize();
    }
    
    this.updateDirectionalAreas();
  },
  
  
  
  /*==========================================================================*\
    Showing and hiding
  \*==========================================================================*/
  show: function () {
    if (this.displayed) return;
    this.displayed = true;
    this.el.classList.add('displayed');
    // this.current.showBuyForm();
  },
  hide: function () {
    if (!this.displayed) return;
    app.router.lightboxMode = false;
    this.displayed = false;
    if (this.slideshow) this.toggleSlideshow();
    fullscreen.release();
    
    this.el.classList.remove('displayed');
  },
  close: function () {
    if (this.buyFormDisplayed) return;
    app.navigate(this.backTo);
  },
  
  
  
  /*==========================================================================*\
    Navigating back and forward
  \*==========================================================================*/
  forward: function (options) {
    if (!this.displayed) return;
    options = options || {};
    
    if (options.preventDefault) options.preventDefault();
    
    if (this.buyFormDisplayed) return;
    
    if (this.current.model.collection !== this.next[0].model.collection) {
      this.collection = this.next[0].model.collection;
      var category = this.collection.parent;
      this.categoryEl.innerHTML = category.name;
      this.categoriesDropdown.innerHTML = templates.lightbox.categories({ 
        categories: category.collection.models,
        active: category,
      });
      // app.navigate(this.next[0].model.app_url, { trigger: true });
      // return;
    }
    
    this.el.classList[options.crossFade ? 'add' : 'remove']('cross-fade');
    this.current.el.classList.remove('current');
    
    if (this.prev.length) {
      this.prev[this.prev.length-1].remove();
      this.prev.pop();
    }
    this.prev.unshift(this.current);
    
    
    this.current = this.next.shift();
    
    this.preloadMore();
    
    // if (!this.next.length) this.preloadMore();
    // else this.updateSlides();
    
    this.current.el.classList.add('current');
    
    this.updateCount();
    this.updateDirectionalAreas();
    
    app.router.navigate(this.current.model.app_url);
    
    if (this.slideshow) {
      clearTimeout(this.slideshowTimer);
      this.slideshowTimer = setTimeout(this.slideshowCallback, 
          this.slideshowSpeed + (options.crossFade ? 1000 : 0));
    }
  },
  back: function (options) {
    if (!this.displayed) return;
    options = options || {};
    if (options.preventDefault) options.preventDefault();
    
    if (this.buyFormDisplayed) return;
    
    if (this.current.model.collection !== this.prev[0].model.collection) {
      this.collection = this.prev[0].model.collection;
      var category = this.collection.parent;
      this.categoryEl.innerHTML = category.name;
      this.categoriesDropdown.innerHTML = templates.lightbox.categories({ 
        categories: category.collection.models,
        active: category,
      });
    }
    
    this.el.classList[options.crossFade ? 'add' : 'remove']('cross-fade');
    this.current.el.classList.remove('current');
    
    if (this.next.length) {
      this.next[this.next.length-1].remove();
      this.next.pop();
    }
    this.next.unshift(this.current);
    
    this.current = this.prev.shift();
    
    if (!this.prev.length) this.preloadMore();
    else this.updateSlides();
    
    this.current.el.classList.add('current');
    
    this.updateCount();
    this.updateDirectionalAreas();
    
    app.router.navigate(this.current.model.app_url);
    
    if (this.slideshow) {
      clearTimeout(this.slideshowTimer);
      this.slideshowTimer = setTimeout(this.slideshowCallback, 
          this.slideshowSpeed + (options.crossFade ? 1000 : 0));
    }
  },
  
  
  /*==========================================================================*\
    Updating the DOM
  \*==========================================================================*/
  updateSlides: function () { // keep only 3 slides in the DOM 
    var i;
    for (i = 0; i < this.prev.length; i++) {
      if (i === 0 && !this.prev[i].el.parentNode) {
        this.slides.appendChild(this.prev[i].el);
      }
      else if (i > 0 && this.prev[i].el.parentNode) {
        this.slides.removeChild(this.prev[i].el);
      }
    }
    for (i = 0; i < this.next.length; i++) {
      if (i === 0 && !this.next[i].el.parentNode) {
        this.slides.appendChild(this.next[i].el);
      }
      else if (i > 0 && this.next[i].el.parentNode) {
        this.slides.removeChild(this.next[i].el);
      }
    }
    if (this.current && !this.current.el.parentNode) {
      this.slides.appendChild(this.current.el);
    }
    
    if (ua.mobile) {
      clearTimeout(this.mouseRestTimer);
      this.mouseRestTimer = setTimeout(this.onMouseRest, this.mouseRestSpeed);
    }
  },
  
  
  /*==========================================================================*\
    Photos Preloading
  \*==========================================================================*/
  preloadMore: function (options) { // preload a photo if needed
    options = options || {};
    if (!this.displayed && !options.inBackground) return;
    if (this.preloading && this.prev.length && this.next.length) {
      return this.updateSlides();
    }
    // var which = options.which || 'next';
    var which = !this.next.length || this.prev.length ? 'next' : 'prev';
    if (which === 'next' && this.next.length >= 45) {
      this.preloading = false;
      return this.updateSlides();
    }
    this.preloading = true;
    // if (which === 'next' && this.next.length >= 2) return this.updateSlides();
    
    var list = this[which];
    var farthest = list.length ? list[list.length-1] : this.current;
    var model = farthest.model[which === 'next' ? 'nextAround' : 'prevAround'];
    var photo = new PhotoView({
      model: model,
      parent: this,
    });
    list.push(photo);
    
    this.updateSlides();
    
    if (this.slideshow && this.slidePending) {
      this.slideshowCallback();
    }
  },
 
  
  
  
  /*==========================================================================*\
    UI Updates
  \*==========================================================================*/
  updateCount: function () { // list counter
    this.count.innerHTML = (this.collection.indexOf(this.current.model) + 1) +
        ' / '+ this.collection.length;
  },
  updateDirectionalAreas: function () { // update coverage of prev/next areas
    var prevWidth = (app.state.width - this.current.imageWidth) / 2;
    if (prevWidth < 43) prevWidth = 43;
    this.prevButton.style.right = (app.state.width - prevWidth) + 'px';
    this.nextButton.style.left = prevWidth + 'px';
  },
  
  // render: function () {
  // },
});
