// app/views/lightbox/photo.js
/*==========================================================================*\
  A single photo with its UI elements in the lightbox
\*==========================================================================*/
"use strict";

var app = require('app');
var ua = require('app/utils/ua');

var Parent = require('app/views/base-view');
var BuyForm = require('./buy-form/buy-form');
var language = require('app/language');

module.exports = Parent.extend(require('bind-transforms')).extend({
  template: require('app/templates').lightbox.photo,
  loaded: false, // whether the main image has been loaded 
  failures: 0, // number of times main image request has failed 
  language: language,
  imageWidth: 0, // dimensions of an image as currently displayed 
  imageHeight: 0, 
  
  props: {
    inCart: ['boolean', true, false],
  },
  
  bindings: {
    inCart: {
      type: function (el, val) {
        this.el.classList[val ? 'add' : 'remove']('in-cart');
        this.query('.action.cart .tip').innerHTML = val ? 'In Cart' : 'Buy Image';
      },
      // type: 'booleanClass',
      // name: 'in-cart',
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.onImageTransitionEnd = this.onImageTransitionEnd.bind(this);
    this.load();
    this.render();
    this.cacheElements({
      imagesEl: '.images',
      img: '.img',
      imgTag: 'img',
      actionsEl: '.actions',
    });
    this.img.addEventListener(ua.events.transitionEnd, this.onImageTransitionEnd);
    this.resize();
  },
  
  events: {
    'click .action.cart': 'showBuyForm',
  },
  
  load: function () {
    if (!this.image) {
      this.image = new Image({language:language});
      this.image.addEventListener('load', this, false);
      this.image.addEventListener('error', this, false);
    }
    this.image.src = this.model.img_url;
  },
  
  handleEvent: function (e) {
    switch (e.type) {
      case 'load':
        this.loaded = true;
        this.trigger('load');
        this.el.classList.add('loaded');
        this.parent.preloading = false;
        this.parent.preloadMore();
        
        this.rq = null;
        this.image.removeEventListener('load', this, false);
        this.image.removeEventListener('error', this, false);
        this.image = null;
        break;
      case 'error':
        console.log('failed to load photo: '+ this.model.img_url, e);
        if (this.failures < 5) {
          this.image.src = this.model.img_url + 
              (this.model.img_url.indexOf('?') > -1 ? '&x=' : '?') + Date.now();
        }
        else {
          this.failed = true;
          this.trigger('error');
          this.rq = null;
        }
        this.failures++;
        break;
    }
  },
  
  resize: function (ops) {
    ops = ops || {};
    var pWidth, pHeight;
    var screenRatio = app.state.width / app.state.height;
    if (app.state.fullscreen || ua.mobile) {
      pHeight = app.state.height;
      pWidth = app.state.width;
    }
    else {
      pHeight = app.state.height - 88 - (this.model.name ? 54 : 0);
      pWidth = app.state.width - 108;
    }
    if (this.parent.buyFormDisplayed) {
      pWidth = Math.floor(app.state.width / 2);
      pHeight = app.state.fullscreen || ua.mobile ? app.state.height : app.state.height - 60;
    }
    var spaceRatio = pWidth / pHeight;
    if (this.model.ratio > spaceRatio) {
      this.imageWidth = pWidth;
      this.imageHeight = this.imageWidth / this.model.ratio;
    }
    else {
      this.imageHeight = pHeight;
      this.imageWidth = this.imageHeight * this.model.ratio;
    }
    if (this.imageWidth > this.model.orig_width ||
        this.imageHeight > this.model.orig_height) {
      this.imageWidth = this.model.orig_width;
      this.imageHeight = this.model.orig_height;
    }
    if (!ops.notReally) {
      this.imgTag.style.top = 0;
      this.imgTag.style.left = 0;
    }
    if (!this.parent.buyFormDisplayed && (app.state.fullscreen || ua.mobile)) {
      if (screenRatio > this.model.ratio && 
          app.state.width - this.imageWidth < 88) {
        this.imageWidth = app.state.width;
        this.imageHeight = Math.ceil(this.imageWidth / this.model.ratio);
        if (!ops.notReally) {
          this.imgTag.style.top = Math.ceil((
              app.state.height - this.imageHeight) / 2) + 'px';
        }
      }
      else if (screenRatio < this.model.ratio && 
               app.state.height - this.imageHeight < 108) {
        this.imageHeight = app.state.height;
        this.imageWidth = Math.ceil(this.imageHeight * this.model.ratio);
        if (!ops.notReally) {
          this.imgTag.style.left = Math.ceil((app.state.width - this.imageWidth) / 2) + 'px';
        }
      }
    }
    
    if (this.buyForm && this.parent.buyFormDisplayed) {
      var formWidth = app.state.width - this.imageWidth - 90;
      var formLeft = 30;
      this.imagesRight = this.parent.buyFormDisplayed ? 30 :
                         app.state.fullscreen || ua.mobile ? 0 : 54;
      if (formWidth > 650) {
        var excessSpace = formWidth - 650;
        formWidth = 650;
        formLeft += Math.ceil(excessSpace / 2);
        this.imagesRight += Math.floor(excessSpace / 2);
      }
      this.buyForm.el.style.width = formWidth + 'px';
      this.buyForm.el.style.left = formLeft + 'px';
    }
    else {
      this.imagesRight = app.state.fullscreen || ua.mobile ? 0 : 54;
    }
    
    
    if (!ops.notReally) {
      this.img.style.width = this.imageWidth +'px';
      this.img.style.height = this.imageHeight +'px';
      this.imgTag.style.width = this.imageWidth +'px';
      this.imgTag.style.height = this.imageHeight +'px';
      this.imagesEl.style.right = this.imagesRight +'px';
      
      // if (this.buyForm) {
      //   this.buyForm.el.style.width = app.state.width - this.imageWidth - 90 + 'px';
      // }
    }
    
  },
  
  remove: function () {
    Parent.prototype.remove.apply(this);
    this.imgTag.removeAttribute('src');
    this.el.innerHTML = '';
    this.stopListening();
    if (this.rq) this.rq.abort();
    if (this.image) {
      this.image.removeEventListener('load', this, false);
      this.image.removeEventListener('error', this, false);
      this.image.removeAttribute('src');
      this.image = null;
    }
  },
  
  showBuyForm: function () {
    if (!this.buyForm) {
      this.buyForm = new BuyForm({
        model: this.model,
        parent: this,
        // el: this.query('.buy-form'),
      });
      this.el.appendChild(this.buyForm.el);
      
      this.buyForm.updatePositions();
    }
    else {
      this.buyForm.saveAsDefault = false;
    }
    
    this.buyForm.el.classList.remove('hidden');
    this.parent.buyFormDisplayed = true;
    
    var imgRect = this.img.getBoundingClientRect();
    this.resize({ notReally: true });
    var newLeft = app.state.width - this.imagesRight - this.imageWidth;
    var newTop = (app.state.height - this.imageHeight) / 2;
    
    var scaleX = this.imageWidth / imgRect.width;
    var scaleY = this.imageHeight / imgRect.height;
    
    this.img.classList.remove('no-transition');
    this.img.style[ua.css.transform] = 
        'translate('+ (newLeft - imgRect.left) +'px, '+ 
        (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
  },
  
  hideBuyForm: function () {
    if (!this.parent.buyFormDisplayed) return;
    
    this.parent.buyFormDisplayed = false;
    
    var imgRect = this.img.getBoundingClientRect();
    this.resize({ notReally: true });
    var newLeft = (app.state.width - this.imageWidth) / 2;
    var newTop = (app.state.height - this.imageHeight - 
                 (this.model.name ? 54 : 0)) / 2;
    // var newTop = (app.state.height - this.imageHeight) / 2;
    
    var scaleX = this.imageWidth / imgRect.width;
    var scaleY = this.imageHeight / imgRect.height;
    
    this.img.classList.remove('no-transition');
    this.img.style[ua.css.transform] = 
        'translate('+ (newLeft - imgRect.left) +'px, '+ 
        (newTop - imgRect.top) +'px) scale('+scaleX+', '+scaleY+')';
  },
  
  
  onImageTransitionEnd: function () {
    this.imagesEl.classList[this.parent.buyFormDisplayed ? 'add' : 'remove'](
        'pushed-right');
    if (!this.parent.buyFormDisplayed) this.buyForm.el.classList.add('hidden');
    this.img.classList.add('no-transition');
    this.img.style[ua.css.transform] = 'none';
    this.resize();
    
    if (!this.parent.buyFormDisplayed && !this.inCart) {
      this.buyForm.remove();
      this.buyForm.stopListening();
      this.buyForm = null;
    }
  },
});
