// app/views/lightbox/purchase-addon-options.js
/*==========================================================================*\
  A purchase option in the buy form
\*==========================================================================*/
"use strict";

var app = require('app');
var ua = require('app/utils/ua');
var Parent = require('app/views/base');
var popup = require('app/views/popup');
var OptionView = require('app/views/lightbox/purchase-addon-option');

module.exports = Parent.extend({
  template: require('app/templates').lightbox['purchase-addon-options'],
  
  props: {
    // active: ['boolean', true, false],
    // price: ['number', true, 0],
  },
  
  bindings: {
    // active: {
    //   type: 'booleanClass',
    //   name: 'active',
    // },
    // price: {
    //   type: function (el, val) {
    //     el.innerHTML = '+' + this.model.formatMoney(val);
    //   },
    //   selector: '.price',
    // },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.render();
  },
  
  events: {
    'click': 'onClick',
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.renderCollection(this.collection, OptionView, '.options');
  },

});
