// app/product/product-model.js
/*==========================================================================*\
  Product Model 
\*==========================================================================*/
"use strict";

var BaseModel = require('app/base/base-model');
var OptionsCollection = require('./option/product-options-collection');
var whRegExp = new RegExp("w=\\d+&h=\\d+", 'ig');

module.exports = BaseModel.extend({
  urlRoot: '/api/products',
  props: {
    id: 'number',
    name: ['string', true, ''],
    description: ['string', true, ''],
    
    img_url: ['string', true, ''],   
    img_width: ['number', true, 0],  
    img_height: ['number', true, 0], 
    
    'sales_tax': ['boolean', true, true], // sales tax should be charged 
    
    // product|gift_cert|album|full_download
    type: ['string', true, 'product'], 
    
    release: 'array',
  },

  collections: {
    options: OptionsCollection,
  },

  derived: {
    includesPrints: {
      fn: function () {
        var option;
        for (var i = 0; i < this.options.length; i++) {
          option = this.options.at(i);
          if (option.required_size_list.length) return true;
        }
        return false;
      },
    }
  },
  
  getImgForWidth: function (width) {
    if (!this.img_url || !this.img_width || !this.img_height) return false;
    width = width || 150;
    var height = Math.round(this.img_height / this.img_width * width);
    return {
      width: width,
      height: height,
      url: this.img_url.replace(whRegExp, '') +'&w='+ width +'&h=' + height,
    };
  },
});
