// app/product/product-item.js
/*==========================================================================*\
  A single product on the event't products listing
\*==========================================================================*/
"use strict";

var app = require('app');
var BaseView = require('app/base/base-view');
var PopupView = require('./product-popup');

module.exports = BaseView.extend({
  viewType: 'product-item',
  template: require('app/templates').product['product-item'],
  
  initialize: function () {
    BaseView.prototype.initialize.apply(this, arguments);
    this.listenTo(app, 'reload-images', this.render);
  },
  
  events: {
    click: 'onClick',
  },
  
  onClick: function (e) {
    if (e && e.preventDefault) e.preventDefault();
    
    var popup = new PopupView({
      model: this.model,
      eventId: this.getPage().eventId,
    });
    
    document.body.insertBefore(popup.el, document.getElementById('popup'));
    // app.view.overlays.appendChild(popup.el);
    
    return false;
  },
});
