// app/models/cart/addon-model.js
/*==========================================================================*\
  A Photo Purchase Addon like paper type or UV protection
\*==========================================================================*/
"use strict";

module.exports = require('app/models/base-model').extend({
  props: {
    id: 'number',
    name: ['string', true, ''], // name to display
    description: ['string', true, ''], // an optional description
    price: ['number', true, 0], // price of an addon in cents
  },
  
  collections: {
    options: require('./addon-options-collection'),
  },
  
  derived: {
    display_price: {
      deps: ['price'],
      fn: function () {
        return this.formatMoney(this.price);
      },
    },
    
    selected_name: {
      deps: ['options', 'name'],
      fn: function () {
        return this.default_option ? this.default_option.name : this.name;
      },
    },
    
    default_option: {
      deps: ['options'],
      fn: function () {
        return this.options.where({ selected: true })[0];
      },
    }
  },
  
  initialize: function () {
    if (this.options.length && !this.default_option) {
      this.options.add({
        id: 0,
        name: 'None', // name to display
        selected: true, // whether the item a default one 
      }, {
        at: 0,
      });
    }
  },
});
