// app/models/cart/cart-addon-model.js
/*==========================================================================*\
  A photo's purchase addon in cart
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/models/base-model');

module.exports = Parent.extend({
  // idAttribute: 'id',
  props: {
    id: 'number', // ID of the addon
    oid: 'number', // selected option ID
  },
  derived: {
    cart: {
      fn: function () {
        return this.collection.parent.cart;
      },
    },
    addon: {
      deps: ['id'],
      fn: function () {
        return app.purchaseAddons.get(this.id);
      },
    },
    price: {
      cache: false,
      // deps: ['oid', 'id'],
      fn: function () {
        var sizeId = this.collection.parent.size_id;
        if (this.addon.options.length <= 1) {
          if (this.addon.options.length) {
            return this.addon.options.at(0).getPrice(sizeId);
          }
          else {
            return this.addon.price;
          }
        }
        else {
          return this.selected_id ? 
                  this.addon.options.get(this.oid).getPrice(sizeId) : 0;
        }
        
        return this.addon.price;
        // return this.oid ? this.addon.options.get(this.oid).price 
        //                 : this.addon.price;
      },
    },
    shipping_price: {
      // deps: ['oid', 'id'],
      cache: false,
      fn: function () {
        var option = this.addon.options.get(this.oid);
        var size = this.collection.parent;
        var shipId = this.cart.shipping_type;
        return option.sizes[size.size_id].shipping.get(shipId).price;
      },
    },
  },
});
