// app/models/cart/cart-model.js
/*==========================================================================*\
  Shopping Cart
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/models/base-model');
var PhotosCollection = require('./cart-photos-collection');

var _ = require('underscore');

module.exports = Parent.extend({
  props: {
    id: 'number',
    eid: 'number',
    email: 'string',
    mtime: ['number', true, 0], // last modification time
    shipping_type: 'number',
  },
  
  collections: {
    photos: PhotosCollection,
    products: PhotosCollection,
  },
  
  session: {
    event_name: 'string',
    subtotal: ['number', true, 0],
    discount: ['number', true, 0],
    shipping_total: ['number', true, 0],
    total: ['number', true, 0],
  },
  
  derived: {
    event: {
      fn: function () {
        return app.eventModels[this.eid];
      },
    },
    is_empty: {
      deps: ['mtime'],
      fn: function () {
        return !this.photos.length && !this.products.length;
      },
    },
    shipping: {
      deps: ['shipping_type'],
      fn: function () {
        return app.shipping.get(this.shipping_type);
      },
    },
  },
  
  initialize: function () {
    // this.listenTo(this.photos, 'add remove change', this.store);
    this.listenTo(this.event, 'change:complete', function () {
      this.event_name = this.event.name;
    });
    
    this.listenTo(app, 'storage:cart'+this.eid, this.storageCallback);
    
    this.listenTo(this.photos, 'add', function (photo) {
      this.trigger('add:photo:'+ photo.pid);
    });
    this.listenTo(this.photos, 'remove', function (photo) {
      this.trigger('remove:photo:'+ photo.pid);
    });
    
    // this.listenTo(this, 'change', this.updateSummary);
    
    var storedData = app.storage['cart'+this.eid];
    if (storedData) this.set(JSON.parse(storedData));
    
    this.store = _.throttle(this.store, 1000);
  },
  
  storageCallback: function (newData) {
    newData = JSON.parse(newData);
    if (newData.mtime > this.mtime) {
      this.set(newData);
    }
  },
  
  store: function () {
    console.log('storing cart', 
        JSON.stringify(this.toJSON()).length * 2 + ' bytes');
    this.mtime = Date.now();
    app.storage['cart'+this.eid] = JSON.stringify(this.toJSON());
  },
  
  save: function () {
    this.updateSummary();
    
    
    
    this.store();
  },
  
  getShippingCost: function () { // only called from updateSummary
    var finalCost = 0; // final shipping cost
    var i, j, k;
    var photo;
    var size;
    var price;
    var maxSizeCost = 0; // max shipping cost for a size 
    var maxAnyCost = 0; // max shipping cost for a size/addon/product option
    var addonsCost = 0; // total addons shipping cost 
    
    for (i = 0; i < this.photos.length; i++) {
      photo = this.photos.at(i);
      for (j = 0; j < photo.options.length; j++) {
        size = photo.options.at(j);
        price = size.shipping_price;
        
        if (price > maxSizeCost) maxSizeCost = price;
        if (price > maxAnyCost) maxAnyCost = price;
        
        for (k = 0; k < size.addons.length; k++) {
          price = size.addons.at(k).shipping_price;
          addonsCost += price;
          if (price > maxAnyCost) maxAnyCost = price;
        }
      }
    }
    
    // fixed cost set per shipping method relative to total
    var baseCost = this.shipping.criteria[0].shipping;
    
    var printsTotal = 0; 
    for (i = 0; i < this.photos.length; i++) {
      printsTotal += this.photos.at(i).subtotal;
    }
    for (i = 0; i < this.shipping.criteria; i++) {
      if (app.config.shipping_method === 'order_percent') {
        if (this.subtotal >= this.shipping.criteria[i].total) break;
      }
      else {
        if (printsTotal >= this.shipping.criteria[i].total) break;
      }
      baseCost = this.shipping.criteria[i].shipping;
    }
    
    switch (app.config.shipping_method) {
      case 'prints_ordered': finalCost = maxSizeCost + addonsCost;
      break;
      case 'items_ordered': finalCost = maxAnyCost;
      break;
      case 'print_total': finalCost = baseCost + addonsCost;
      break;
      case 'order_total': finalCost = baseCost;
      break;
      case 'print_percent': finalCost = printsTotal * baseCost + addonsCost;
      break;
      case 'order_percent': finalCost = this.subtotal * baseCost;
      break;
    }
    
    if (finalCost < this.shipping.min) finalCost = this.shipping.min;
    if (finalCost > this.shipping.max) finalCost = this.shipping.max;
    
    return finalCost;
  },
  
  updateSummary: function () {
    var i, j;
    var subtotal = 0; 
    for (i = 0; i < this.photos.length; i++) {
      subtotal += this.photos.at(i).subtotal;
    }
    this.subtotal = subtotal;
    
    this.shipping_total = this.getShippingCost();
    this.total = subtotal - this.discount + this.shipping_total;
  },
  
  // parse: function (res) {
  //   // Parent.parse.apply(this, arguments);
  // },
  
});
