// app/models/cart/cart-photo-model.js
/*==========================================================================*\
  A Photo in cart 
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/models/base-model');

var Options = require('app/models/base-collection').extend({
  mainIndex: 'size_id',
  model: require('./cart-option-model'),
});

module.exports = Parent.extend({
  idAttribute: 'pid',
  extraProperties: 'ignore',
  
  // prevent ampersand from sorting the cart photos collection 
  isNew: function () { return false; },
  
  props: {
    cat_id: 'number', // photo's category ID
    pid: 'number', // original ID of the event's photo
    color_id: 'number', // color variant ID 
    
    // options: {
    //   type: 'object',
    //   default: function () {
    //     return {};
    //   },
    // },
  },
  
  collections: {
    options: Options,
  },
  
  derived: {
    
    eid: {
      fn: function () {
        return this.cart.eid;
      },
    },
    cart: {
      fn: function () {
        return this.collection.parent;
      },
    },
    photo: {
      // cache: false,
      fn: function () {
        // if (!app.eventModels[this.eid].categories.length) return null;
        if (!app.eventModels[this.eid]) {
          console.log('AAAA!', this.toJSON(), app.eventModels);
          console.trace();
        }
        return app.eventModels[this.eid].categories.get(this.cat_id).
               photos.get(this.pid);
      }
    },
    url: {
      fn: function () {
        return '#events/'+ this.eid +'/'+ this.cat_id +'/'+ this.pid +'buy';
      },
    },
    subtotal: {
      cache: false,
      fn: function () {
        // if (!this.photo) return 0;
        var i;
        var subtotal = 0;
        for (i = 0; i < this.options.length; i++) {
          subtotal += this.options.at(i).price;
        }
        return subtotal;
      },
    },
  },
  
  // parse: function (res) {
  // },
  
});
