// app/views/cart/cart-photo.js
/*==========================================================================*\
  A single photo in the cart 
\*==========================================================================*/
"use strict";

var app = require('app');

var Parent = require('app/views/base-view');
var OptionView = require('./cart-photo-option');

module.exports = Parent.extend(require('bind-transforms')).extend({
  template: require('app/templates').cart['cart-photo'],
  
  props: {
    imageWidth: 'number', // dimensions of an image as currently displayed 
    imageHeight: 'number',
    
    currentCrop: 'object',
  },
  
  bindings: {
    imageWidth: {
      type: function (el, val) {
        el.style.width = val + 'px';
      },
      selector: '.img',
    },
    imageHeight: {
      type: function (el, val) {
      el.classList.toggle('small', val <= 90);
        el.style.height = val + 'px';
      },
      selector: '.img',
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.render();
    this.imageWidth = 100;
    this.imageHeight = 100 / this.model.photo.ratio;
    this.cacheElements({
    });
  },
  
  events: {
  },
  
  render: function () {
    this.renderWithTemplate(this);
    this.renderCollection(this.model.options, OptionView, this.query('.items'));
  },
});
