// app/fake-api/fake-config.js
/*==========================================================================*\
  Fake Main App Config
\*==========================================================================*/
module.exports = {
  // grid|list|login: what the visitors see when they visit the homepage
  style: 'grid',
  
  // date|name|type : Determines how the events are sorted by default
  sort: 'date', 
  
  // current|future|all : Determines which events are displayed by default
  filter: 'all', 
  
  // Boolean value that indicates whether the search area should be displayed
  display_search: true,
  
  // Boolean value that indicates whether the "Home" link should be displayed 
  display_home: false, 
  
  // Boolean value that indicates whether the "Home" link should be displayed
  display_custom: false, 
  
  // The text that should be displayed for the custom link
  display_custom_label: '',
  
  // The URL that clicking the custom link should take the user to 
  display_custom_url: '', 
  
  // whether clicking the logo should take visitor to the 
  // photographer's personal website
  link_to_company: false, 
  
  // indicates whether the products should be shown above (1) 
  // or below (0) the image categories within the event
  products_on_top: false, 
  
  // whether product names should be shown over the images in thumbnail view
  product_names: true,
  
  // none|view|custom|view_custom : Determines whether the visitor can 
  // select a crop for their ordered prints 
  custom_crop: 'none',
  
  // whether the storefront is in vacation mode. When in vacation mode, 
  // visitors can view the homepage as normal, but when an event is 
  // tried to be entered the user is shown a vacation message, is asked 
  // for their email address, and cannot view the images within the event 
  vacation: false, 
  
  // The message to display when the storefront is in vacation mode and a 
  // person tries to enter an event
  vacation_message: '', 
  
  // whether visitors can see the "Email Photo" option with the share links
  email_photo: false, 
  
  // Numeric value used to determine which sharing links 
  // (Facebook, Twitter, Digg, etc.) should be displayed with the images
  share_options: 0,
  
  // Displayed on the purchase page with options 
  // that have add-ons available for them 
  addon_label: '',
  
  // The minimum order size allowed before the "Checkout" 
  // option should be displayed 
  min_order: 1, 
  
  // Determines whether American Express card 
  // will be accepted on large orders (<$2000)
  allow_amex: true, 
  
  // Message to display if a person enters an event that contains no images
  empty_message: '', 
  
  // Message to display if a person enters an event that has expired
  expired_message: '', 
  
  // Message displayed on the receipt after an order is placed
  receipt_message: '', 
  
  // Description used in the <head> portion of the website
  meta_description: '', 
  
  // Text to display in place of "Sales Tax" on the checkout and receipt pages
  tax_label: '', 
};
