// app/fake-api/fake-purchase-addons.js
/*==========================================================================*\
  Fake purchase options, such as paper types, web-rights etc.
\*==========================================================================*/
var uid = 1;

module.exports = [
  { 
    id: 1,
    name: 'Paper Type',
    options: [
      {
        id: uid++,
        name: 'Matte Paper',
        description: 'Unlike glossy paper, matte paper produces nice quality '+
            'prints but does not have that vibrant color finish. Matte paper '+
            'is not vulnerable to being marred by fingerprints and it produces '+
            'a non-glare photo. ',
        img_url: '/img/purchase-options/matte.jpg',
        img_width: 300,
        img_height: 200,
        price: 0,
        selected: true,
      },
      {
        id: uid++,
        name: 'Deep Matte Paper',
        description: 'This lusterless surface has a unique visual quality '+
            'which produces incredibly elegant, warm images with most '+
            'accurate color reproduction and pure whiteness. '+
            'Great with soft, light images, skin tones, and black and whites.',
        img_url: '/img/purchase-options/deep.jpg',
        img_width: 300,
        img_height: 200,
        price: 20,
      },
      {
        id: uid++,
        name: 'Metallic Paper',
        description: 'The gloss finish and metallic appearance creates images '+
            'with exceptionally unique visual interest & depth. '+
            'It delivers rich, vibrant colors and flattering flesh tones.',
        img_url: '/img/purchase-options/metallic.jpg',
        img_width: 300,
        img_height: 200,
        price: 20,
      },
      {
        id: uid++,
        name: 'Luster Paper',
        description: 'This semi-gloss paper provides more vivid color '+
            'reproduction, plus improved highlight detail.',
        img_url: '/img/purchase-options/luster.jpg',
        img_width: 300,
        img_height: 200,
        price: 20,
      },
    ],
  },
  { 
    id: 2,
    name: 'Web Use Rights',
    description: 'License to use this image on the internet.',
    price: 150, // price of an option in cents
  },
];
