// app/fake-api/fake-purchase-options.js
/*==========================================================================*\
  Fake purchase options, such as sizes, web use rights etc.
\*==========================================================================*/
var uid = 1;

module.exports = [
  { 
    id: uid++,
    type: 'digital',
    name: 'Instant Download',
    price: 50, // price of an option in cents
    no_qty: true,
    addon_ids: [2],
  },
  { 
    id: uid++,
    type: 'print',
    price: 100, // price of an option in cents
    width: 4,  // width of a print
    height: 6, // height of a print
    addon_ids: [1,3],
  },
  { 
    id: uid++,
    type: 'print',
    price: 110, // price of an option in cents
    width: 5,  // width of a print
    height: 7, // height of a print
    addon_ids: [1,3],
  },
  { 
    id: uid++,
    type: 'print',
    price: 120, // price of an option in cents
    width: 8,  // width of a print
    height: 10, // height of a print
    addon_ids: [1,3],
  },
  { 
    id: uid++,
    type: 'print',
    price: 130, // price of an option in cents
    width: 11,  // width of a print
    height: 14, // height of a print
    addon_ids: [1,3],
  },
  { 
    id: uid++,
    type: 'print',
    price: 150, // price of an option in cents
    width: 16,  // width of a print
    height: 20, // height of a print
    addon_ids: [1],
  },
  { 
    id: uid++,
    type: 'print',
    price: 200, // price of an option in cents
    width: 20,  // width of a print
    height: 24, // height of a print
    addon_ids: [1],
  },
  { 
    id: uid++,
    type: 'print',
    name: '8 Wallets',
    full_name: '8 wallet-sized 2x3.5" prints',
    img_url: '/img/purchase-options/wallets.jpg',
    img_width: 300,
    img_height: 300,
    description: 
        "Keep your favorite shots ready to be bragged about at "+
        "a moment's notice.",
    price: 150, // price of an option in cents
    width: 2,  // width of a print
    height: 3.5, // height of a print
  },
  { 
    id: uid++,
    type: 'print',
    name: '12 Wallets',
    full_name: '12 wallet-sized 2x3.5" prints',
    img_url: '/img/purchase-options/wallets.jpg',
    img_width: 300,
    img_height: 300,
    description: 
        "Keep your favorite shots ready to be bragged about at "+
        "a moment's notice.",
    price: 200, // price of an option in cents
    width: 2,  // width of a print
    height: 3.5, // height of a print
  },
  { 
    id: uid++,
    type: 'print',
    name: '8" x 10" Canvas',
    price: 1000, // price of an option in cents
    width: 8,  // width of a print
    height: 10, // height of a print
  },
  { 
    id: uid++,
    type: 'print',
    name: '11" x 14" Canvas',
    price: 1200, // price of an option in cents
    width: 11,  // width of a print
    height: 14, // height of a print
  },
];
