// app/fake-api/fake-photographer.js
/*==========================================================================*\
  Fake Photographer's Info
\*==========================================================================*/
module.exports = {
  company_name: 'Brand Name Photography', // The name of the company
  
  // contact person (generally the photographer)
  // first_name: '',
  // last_name: '',
  
  email: 'info@instaproofs.com', // The photographer's email address
  phone: '(435) 258-6197', // The photographer's phone number
  storefront_url: 'http://www.instaproofs.com', // photographer's storefront
  facebook_url: 'http://www.facebook.com', // The company's facebook URL
  twitter_url: 'http://www.twitter.com', // The company's twitter URL
  linkedin_url: 'http://www.linkedin.com', // The company's linkedin URL
  instagram_url: 'http://www.instagram.com', // The company's instagram URL
  website_url: 'http://www.instaproofs.com', // The company's website URL
  address: 'PO BOX 14',
  city: 'St. George', // The city the company is located in
  state: 'Utah', // The state the company is located in
  zip: '84771',
  country: 'USA',
};
