// app/fake-api/fake-events.js
/*==========================================================================*\
  Fake Events Data
\*==========================================================================*/
"use strict";

var eventID = 1;
var catID = 1;
var photoID = 1;
var uid = 0;

var events = [];
// photos are separated as they are not included for events listing
var photos = {};

// 68 male and 68 female names 
var names1 = ['Noah','Liam','Jacob','Mason','William','Ethan','Michael', 
'Alexander','Jayden','Daniel','Elijah','Aiden','James','Benjamin','Matthew',
'Jackson','Logan','David','Anthony','Joseph','Joshua','Andrew','Lucas',
'Gabriel','Samuel','Christopher','John','Dylan','Isaac','Ryan','Nathan',
'Carter','Caleb','Luke','Christian','Hunter','Henry','Owen','Landon','Jack',
'Wyatt','Jonathan','Eli','Isaiah','Sebastian','Jaxon','Julian','Brayden',
'Gavin','Levi','Aaron','Oliver','Jordan','Nicholas','Evan','Connor','Charles',
'Jeremiah','Cameron','Adrian','Thomas','Robert','Tyler','Colton','Austin',
'Jace','Angel','Dominic'];
var names2 = ['Sophia','Emma','Olivia','Isabella','Ava','Mia','Emily','Abigail',
'Madison','Elizabeth','Charlotte','Avery','Sofia','Chloe','Ella','Harper',
'Amelia','Aubrey','Addison','Evelyn','Natalie','Grace','Hannah','Zoey',
'Victoria','Lillian','Lily','Brooklyn','Samantha','Layla','Zoe','Audrey','Leah',
'Allison','Anna','Aaliyah','Savannah','Gabriella','Camila','Aria','Kaylee',
'Scarlett','Hailey','Arianna','Riley','Alexis','Nevaeh','Sarah','Claire',
'Sadie','Peyton','Aubree','Serenity','Ariana','Genesis','Penelope','Alyssa',
'Bella','Taylor','Alexa','Kylie','Mackenzie','Caroline','Kennedy','Autumn',
'Lucy','Ashley','Madelyn'];
// names2.pop(); // remove one so that combinations will shift on iteration

var types = ['',
  'Wedding',
  'Bridal Shower',
  'Marriage Proposal',
  'Birthday',
  'Concert',
  'Reunion'];

var catTitles = ['Bride & Groom', 'Family Formals', 'Rose Ceremony',
      'Declaration of Intent',
      'Lighting of the Unity Candle', 'Exchange of Vows', 
      'Bridal Processional', 'Exchange of Rings', 'Recessional'];
var photoTitles = ['Bride & Groom', 'Family Formals', 'Rose Ceremony',
    'Declaration of Intent',
    'Lighting of the Unity Candle', 'Exchange of Vows', 
    'Bridal Processional', 'Exchange of Rings', 
    'A pretty long title of a photo to see how it goes here'];
var tracks = [
  // '/audio/Sketch-7.mp3',
  '/audio/01_Sleepwalker.mp3',
  '/audio/05_Morning_Talk_Supersymmetry.mp3',
  '/audio/06_Some_Other_Place.mp3',
  '/audio/07_Song_On_The_Beach.mp3',
  '/audio/13_Dimensions.mp3',
  // '/audio/between_the_lines.mp3',
  '/audio/days_to_come.mp3',
  '/audio/going_brown_again.mp3',
  // '/audio/so-it-goes.mp3',
  '/audio/steel_for_pappa.mp3',
  '/audio/tomorrow_untrodden.mp3',
  '/audio/track1.mp3',
];   

var date = new Date('12 February 2015 14:48 UTC');
function pad(number) { return number < 10 ? '0' + number : number; }
function toISO(date) {
  return date.getUTCFullYear() +'-'+ pad(date.getUTCMonth() + 1) +'-'+ 
         pad(date.getUTCDate());
}

function createEvent(index) {
  var img = index % 19 + 1;
  var ev = {
    id: eventID++,
    img_url: '/img/pics/img'+ img +'.jpg?'+ (uid++),
    city: 'New York', 
    state: 'NY', 
    country: 'US',
    show: true,
    display_image: true, 
    showcase: true,
    
    categories: [],
    
    // alias: '', 
    // future: false, 
  };
  switch (img) {
    case 11: ev.type_id = 5; break; // Concert
    case 12: case 13: ev.type_id = 6; break; // Reunion
    case 14: ev.type_id = 2; break; // Bridal Shower
    case 15: case 16: case 17: case 18: ev.type_id = 3; break; // Proposal
    case 19: ev.type_id = 4; break; // Birthday
    default: ev.type_id = 1; // Wedding
  }
  ev.type = types[ev.type_id];
  
  ev.name = names1[index] + (ev.type_id < 4 ? ' & '+ names2[index] : '');
  
  ev.date = toISO(date);
  date = new Date(date.valueOf() - Math.max(1, Math.ceil(index/6)) * 86400000);
  
  if ((index+1) % 20 === 0) ev.email = ev.password = true;
  
  
  var catsNum = (catTitles.length + index - 2) % (catTitles.length-1) + 2;
  for (var i = 0; i < catsNum; i++) {
    addCategory(ev, i);
  }
  
  return ev;
}

var landscapePhotoRE = new RegExp("^(1|3|4|5|13|16|17|18|19|20|42)$");
function addCategory(ev, index) {
  var img, landscape;
  var photosStart = (71 + ev.id * catID * 2) % 71 + 1;
  var photosNum = (40 + ev.id * index * 30) % 250;
  if (photosNum < 20) photosNum = 20;
  var photosEnd = photosStart + photosNum;
  
  photos['cat'+ catID] = [];
  ev.categories[index] = {
    id: catID++,
    name: catTitles[index],
    img_url: "/img/pics/img" + (1 + (catID % 19)) +'.jpg?' + (uid++),
    img_count: photosEnd - photosStart + 1,
    music: ev.id % 3 === 2 ? tracks[ev.id % tracks.length] :
           ev.id % 3 !== 0 ? tracks[catID % tracks.length] : '',
  };
  
  for (var i = photosStart; i <= photosEnd; i++) {
    img = (70 + i) % 71 + 1;
    landscape = landscapePhotoRE.test(img);
    photos['cat'+ ev.categories[index].id].push({
      id: photoID++,
      name: img % 3 === 0 ? photoTitles[(img / 3) % photoTitles.length] : '',
      download: img % 4 === 0,
      orig_width: landscape ? 1280 : 941,
      orig_height: landscape ? 941 : 1280,
      img_url: '/img/pics/events/1/img'+ img +'.jpg?'+ (uid++),
      bw: true,
      sepia: true,
      
      // product: false,
      // featured: /^(1|3|5|16|20|42|47|48|54|62|50|70)$/.test(n),
      // displayed: true,
      // fav_id: 0,
      // can_hide: false,
      // hidden: false,
    });
  }
}

for (var i = 0; i < 68; i++) {
  events.push(createEvent(i));
}

module.exports = {
  events: events,
  photos: photos,
};
