// app/views/global/misc-popup/contact-popup.js
/*==========================================================================*\
  A popup with a contact form
\*==========================================================================*/
"use strict";
var app = require('app');
var ua = require('app/utils/ua');
var xhr = require('xhr');
var Parent = require('./misc-popup');

module.exports = Parent.extend({
  template: require('app/templates').global['misc-popup'].contact,
  
  // initialize: function () {
  //   Parent.prototype.initialize.apply(this, arguments);
  // },
  
  // props: {
  // },
  
  // events: {
  //   click: 'onClick',
  // },
  
  // bindings: {
  // },
  
  save: function () {
    if (this.xhr) return;
    
    var name = this.query('input.name').value;
    var email = this.query('input.email').value;
    var message = this.query('textarea.message').value;
    if (!name) { this.query('input.name').focus(); return; }
    if (!email) { this.query('input.email').focus(); return; }
    if (!message) { this.query('textarea.message').focus(); return; }
    
    var self = this;
    this.query('.button.save').classList.add('loading');
    
    var data = {
      sender_email: email,
      sender_name: name,
      message: message,
    };
    if (app.currentPage.eventId) {
      data.event_id = app.currentPage.eventId;
    }
    
    // this.xhr = true;
    // setTimeout(function () {
    //   self.sent = true;
    // }, 500);
    // return;
    
    this.xhr = xhr({
      uri: 'http://www.instaproofs.com/api/rest2/email/contact?url='+ app.url,
      method: 'POST',
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }, function (err, resp, body) {
      body = JSON.parse(body);
      if (body.status === 200) {
        self.sent = true;
      }
      else {
        alert('Failed to send the message. Please check your internet '+
              'connection and try again.');
      }
      self.xhr = null;
    });
  },
});
