// app/views/global/misc-popup/misc-popup.js
/*==========================================================================*\
  A popup with FAQ, Contact form etc.
\*==========================================================================*/
"use strict";
var app = require('app');
var ua = require('app/utils/ua');
var Parent = require('app/views/base-view');

module.exports = Parent.extend({
  autoRender: true,
  template: require('app/templates').event['product-popup'],
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
  },
  
  props: {
    animatedHeight: ['boolean', true, false],
    sent: ['boolean', true, false],
    data: 'object',
  },
  
  events: {
    click: 'onClick',
  },
  
  bindings: {
    animatedHeight: {
      type: 'booleanClass',
      name: 'animated-height',
    },
    sent: {
      type: 'booleanClass',
      name: 'sent',
    },
  },
  
  onClick: function (e) {
    if (!e.target.closest('.box') || e.target.closest('.close') ||
        e.target.closest('.cancel')) {
      this.hide();
    }
    if (e.target.closest('.button.save')) {
      this.save();
    }
  },
  
  hide: function () {
    this.el.classList.add('hidden');
    this.after(400, function () {
      this.stopListening();
      this.remove();
    });
  },
  
  render: function () {
    this.renderWithTemplate(this);
    // this.cacheElements({
    //   options: '.options',
    // });
    
    this.after(50, function () {
      this.updateHeight();
      var nameInput = this.query('input.name');
      if (nameInput) nameInput.focus();
    });
    
  },
  
  updateHeight: function () {
    var contentHeight = this.query('.misc-content').offsetHeight;
    this.query('.box').style.maxHeight = contentHeight + 74 + 'px';
  },
});
