// app/pages/event-page.js
/*==========================================================================*\
  Event page with a list of categories 
\*==========================================================================*/
"use strict";

var _ = require('lodash');
var CategoriesView = require('app/views/event/categories');
var ProductsView = require('app/views/event/products');
var Page = require('app/pages/base');
var language = require('app/language');

module.exports = Page.extend({
  pageTitle: '',
  pageType: 'event',
  
  template: require('app/templates').pages.event,
  
  events: {
    // 'click .sorting a': 'sort',
    'click .load-more > .label': 'loadMore',
  },
  
  subviews: _.merge({
    categories: {
      container: '.categories',
      waitFor: 'model.name',
      prepareView: function (el) {
        this.triggerReady();
        this.categories = new CategoriesView({
          el: el,
          parent: this,
          collection: this.model.categories,
          language: language,
        });
        return this.categories;
      },
    },
    products: {
      container: '.products',
      waitFor: 'model.name',
      prepareView: function (el) {
        this.triggerReady();
        this.products = new ProductsView({
          el: el,
          parent: this,
          collection: this.model.categories, // TODO: this should be product model, hence we are fetching categories in event model so we have to fetch it from product model
          language: language,
        });
        return this.products;
      },
    },
  }, Page.prototype.subviews),
  
  bindings: {
    'model.name': '[data-hook=event-name]',
  },
  
  initialize: function (ops) {
    Page.prototype.initialize.apply(this, arguments);
    ops = ops || {};
    
    if (!this.model.name || !this.model.complete) {
      this.model.fetch();
    } 
    else {
      this.triggerReady();
    }
    
    this.setTitle(this.model.name);
    this.listenTo(this.model, 'change:name', function () {
      this.setTitle(this.model.name);
    });
    
    this.render();
  },
  
});
