// app/views/global/footer.js
/*==========================================================================*\
  Global footer
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');
var i18n = require('app/i18n/i18n');
var popup = require('./popup');
var LanguagesView = require('./languages');

module.exports = Parent.extend({
  props: {
    i18n: 'object',
  },
  
  template: require('app/templates').global.footer,
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.model = app.photographer;
    this.i18n = i18n;
    this.listenTo(i18n, 'select', this.updateLanguage);
  },
  
  events: {
    'click .language': 'showLanguages',
  },
  
  bindings: {
    'i18n.id': {
      hook: 'language',
      type: function () {
        this.updateLanguage();
      },
    },
    
    'model.name': '[data-hook=name]',
    'model.combinedAddress': {
      hook: 'address',
      type: function (el, val) {
        el.href = "http://maps.google.com?q=" + val;
        el.innerHTML = val;
      },
    },
    'model.phone': {
      hook: 'phone',
      type: function (el, val) {
        el.href = "tel:" + val;
        el.innerHTML = val;
      },
    },
    'model.email': {
      hook: 'email',
      type: function (el, val) {
        el.href = "mailto:" + val;
        el.innerHTML = val;
      },
    },
    'model.website_url': {
      hook: 'web',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val.replace(/^https?:\/\//i, '');
      },
    },
    
    'model.facebook_url': {
      hook: 'facebook',
      type: function (el, val) {
        el.href = val;
        // el.innerHTML = val;
      },
    },
    'model.twitter_url': {
      hook: 'twitter',
      type: function (el, val) {
        el.href = val;
        // el.innerHTML = val;
      },
    },
    'model.linkedin_url': {
      hook: 'linkedin',
      type: function (el, val) {
        el.href = val;
        // el.innerHTML = val;
      },
    },
    'model.instagram_url': {
      hook: 'instagram',
      type: function (el, val) {
        el.href = val;
        // el.innerHTML = val;
      },
    },
  },
  
  updateLanguage: function () {
    this.queryByHook('language').textContent = this.i18n.locale.name;
    this.query('.language').className = 'language '+ this.i18n.id;
  },
  
  showLanguages: function (e) {
    if (e && e.preventDefault) e.preventDefault();
    
    if (!this.languagesView) {
      this.languagesView = new LanguagesView({
        parent: this,
      });
    }
    
    if (popup.displayedId && popup.displayedId === this.popupId) return;
    this.popupId = popup.show({
      targetEl: this.query('.language'),
      content: this.languagesView.el,
      minWidth: 1,
    });
  },
});
