// app/views/global/footer.js
/*==========================================================================*\
  Global footer
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');

module.exports = Parent.extend({
  
  template: require('app/templates').global.footer,
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    this.model = app.photographer;
  },
  
  events: {
    
  },
  
  bindings: {
    'model.name': '[data-hook=name]',
    'model.combinedAddress': {
      hook: 'address',
      type: function (el, val) {
        el.href = "http://maps.google.com?q=" + val;
        el.innerHTML = val;
      },
    },
    'model.phone': {
      hook: 'phone',
      type: function (el, val) {
        el.href = "tel:" + val;
        el.innerHTML = val;
      },
    },
    'model.email': {
      hook: 'email',
      type: function (el, val) {
        el.href = "mailto:" + val;
        el.innerHTML = val;
      },
    },
    'model.website_url': {
      hook: 'web',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val.replace(/^https?:\/\//i, '');
      },
    },
    
    'model.facebook_url': {
      hook: 'facebook',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val;
      },
    },
    'model.twitter_url': {
      hook: 'twitter',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val;
      },
    },
    'model.linkedin_url': {
      hook: 'linkedin',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val;
      },
    },
    'model.instagram_url': {
      hook: 'instagram',
      type: function (el, val) {
        el.href = val;
        el.innerHTML = val;
      },
    },
  },
});
