// app/views/global/lightbox-notices.js
/*==========================================================================*\
  Dropdown popup with side-wide notices 
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');
var NoticeView = require('./notice');

module.exports = Parent.extend({
  autoRender: true,
  template: require('app/templates').global['lightbox-notices'],
  
  events: {
    'click .button.hide': 'hide',
  },
  
  bindings: {
    // 'model.darkMode': {
    //   type: function (el, val) {
    //     el.textContent = val ? 'Switch to light color scheme'
    //                          : 'Switch to dark color scheme';
    //   },
    //   selector: '.color-scheme',
    // },
  },
  
  // initialize: function () {
  //   Parent.prototype.initialize.apply(this, arguments);
  // },
  
  render: function () {
    this.renderWithTemplate(this);
    this.renderCollection(app.notices, NoticeView, '.items');
  },
  
  hide: function () {
    app.view.hideNotices();
  },
  
});
