// app/views/global/menu.js
/*==========================================================================*\
  Dropdown menu for the header navbar
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');

module.exports = Parent.extend({
  dictionary: 'nav',
  // autoRender: true,
  template: require('app/templates').global.menu,
  
  events: {
    'click .color-scheme': 'toggleScheme',
    'click .transitions': 'toggleTransitions',
  },
  
  bindings: {
    'model.darkMode': {
      type: function (el, val) {
        el.textContent = val ? 'Switch to light color scheme'
                             : 'Switch to dark color scheme';
      },
      selector: '.color-scheme',
    },
    'model.noTransitions': {
      type: function (el, val) {
        el.textContent = val ? 'Turn on page transitions'
                             : 'Turn off page transitions';
      },
      selector: '.transitions',
    },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    
    this.render();
  },
  
  toggleScheme: function (e) {
    this.model.darkMode = !this.model.darkMode;
    app.storage.darkMode = this.model.darkMode ? '1' : '';
    this.parent.toggleMenu();
    e.preventDefault();
    return false;
  },
  toggleTransitions: function (e) {
    app.state.noTransitions = !app.state.noTransitions;
    app.storage.noTransitions = app.state.noTransitions ? '1' : '';
    this.parent.toggleMenu();
    e.preventDefault();
    window.location.reload();
    return false;
  },
});
