// app/views/global/menu.js
/*==========================================================================*\
  Dropdown menu for the header navbar
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');
var FaqPopup = require('./misc-popup/faq-popup');
var ContactPopup = require('./misc-popup/contact-popup');
var OrderStatusPopup = require('./misc-popup/order-status-popup');

window.toggleScheme = function () {
  app.state.darkMode = !app.state.darkMode;
  app.storage.darkMode = app.state.darkMode ? '1' : '';
  return 'Switched to '+ (app.state.darkMode ? 'dark' : 'light') + 
         ' color scheme';
};
window.toggleTransitions = function () {
  app.state.noTransitions = !app.state.noTransitions;
  app.storage.noTransitions = app.state.noTransitions ? '1' : '';
  window.location.reload();
  return 'Page transitions are '+ (app.state.noTransitions ? 'off' : 'on');
};

module.exports = Parent.extend({
  dictionary: 'nav',
  // autoRender: true,
  template: require('app/templates').global.menu,
  
  events: {
    // 'click .color-scheme': 'toggleScheme',
    // 'click .transitions': 'toggleTransitions',
    'click .faq': 'showFaq',
    'click .contact': 'showContacts',
    'click .order-status': 'showOrderStatus',
  },
  
  bindings: {
    // 'model.darkMode': {
    //   type: function (el, val) {
    //     el.textContent = val ? 'Switch to light color scheme'
    //                          : 'Switch to dark color scheme';
    //   },
    //   selector: '.color-scheme',
    // },
    // 'model.noTransitions': {
    //   type: function (el, val) {
    //     el.textContent = val ? 'Turn on page transitions'
    //                          : 'Turn off page transitions';
    //   },
    //   selector: '.transitions',
    // },
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    
    if (app.state.bootstrapped) {
      this.render();
    }
    else {
      var self = this;
      this.renderWithTemplate('');
      app.state.on('change:bootstrapped', function () {
        self.render();
      });
    }
  },
  
  showPopup: function (popup, e) {
    if (e && e.preventDefault) e.preventDefault();
    document.body.insertBefore(popup.el, document.getElementById('popup'));
    this.parent.menuDisplayed = false;
    return false;
  },
  showFaq: function (e) { return this.showPopup(new FaqPopup(), e); },
  showContacts: function (e) { return this.showPopup(new ContactPopup(), e); },
  showOrderStatus: function (e) { 
    return this.showPopup(new OrderStatusPopup(), e);
  },
  
  
});
