// app/views/global/nav.js
/*==========================================================================*\
  Fixed header navbar
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');
var ViewSwitcher = require('app/utils/view-switcher');

module.exports = Parent.extend({
  dictionary: 'nav',
  // autoRender: true,
  template: require('app/templates').global.nav,
  
  events: {
    'click a': 'onLinkClick'
  },
  
  initialize: function () {
    Parent.prototype.initialize.apply(this, arguments);
    
    // this.renderWithTemplate();
    this.el.innerHTML = this.template();
    this.viewSwitcher = new ViewSwitcher(this.query('.contents'), 
        function (done, newView, oldView) {
      setTimeout(function () {
        done();
      }, 300);
      newView.el.classList.remove('hidden');
      if (oldView) oldView.el.classList.add('hidden');
    });
    
    this.after(100, function () {
      this.query('.logo').addEventListener('click', function (e) {
        app.state.darkMode = !app.state.darkMode;
        app.storage.darkMode = app.state.darkMode ? '1' : '';
        e.preventDefault();
        return false;
      }, false);
      // this.query('.cart').addEventListener('click', function (e) {
      //   app.state.noTransitions = !app.state.noTransitions;
      //   app.storage.noTransitions = app.state.noTransitions ? '1' : '';
      //   e.preventDefault();
      //   window.location.reload();
      //   return false;
      // }, false);
    });
  },
  
  setContents: function (newView) { // set nav contents view
    this.viewSwitcher.set(newView);
  },
  
  onLinkClick: function (e) {
    e.stopPropagation();
  },
});
