// app/views/category/category.js
/*==========================================================================*\
  Event's Category
\*==========================================================================*/
"use strict";

var app = require('app');
var Parent = require('app/views/base-view');
var PhotosView = require('./thumbs');

module.exports = Parent.extend({
  
  template: require('app/templates').category.category,
  
  preview: false, // whether this is a preview view on 'All Categories' page
  
  initialize: function (ops) {
    Parent.prototype.initialize.apply(this, arguments);
    ops = ops || {};
    this.preview = !!ops.preview;
  },
  
  events: {
    'click .category-slideshow': 'onSlideshowClick',
  },
  
  bindings: {
    'model.name': {
      type: function (el, name) {
        if (!this.model.isFavModel) el.textContent = name;
      },
      hook: 'category-name',
    },
    
    'model.mtime': {
      type: function () {
        this.updateSledeshowLink();
      },
    },
    
    'model.photos.length': {
      type: 'booleanClass',
      selector: '.event-category',
      yes: 'loaded',
      no: 'loading',
    },
    'model.photos': {
      type: function (el, list) {
        if (list.length) {
          el.setAttribute('href', list.parent.app_url +'/'+ list.at(0).id);
        }
      },
      selector: '.category-slideshow',
    },
  },
  
  render: function () {
    this.scheduleWrite(function () {
      this.renderWithTemplate(this);
      this.photos = new PhotosView({
        el: this.query('.photos'),
        collection: this.model.photos,
        parent: this,
        preview: this.preview,
      });
      this.cacheElements({
        slideshowEl: '.category-slideshow',
      });
      this.updateSledeshowLink();
    });
  },
  
  updateSledeshowLink: function () {
    if (this.slideshowEl && this.model.isFavModel && this.model.photos.length) {
      this.query('.category-slideshow').setAttribute('href',
          this.model.app_url +'/'+ this.model.photos.at(0).id);
    }
  },
  
  onPageShow: function () {
    if (this.photos && this.photos.collection && 
        this.photos.collection.length) {
      this.showMore();
      this.photos.loadPhotos();
    }
  },
  
  onSlideshowClick: function () {
    app.router.enterSlideshow = true;
  },
  
  showMore: function () {
    this.photos.showMore();
    return this;
  },
  
});
