// app/router.js
/*==========================================================================*\
  App routes
\*==========================================================================*/
"use strict";

var app = require('app');
// var CartModel = require('app/models/cart/cart-model');
var EventModel = require('app/models/event-model');
var Lightbox = require('app/views/lightbox/lightbox');
var NavContents = require('app/views/global/nav-contents');
var EventNav = require('app/views/event/event-nav');
var pages = {}; // list of initialized page views
var views = {
  home: require('app/views/events/events-page'),
  event: require('app/views/event/event-page'),
  category: require('app/views/category/category-page'),
  favorites: require('app/views/category/favorites-page'),
  cart: require('app/views/cart/cart-page'),
};
var lightbox; // lightbox view
var navs = { // navbar views
  global: new NavContents(),
}; 

function parseQuery(query) {
  if (!query) return {};
  var vars = query.split('&');
  var data = {};
  var pair;
  for (var i = 0; i < vars.length; i++) {
    pair = vars[i].split('=');
    data[pair[0]] = pair[1];
  }
  return data;
}

module.exports = require('ampersand-router').extend({
  
  currentPage: null, // current type of page opened 
  enterSlideshow: false, // enter the slideshow on next lightbox display 
  
  setNav: function (eid) { // sets navbar to global or event's one
    var view;
    if (!eid) view = navs.global;
    else if (navs['event'+eid]) view = navs['event'+eid];
    else {
      view = new EventNav({
        id: eid,
        model: app.eventModels[eid],
      });
      navs['event'+eid] = view;
    }
    
    app.view.nav.setContents(view);
  },
  
  openPage: function (page, ops) {
    ops = ops || {};
    var id = ops.id ? page + ops.id : page;
    delete ops.id;
    var view = pages[id] ? pages[id].reset(ops) : new views[page](ops);
    pages[id] = view;
    
    this.currentPage = page;
    
    this.trigger('page', view, id);
    this.trigger('page:'+id, view, id);
  },
  showLightbox: function (eid, cid, pid, back, buy, crop) {
    var event = app.eventModels[eid]
        || (pages.home && pages.home.list.collection.get(eid));
    console.log(cid);
    var collection = cid === 'favorites' ? event.favorites.photos
                                         : event.categories.get(cid).photos;
    var photo = collection.get(pid);
        
    if (!lightbox) {
      this.lightbox = lightbox = new Lightbox({ 
        // el: app.view.query('#lightbox') 
      });
    }
    
    lightbox.reset({
      event: event,
      collection: collection,
      photo: photo,
      backTo: back || (cid === 'favorites' ? '#events/'+eid+'/favorites'
                                           : '#events/'+eid+'/'+cid),
      slideshow: this.enterSlideshow,
      buyForm: buy,
      crop: crop,
    });
    lightbox.show();
    this.enterSlideshow = false;
  },
  
  login: function (event) {
    var splash = app.view.splashView;
    var keychain = app.keychain.events.get(event.id);
    if (event.name && event.locked) {
      if (!keychain || !keychain.verified || (event.email && !keychain.email) || 
          (event.password && !keychain.password) ) {
        splash.showLoginForm(event);
      }
    }
  },
  
  initialize: function() {
    this.lightbox = {};
    this.productLightbox = {};
    // homepage ==============================================================
    this.route(/^events(?:\?(.*?))?$/, function (query) {
      if (lightbox) lightbox.hide();
      query = parseQuery(query);
      
      this.openPage('home', {
        sort: query.sort || 'date',
        filter: query.show || 'all',
      });
      this.setNav();
    });
    
    // event page ============================================================
    this.route(/^events\/(\d+)$/, function (eid) {
      if (lightbox) lightbox.hide();
      eid = +eid;
      var model = app.eventModels[eid]
          || (pages.home && pages.home.list.collection.get(eid))
          || new EventModel({ id: eid });
      app.eventModels[eid] = model;
      this.login(model);
      this.openPage('event', {
        id: eid,
        model: model,
      });
      this.setNav(eid);
    });
    
    // event category page ===================================================
    this.route(/^events\/(\d+)\/(\d+)$/, function (eid, cid) {
      var lightboxDisplayed = !!this.lightbox.displayed;
      if (lightbox) lightbox.hide();
      
      eid = +eid;
      cid = +cid;
      var model = app.eventModels[eid]
          || (pages.home && pages.home.list.collection.get(eid))
          || new EventModel({ id: eid });
      app.eventModels[eid] = model;
      
      this.login(model);
      
      this.openPage('category', {
        id: eid+'_'+cid,
        cid: cid,
        model: model,
        showMore: lightboxDisplayed,
      });
      this.setNav(eid);
    });
    
    // event favorites page ====================================================
    this.route(/^events\/(\d+)\/favorites$/, function (eid) {
      var lightboxDisplayed = !!this.lightbox.displayed;
      if (lightbox) lightbox.hide();
      
      eid = +eid;
      var model = app.eventModels[eid]
          || (pages.home && pages.home.list.collection.get(eid))
          || new EventModel({ id: eid });
      app.eventModels[eid] = model;
      
      this.login(model);
      
      this.openPage('favorites', {
        id: eid,
        model: model,
        showMore: lightboxDisplayed,
      });
      this.setNav(eid);
    });

    // single photo ==========================================================
    this.route(/^events\/(\d+)\/(\d+|favorites)\/(\d+)(?:(buy)|(crop)(\d+)(?::(\d+))?)?$/,
        function (eid, cid, pid, buy, crop, id1, id2) {
      // if (cid === 'favorites') return app.navigate('events/'+eid+'/favorites');
      this.lightboxMode = true;
      this.willShowLightbox = true;
      var self = this;
      eid = +eid;
      cid = cid === 'favorites' ? 'favorites' : +cid;
      pid = +pid;
      var event = app.eventModels[eid]
          || (pages.home && pages.home.list.collection.get(eid))
          || new EventModel({ id: eid });
      app.eventModels[eid] = event;
      
      this.login(event);
      
      if (crop) { // sizeId or productId:sizeId
        crop = id2 ? [+id1, +id2] : +id1;
      }
      var mode = crop || buy;
      
      if (buy || crop) { // these URLs are only for internal use 
        this.navigate('events/'+eid+'/'+cid+'/'+pid, {
          trigger: false,
          replace: true,
        });
      }
      
      // if (!mode || this.currentPage !== 'cart') this.openPage('category', {
      if (!mode) {
        if (cid === 'favorites') {
          this.openPage('favorites', {
            id: eid,
            model: event,
            showMore: false,
          });
        }
        else {
          this.openPage('category', {
            id: eid+'_'+cid,
            cid: cid,
            model: event,
            showMore: false,
          });
        }
      }
      this.setNav(eid);
      
      if (!app.isReady) {
        app.ready(function () {
          self.showLightbox(eid, cid, pid);
        });
      }
      else {
        this.showLightbox(eid, cid, pid, 
            mode && this.currentPage === 'cart' && ('#events/'+eid+'/cart'),
            !!buy, crop);
      }
      
    });
    
    // cart page =============================================================
    this.route(/^events\/(\d+)\/cart$/, function (eid) {
      if (lightbox) lightbox.hide();
      
      // var cart = carts['event'+eid];
      // if (!cart) {
      //   cart = carts['event'+eid] = new CartModel({})
      // }
      var event = app.eventModels[eid]
          || (pages.home && pages.home.list.collection.get(eid))
          || new EventModel({ id: eid });
      app.eventModels[eid] = event;
      
      if (!event.complete && !event.fetching) {
        event.fetch();
      }
      
      this.login(event);
      
      this.openPage('cart', {
        id: eid,
        model: app.eventModels[eid].cart,
      });
      this.setNav(eid);
    });
    
  },
  
  routes: {
    '(*path)': function () { this.redirectTo('events?show=all&sort=date'); },
  },
});
