// app/state.js
/*==========================================================================*\
  Global state of the app including stuff like dimensions
\*==========================================================================*/
"use strict";

module.exports = require('ampersand-state').extend({
  props: {
    
    // set to true when global data such as config and photographer's info
    // has been loaded 
    bootstrapped: ['boolean', true, false],
    
    width: 'number',
    height: 'number',
    contentWidth: 'number', // width of standard content blocks with width limit
    contentHeight: 'number', // window height minus nav height
    
    fullscreen: ['boolean', true, false], // the app is currently fullscreen
    
    hiddenNav: ['boolean', true, false], // whether fixed nav should be hidden
    
    scrolling: ['boolean', true, false], // whether user currently scrolling 
    
    pageTransition: 'string', // kind of next page transition (back, fwd etc.)
    
    // pointer events are disabled (e.g. during scroll)
    noPointerEvents: ['boolean', true, false], 
    
    // a mode with no page transitions and a single global scroll
    noTransitions: ['boolean', true, false],
    
    // Dark nav bars mode
    darkMode: ['boolean', true, false], 
    
    splashState: 'string', // 'locked', 'unlocked', or '' 
    
    // list of URLs for images that has been already loaded 
    loadedImages: {
      type: 'object',
      default: function () { return {}; }
    },
  },
  
  initialize: function () {
    window.addEventListener('resize', this.resize.bind(this));
    this.resize();
  },
  
  resize: function () {
    var i, keys = ['width', 'height', 'contentWidth', 'contentHeight'];
    var oldValues = {};
    for (i = 0; i < keys.length; i++) {
      oldValues[keys[i]] = this[keys[i]];
    }
    
    this.width = document.documentElement.clientWidth;
    this.height = document.documentElement.clientHeight;
    this.contentHeight = this.height - 
        (this.width <= 750 || this.height <= 800 ? 44 : 80);
    this.contentWidth = this.width >= 1304 ? 1200 
                                           : Math.round(this.width * 0.92);
    
    for (i = 0; i < keys.length; i++) {
      if (oldValues[keys[i]] !== this[keys[i]]) {
        this.trigger('resize');
        break;
      }
    }
    
    return this;
  },
});
