// app/templates.js
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        define([], factory);
    } else if (typeof exports === 'object') {
        module.exports = factory();
    } else if (typeof root === 'undefined' || root !== Object(root)) {
        throw new Error('templatizer: window does not exist or is not an object');
    } else {
        root.templatizer = factory();
    }
}(this, function () {
    var jade=function(){function r(r){return null!=r&&""!==r}function n(e){return Array.isArray(e)?e.map(n).filter(r).join(" "):e}var e={};return e.merge=function t(n,e){if(1===arguments.length){for(var a=n[0],s=1;s<n.length;s++)a=t(a,n[s]);return a}var i=n["class"],l=e["class"];(i||l)&&(i=i||[],l=l||[],Array.isArray(i)||(i=[i]),Array.isArray(l)||(l=[l]),n["class"]=i.concat(l).filter(r));for(var o in e)"class"!=o&&(n[o]=e[o]);return n},e.joinClasses=n,e.cls=function(r,t){for(var a=[],s=0;s<r.length;s++)a.push(t&&t[s]?e.escape(n([r[s]])):n(r[s]));var i=n(a);return i.length?' class="'+i+'"':""},e.attr=function(r,n,t,a){return"boolean"==typeof n||null==n?n?" "+(a?r:r+'="'+r+'"'):"":0==r.indexOf("data")&&"string"!=typeof n?" "+r+"='"+JSON.stringify(n).replace(/'/g,"&apos;")+"'":t?" "+r+'="'+e.escape(n)+'"':" "+r+'="'+n+'"'},e.attrs=function(r,t){var a=[],s=Object.keys(r);if(s.length)for(var i=0;i<s.length;++i){var l=s[i],o=r[l];"class"==l?(o=n(o))&&a.push(" "+l+'="'+o+'"'):a.push(e.attr(l,o,!1,t))}return a.join("")},e.escape=function(r){var n=String(r).replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/"/g,"&quot;");return n===""+r?r:n},e.rethrow=function a(r,n,e,t){if(!(r instanceof Error))throw r;if(!("undefined"==typeof window&&n||t))throw r.message+=" on line "+e,r;try{t=t||require("fs").readFileSync(n,"utf8")}catch(s){a(r,null,e)}var i=3,l=t.split("\n"),o=Math.max(e-i,0),c=Math.min(l.length,e+i),i=l.slice(o,c).map(function(r,n){var t=n+o+1;return(t==e?"  > ":"    ")+t+"| "+r}).join("\n");throw r.path=n,r.message=(n||"Jade")+":"+e+"\n"+i+"\n\n"+r.message,r},e}();

    var templatizer = {};
    templatizer["events"] = {};
    templatizer["lightbox"] = {};
    templatizer["pages"] = {};

    // events/category-footer.jade compiled template
    templatizer["events"]["category-footer"] = function tmpl_events_category_footer(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            var event = model && model.collection.parent;
            var prev = model && model.prev;
            var next = model && model.next;
            buf.push('<div class="fixed-footer event-category"><div class="content"><a data-hook="prev-category"' + jade.attr("href", prev ? "#events/" + event.id + "/" + prev.id : null, true, false) + jade.cls([ "prev", !prev ? "hidden" : null ], [ null, true ]) + "><span>" + jade.escape(null == (jade_interp = model && model.prev && model.prev.name) ? "" : jade_interp) + '</span></a><a data-hook="next-category"' + jade.attr("href", next ? "#events/" + event.id + "/" + next.id : null, true, false) + jade.cls([ "next", !next ? "hidden" : null ], [ null, true ]) + "><span>" + jade.escape(null == (jade_interp = model && model.next && model.next.name) ? "" : jade_interp) + "</span></a></div></div>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/category-item.jade compiled template
    templatizer["events"]["category-item"] = function tmpl_events_category_item(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, Math) {
            var event = model.collection.parent;
            var imageURL = "/img/pics/thumbs/img" + (1 + model.id % 20) + ".jpg";
            var nl = model.name.length;
            var titleClass = nl >= 25 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push("<a" + jade.attr("href", "#events/" + event.id + "/" + model.id, true, false) + ' class="event-item"><div' + jade.attr("style", "background-image:url(" + imageURL + ");", true, false) + ' class="bg"></div><div class="spacer"></div><div class="info"><h3' + jade.cls([ titleClass ], [ true ]) + ">" + jade.escape(null == (jade_interp = model.locked ? "Private Event" : model.name) ? "" : jade_interp) + '</h3><div class="date">' + jade.escape(null == (jade_interp = model.img_count + " Images") ? "" : jade_interp) + "</div></div></a>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined);
        return buf.join("");
    };

    // events/category.jade compiled template
    templatizer["events"]["category"] = function tmpl_events_category(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            buf.push('<div class="event-category"><div class="heading"><div class="line"></div><div data-hook="category-name" class="label">' + jade.escape(null == (jade_interp = model && model.name) ? "" : jade_interp) + '</div><div class="line"></div></div><a class="category-slideshow">Slideshow</a><div class="photos"></div><div class="load-more"><div class="label">Show More Images</div><div class="loader"></div></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/group.jade compiled template
    templatizer["events"]["group"] = function tmpl_events_group(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(classes, label) {
            buf.push("<div" + jade.cls([ "events-group", classes ], [ null, true ]) + '><div class="heading group-header"><div class="line"></div><div class="label">' + (null == (jade_interp = label) ? "" : jade_interp) + '</div><div class="line"></div></div><div class="group-events"></div></div>');
        }).call(this, "classes" in locals_for_with ? locals_for_with.classes : typeof classes !== "undefined" ? classes : undefined, "label" in locals_for_with ? locals_for_with.label : typeof label !== "undefined" ? label : undefined);
        return buf.join("");
    };

    // events/list-item.jade compiled template
    templatizer["events"]["list-item"] = function tmpl_events_list_item(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, Math) {
            var classes = [];
            if (model.showcase) classes.push("showcased");
            if (model.locked) classes.push("locked");
            var imageURL = "/img/pics/thumbs/" + (model.locked ? "blur/" : "") + model.img;
            var nl = model.name.length;
            var titleClass = nl >= 25 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push("<a" + jade.attr("href", "#events/" + model.id, true, false) + jade.cls([ "event-item", classes ], [ null, true ]) + "><div" + jade.attr("style", "background-image:url(" + imageURL + ");", true, false) + ' class="bg"></div><div class="spacer"></div><div class="info"><h3' + jade.cls([ titleClass ], [ true ]) + ">" + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h3><div class="date">' + jade.escape(null == (jade_interp = model.date.format("MMMM Do, YYYY")) ? "" : jade_interp) + '</div><div class="type">' + jade.escape(null == (jade_interp = model.type) ? "" : jade_interp) + "</div></div></a>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined);
        return buf.join("");
    };

    // events/photo.jade compiled template
    templatizer["events"]["photo"] = function tmpl_events_photo(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            buf.push('<div class="grid-photo"><div class="image"></div><div class="fill borders"></div><div class="overlay"><a' + jade.attr("href", model.app_url, true, false) + ' class="view-large"><span>View Larger Image</span></a><div class="icons"><div class="icon cart"><div class="tip">Buy</div></div><div class="icon fav"><div class="tip">To Favorites</div></div>');
            if (model.download) {
                buf.push("<a" + jade.attr("href", model.url, true, false) + ' target="_blank" bypass="bypass" class="icon download"><div class="tip">Download</div></a>');
            }
            buf.push('<div class="icon share"><div class="tip">Share</div></div></div></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // footer.jade compiled template
    templatizer["footer"] = function tmpl_footer(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(Date, model) {
            buf.push('<div class="page-footer"><div class="content"><div class="copyright"><div class="line"></div><div class="message"><i>All images copyright </i><span>' + jade.escape(null == (jade_interp = "© " + new Date().getFullYear() + " ") ? "" : jade_interp) + '</span><span data-hook="name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</span></div><div class="line"></div></div><div class="logo-container"><div class="line"></div><a href="#" class="logo"><img src="/img/logo.png"/></a><div class="line"></div></div><h1 data-hook="name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h1><div class="contact-links"><a' + jade.attr("href", "http://maps.google.com?q=" + model.address, true, false) + ' data-hook="address" target="_blank" class="address">' + jade.escape(null == (jade_interp = model.address) ? "" : jade_interp) + "</a><a" + jade.attr("href", "tel:" + model.phone, true, false) + ' data-hook="phone" class="phone">' + jade.escape(null == (jade_interp = model.phone) ? "" : jade_interp) + "</a><a" + jade.attr("href", "mailto:" + model.email, true, false) + ' data-hook="email" class="email">' + jade.escape(null == (jade_interp = model.email) ? "" : jade_interp) + "</a><a" + jade.attr("href", model.website_url, true, false) + ' data-hook="web" target="_blank" class="web">' + jade.escape(null == (jade_interp = model.website_url) ? "" : jade_interp) + '</a></div><div class="social"><a href="http://www.facebook.com" data-hook="facebook" target="_blank" class="facebook"></a><a href="http://www.twitter.com" data-hook="twitter" target="_blank" class="twitter"></a><a href="http://www.linkedin.com" data-hook="linkedin" target="_blank" class="linkedin"></a><a href="http://www.instagram.com" data-hook="instagram" target="_blank" class="instagram"></a></div></div></div>');
        }).call(this, "Date" in locals_for_with ? locals_for_with.Date : typeof Date !== "undefined" ? Date : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // lightbox/categories.jade compiled template
    templatizer["lightbox"]["categories"] = function tmpl_lightbox_categories(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(categories, active) {
            (function() {
                var $obj = categories;
                if ("number" == typeof $obj.length) {
                    for (var $index = 0, $l = $obj.length; $index < $l; $index++) {
                        var cat = $obj[$index];
                        buf.push("<a" + jade.attr("href", cat.app_url + "/" + cat.prod_list.at(0).id, true, false) + jade.cls([ cat === active ? "active" : "" ], [ true ]) + ">" + jade.escape(null == (jade_interp = cat.name) ? "" : jade_interp) + "</a>");
                    }
                } else {
                    var $l = 0;
                    for (var $index in $obj) {
                        $l++;
                        var cat = $obj[$index];
                        buf.push("<a" + jade.attr("href", cat.app_url + "/" + cat.prod_list.at(0).id, true, false) + jade.cls([ cat === active ? "active" : "" ], [ true ]) + ">" + jade.escape(null == (jade_interp = cat.name) ? "" : jade_interp) + "</a>");
                    }
                }
            }).call(this);
        }).call(this, "categories" in locals_for_with ? locals_for_with.categories : typeof categories !== "undefined" ? categories : undefined, "active" in locals_for_with ? locals_for_with.active : typeof active !== "undefined" ? active : undefined);
        return buf.join("");
    };

    // lightbox/photo.jade compiled template
    templatizer["lightbox"]["photo"] = function tmpl_lightbox_photo(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(require, model, Math) {
            var app = require("app");
            var nl = model.name.length;
            var titleClass = nl >= 35 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push('<div class="lightbox-photo"><div class="images"><div class="img"><div class="loader white"></div>');
            if (app.loadedImages[model.thumb_url]) {
                buf.push("<div" + jade.attr("style", "background-image:url(" + model.thumb_url + ")", true, false) + ' class="thumb"></div>');
            }
            buf.push("<img" + jade.attr("src", model.url, true, false) + '/><div class="borders"></div></div>');
            if (model.name) {
                buf.push("<div" + jade.cls([ "title", titleClass ], [ null, true ]) + '><div class="text">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + "</div></div>");
            }
            buf.push('</div><div class="actions"><div class="icon cart"><div class="tip">Buy Image</div></div><div class="icon fav"><div class="tip">Add To Favorites</div></div>');
            if (model.download) {
                buf.push("<a" + jade.attr("href", model.url, true, false) + ' target="_blank" bypass="bypass" class="icon download"><div class="tip">Download</div></a>');
            }
            buf.push('<div class="icon share"><div class="tip">Share</div></div></div></div>');
        }).call(this, "require" in locals_for_with ? locals_for_with.require : typeof require !== "undefined" ? require : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined);
        return buf.join("");
    };

    // nav.jade compiled template
    templatizer["nav"] = function tmpl_nav() {
        return '<nav><div class="content"><div class="body"><div class="right"><a href="#" class="search"><span>Search Events</span></a><a href="#favorites" class="favs"><span>Favorites</span></a><a href="#cart" class="cart"><span>Cart</span></a><a href="#" class="menu"><span>Menu</span></a></div><a href="#" class="logo"><img src="/img/logo.png"/></a></div></div></nav>';
    };

    // pages/event.jade compiled template
    templatizer["pages"]["event"] = function tmpl_pages_event(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(pageType, model) {
            buf.push("<div" + jade.cls([ "page", "event", "inactive", pageType === "event-category" ? "category" : "" ], [ null, null, null, true ]) + '><div class="page-scroller"><div class="page-scroll"><div class="page-content"><div class="content"> ');
            if (model) {
                buf.push('<div class="header"><h1 data-hook="event-name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h1><div class="page-path"> <span class="wide-only">You are here:</span><a href="#">Home</a><span>&gt;</span>');
                if (pageType === "event") {
                    buf.push('<span data-hook="event-name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + "</span>");
                } else {
                    buf.push('<a data-hook="event-name"' + jade.attr("href", "#events/" + model.id, true, false) + ">" + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</a><span>&gt;</span><span data-hook="category-name"></span>');
                }
                buf.push('</div></div><div class="events-group"><div class="categories group-events"></div></div>');
            }
            buf.push('</div></div><div class="page-footer"> </div></div></div></div>');
        }).call(this, "pageType" in locals_for_with ? locals_for_with.pageType : typeof pageType !== "undefined" ? pageType : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // pages/events.jade compiled template
    templatizer["pages"]["events"] = function tmpl_pages_events() {
        return '<div class="page events complete reloading inactive"><div class="page-scroller"><div class="page-scroll"><div class="page-content"><div class="content"> <div class="header"><div class="filters"><div class="inner"><a href="#" data-filter="showcased" class="showcased">Showcased<i> Events</i></a><a href="#" data-filter="current" class="current">Current</a><a href="#" data-filter="upcoming" class="upcoming">Upcoming</a><a href="#" data-filter="all" class="all active">All<i class="all"> Events</i></a></div></div><div class="sorting"><div class="inner"><span>Sort events by:</span><a href="#" data-sort="name" class="name"><i>Event </i>Name</a><a href="#" data-sort="date" class="date active"><i>Event </i>Date</a><a href="#" data-sort="type" class="type"><i>Event </i>Type</a></div></div><div class="clear"></div><div class="label"><div class="icon"></div><div class="loader grey-bg"></div></div></div><div class="events-list"></div><div class="load-more"><div class="label">Load More Events</div><div class="loader"></div></div></div></div><div class="page-footer"></div></div></div></div>';
    };

    return templatizer;
}));
