// app/templates.js
(function (root, factory) {
    if (typeof define === 'function' && define.amd) {
        define([], factory);
    } else if (typeof exports === 'object') {
        module.exports = factory();
    } else if (typeof root === 'undefined' || root !== Object(root)) {
        throw new Error('templatizer: window does not exist or is not an object');
    } else {
        root.templatizer = factory();
    }
}(this, function () {
    var jade=function(){function e(e){return null!=e&&""!==e}function n(t){return(Array.isArray(t)?t.map(n):t&&"object"==typeof t?Object.keys(t).filter(function(e){return t[e]}):[t]).filter(e).join(" ")}var t={};return t.merge=function r(n,t){if(1===arguments.length){for(var a=n[0],i=1;i<n.length;i++)a=r(a,n[i]);return a}var o=n["class"],s=t["class"];(o||s)&&(o=o||[],s=s||[],Array.isArray(o)||(o=[o]),Array.isArray(s)||(s=[s]),n["class"]=o.concat(s).filter(e));for(var l in t)"class"!=l&&(n[l]=t[l]);return n},t.joinClasses=n,t.cls=function(e,r){for(var a=[],i=0;i<e.length;i++)a.push(r&&r[i]?t.escape(n([e[i]])):n(e[i]));var o=n(a);return o.length?' class="'+o+'"':""},t.style=function(e){return e&&"object"==typeof e?Object.keys(e).map(function(n){return n+":"+e[n]}).join(";"):e},t.attr=function(e,n,r,a){return"style"===e&&(n=t.style(n)),"boolean"==typeof n||null==n?n?" "+(a?e:e+'="'+e+'"'):"":0==e.indexOf("data")&&"string"!=typeof n?(-1!==JSON.stringify(n).indexOf("&")&&console.warn("Since Jade 2.0.0, ampersands (`&`) in data attributes will be escaped to `&amp;`"),n&&"function"==typeof n.toISOString&&console.warn("Jade will eliminate the double quotes around dates in ISO form after 2.0.0")," "+e+"='"+JSON.stringify(n).replace(/'/g,"&apos;")+"'"):r?(n&&"function"==typeof n.toISOString&&console.warn("Jade will stringify dates in ISO form after 2.0.0")," "+e+'="'+t.escape(n)+'"'):(n&&"function"==typeof n.toISOString&&console.warn("Jade will stringify dates in ISO form after 2.0.0")," "+e+'="'+n+'"')},t.attrs=function(e,r){var a=[],i=Object.keys(e);if(i.length)for(var o=0;o<i.length;++o){var s=i[o],l=e[s];"class"==s?(l=n(l))&&a.push(" "+s+'="'+l+'"'):a.push(t.attr(s,l,!1,r))}return a.join("")},t.escape=function(e){var n=String(e).replace(/&/g,"&amp;").replace(/</g,"&lt;").replace(/>/g,"&gt;").replace(/"/g,"&quot;");return n===""+e?e:n},t.rethrow=function a(e,n,t,r){if(!(e instanceof Error))throw e;if(!("undefined"==typeof window&&n||r))throw e.message+=" on line "+t,e;try{r=r||require("fs").readFileSync(n,"utf8")}catch(i){a(e,null,t)}var o=3,s=r.split("\n"),l=Math.max(t-o,0),f=Math.min(s.length,t+o),o=s.slice(l,f).map(function(e,n){var r=n+l+1;return(r==t?"  > ":"    ")+r+"| "+e}).join("\n");throw e.path=n,e.message=(n||"Jade")+":"+t+"\n"+o+"\n\n"+e.message,e},t}();

    var templatizer = {};
    templatizer["events"] = {};
    templatizer["lightbox"] = {};
    templatizer["pages"] = {};

    // events/category-footer.jade compiled template
    templatizer["events"]["category-footer"] = function tmpl_events_category_footer(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            var event = model && model.collection.parent;
            var prev = model && model.prev;
            var next = model && model.next;
            buf.push('<div class="fixed-footer event-category"><div class="content"><a data-hook="prev-category"' + jade.attr("href", prev ? "#events/" + event.id + "/" + prev.id : null, true, false) + jade.cls([ "prev", !prev ? "hidden" : null ], [ null, true ]) + "><span>" + jade.escape(null == (jade_interp = model && model.prev && model.prev.name) ? "" : jade_interp) + '</span></a><a data-hook="next-category"' + jade.attr("href", next ? "#events/" + event.id + "/" + next.id : null, true, false) + jade.cls([ "next", !next ? "hidden" : null ], [ null, true ]) + "><span>" + jade.escape(null == (jade_interp = model && model.next && model.next.name) ? "" : jade_interp) + "</span></a></div></div>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/category-item.jade compiled template
    templatizer["events"]["category-item"] = function tmpl_events_category_item(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(Math, model) {
            var event = model.collection.parent;
            var imageURL = model.thumb_url;
            var nl = model.name.length;
            var titleClass = nl >= 25 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push("<a" + jade.attr("href", "#events/" + event.id + "/" + model.id, true, false) + ' class="event-item"><div' + jade.attr("style", "background-image:url(" + imageURL + ");", true, false) + ' class="bg"></div><div class="spacer"></div><div class="info"><h3' + jade.cls([ titleClass ], [ true ]) + ">" + jade.escape(null == (jade_interp = model.locked ? "Private Event" : model.name) ? "" : jade_interp) + '</h3><div class="date">' + jade.escape(null == (jade_interp = model.img_count + " Images") ? "" : jade_interp) + "</div></div></a>");
        }).call(this, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/category.jade compiled template
    templatizer["events"]["category"] = function tmpl_events_category(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            buf.push('<div class="event-category"><div class="heading"><div class="line"></div><div data-hook="category-name" class="label">' + jade.escape(null == (jade_interp = model && model.name) ? "" : jade_interp) + '</div><div class="line"></div></div><a class="category-slideshow">Slideshow</a><div class="photos"></div><div class="load-more"><div class="label">Show More Images</div><div class="loader"></div></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/group.jade compiled template
    templatizer["events"]["group"] = function tmpl_events_group(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(classes, label) {
            buf.push("<div" + jade.cls([ "events-group", classes ], [ null, true ]) + '><div class="heading group-header"><div class="line"></div><div class="label">' + (null == (jade_interp = label) ? "" : jade_interp) + '</div><div class="line"></div></div><div class="group-events"></div></div>');
        }).call(this, "classes" in locals_for_with ? locals_for_with.classes : typeof classes !== "undefined" ? classes : undefined, "label" in locals_for_with ? locals_for_with.label : typeof label !== "undefined" ? label : undefined);
        return buf.join("");
    };

    // events/list-item.jade compiled template
    templatizer["events"]["list-item"] = function tmpl_events_list_item(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(Math, model) {
            var classes = [];
            if (model.showcase) classes.push("showcased");
            if (model.locked) classes.push("locked");
            var imageURL = model.locked ? model.thumb_blur_url : model.thumb_url;
            var nl = model.name.length;
            var titleClass = nl >= 25 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push("<a" + jade.attr("href", "#events/" + model.id, true, false) + jade.cls([ "event-item", classes ], [ null, true ]) + "><div" + jade.attr("style", "background-image:url(" + imageURL + ");", true, false) + ' class="bg"></div><div class="spacer"></div><div class="info"><h3' + jade.cls([ titleClass ], [ true ]) + ">" + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h3><div class="date">' + jade.escape(null == (jade_interp = model.date.format("MMMM Do, YYYY")) ? "" : jade_interp) + '</div><div class="type">' + jade.escape(null == (jade_interp = model.type) ? "" : jade_interp) + "</div></div></a>");
        }).call(this, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // events/photo.jade compiled template
    templatizer["events"]["photo"] = function tmpl_events_photo(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            buf.push('<div class="grid-photo"><div class="image"></div><div class="fill borders"></div><div class="overlay"><a' + jade.attr("href", model.app_url, true, false) + ' class="view-large"><span>View Larger Image</span></a><div class="icons"><div class="icon cart"><div class="tip">Buy</div></div><div class="icon fav"><div class="tip">To Favorites</div></div>');
            if (model.download) {
                buf.push("<a" + jade.attr("href", model.url, true, false) + ' target="_blank" bypass="bypass" class="icon download"><div class="tip">Download</div></a>');
            }
            buf.push('<div class="icon share"><div class="tip">Share</div></div></div></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // footer.jade compiled template
    templatizer["footer"] = function tmpl_footer(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(Date, model) {
            buf.push('<div class="page-footer"><div class="content"><div class="copyright"><div class="line"></div><div class="message"><i>All images copyright </i><span>' + jade.escape(null == (jade_interp = "© " + new Date().getFullYear() + " ") ? "" : jade_interp) + '</span><span data-hook="name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</span></div><div class="line"></div></div><div class="logo-container"><div class="line"></div><a href="#" class="logo"><img src="/img/logo.png"/></a><div class="line"></div></div><h1 data-hook="name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h1><div class="contact-links"><a' + jade.attr("href", "http://maps.google.com?q=" + model.address, true, false) + ' data-hook="address" target="_blank" class="address">' + jade.escape(null == (jade_interp = model.address) ? "" : jade_interp) + "</a><a" + jade.attr("href", "tel:" + model.phone, true, false) + ' data-hook="phone" class="phone">' + jade.escape(null == (jade_interp = model.phone) ? "" : jade_interp) + "</a><a" + jade.attr("href", "mailto:" + model.email, true, false) + ' data-hook="email" class="email">' + jade.escape(null == (jade_interp = model.email) ? "" : jade_interp) + "</a><a" + jade.attr("href", model.website_url, true, false) + ' data-hook="web" target="_blank" class="web">' + jade.escape(null == (jade_interp = model.website_url) ? "" : jade_interp) + '</a></div><div class="social"><a href="http://www.facebook.com" data-hook="facebook" target="_blank" class="facebook"></a><a href="http://www.twitter.com" data-hook="twitter" target="_blank" class="twitter"></a><a href="http://www.linkedin.com" data-hook="linkedin" target="_blank" class="linkedin"></a><a href="http://www.instagram.com" data-hook="instagram" target="_blank" class="instagram"></a></div></div></div>');
        }).call(this, "Date" in locals_for_with ? locals_for_with.Date : typeof Date !== "undefined" ? Date : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // lightbox/buy-form.jade compiled template
    templatizer["lightbox"]["buy-form"] = function tmpl_lightbox_buy_form() {
        return '<div class="buy-form"><div class="options-wrapper"><div class="scroller"><div class="options"></div></div></div><div class="summary"><div class="total"><div class="make-default"> <div class="checkbox"> </div>Save this as my default set</div><div>Total: </div><div class="total-price"></div></div><div class="buttons"><div class="button cancel">Cancel</div><div class="right-buttons"><div class="button remove">Remove From Cart</div><div class="button save">Add To Cart</div></div></div></div></div>';
    };

    // lightbox/categories.jade compiled template
    templatizer["lightbox"]["categories"] = function tmpl_lightbox_categories(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(active, categories, undefined) {
            (function() {
                var $obj = categories;
                if ("number" == typeof $obj.length) {
                    for (var $index = 0, $l = $obj.length; $index < $l; $index++) {
                        var cat = $obj[$index];
                        buf.push("<a" + jade.attr("href", cat.app_url + "/" + cat.photos.at(0).id, true, false) + jade.cls([ cat === active ? "active" : "" ], [ true ]) + ">" + jade.escape(null == (jade_interp = cat.name) ? "" : jade_interp) + "</a>");
                    }
                } else {
                    var $l = 0;
                    for (var $index in $obj) {
                        $l++;
                        var cat = $obj[$index];
                        buf.push("<a" + jade.attr("href", cat.app_url + "/" + cat.photos.at(0).id, true, false) + jade.cls([ cat === active ? "active" : "" ], [ true ]) + ">" + jade.escape(null == (jade_interp = cat.name) ? "" : jade_interp) + "</a>");
                    }
                }
            }).call(this);
        }).call(this, "active" in locals_for_with ? locals_for_with.active : typeof active !== "undefined" ? active : undefined, "categories" in locals_for_with ? locals_for_with.categories : typeof categories !== "undefined" ? categories : undefined, "undefined" in locals_for_with ? locals_for_with.undefined : typeof undefined !== "undefined" ? undefined : undefined);
        return buf.join("");
    };

    // lightbox/photo.jade compiled template
    templatizer["lightbox"]["photo"] = function tmpl_lightbox_photo(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(Math, model, require) {
            var app = require("app");
            var nl = model.name.length;
            var titleClass = nl >= 35 ? "long-" + Math.min(10, Math.floor((nl - 25) / 10)) : "";
            buf.push('<div class="lightbox-photo"><div class="images"><div class="img"><div class="loader white"></div>');
            if (app.loadedImages[model.thumb_url]) {
                buf.push("<div" + jade.attr("style", "background-image:url(" + model.thumb_url + ")", true, false) + ' class="thumb"></div>');
            }
            buf.push("<img" + jade.attr("src", model.img_url, true, false) + '/><div class="borders"></div></div>');
            if (model.name) {
                buf.push("<div" + jade.cls([ "title", titleClass ], [ null, true ]) + '><div class="text">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + "</div></div>");
            }
            buf.push('</div><div class="actions"><div class="action cart"><div class="tip">Buy Image</div></div><div class="action fav"><div class="tip">Add To Favorites</div></div>');
            if (model.download) {
                buf.push("<a" + jade.attr("href", model.url, true, false) + ' target="_blank" bypass="bypass" class="action download"><div class="tip">Download</div></a>');
            }
            buf.push('<div class="action share"><div class="tip">Share</div></div><div class="action variants"><div class="tip">Variants</div></div></div></div>');
        }).call(this, "Math" in locals_for_with ? locals_for_with.Math : typeof Math !== "undefined" ? Math : undefined, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "require" in locals_for_with ? locals_for_with.require : typeof require !== "undefined" ? require : undefined);
        return buf.join("");
    };

    // lightbox/purchase-addon-option.jade compiled template
    templatizer["lightbox"]["purchase-addon-option"] = function tmpl_lightbox_purchase_addon_option(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model) {
            buf.push("<div" + jade.cls([ "purchase-addon-option", model.img_url ? "has-image" : "" ], [ null, true ]) + '><div class="text"><div class="title"><div class="checkbox"></div>' + jade.escape(null == (jade_interp = model.name + " (+" + model.formatMoney(model.price) + ")") ? "" : jade_interp) + '</div><div class="description">' + jade.escape(null == (jade_interp = model.description) ? "" : jade_interp) + "</div></div>");
            if (model.img_url) {
                buf.push("<div" + jade.attr("style", "background-image:url(" + model.img_url + ")", true, false) + ' class="image-container"></div>');
            }
            buf.push("</div>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined);
        return buf.join("");
    };

    // lightbox/purchase-addon-options.jade compiled template
    templatizer["lightbox"]["purchase-addon-options"] = function tmpl_lightbox_purchase_addon_options(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, undefined) {
            buf.push("<div" + jade.attr("cid", this.cid, true, false) + ' class="purchase-addon-options"><div class="header">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</div><div class="options"></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "undefined" in locals_for_with ? locals_for_with.undefined : typeof undefined !== "undefined" ? undefined : undefined);
        return buf.join("");
    };

    // lightbox/purchase-addon.jade compiled template
    templatizer["lightbox"]["purchase-addon"] = function tmpl_lightbox_purchase_addon(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, require) {
            var ua = require("app/utils/ua");
            var classes = [];
            if (model.options) classes.push("no-qty");
            buf.push("<div" + jade.cls([ "purchase-addon", classes ], [ null, true ]) + ">");
            if (model.options.length) {
                buf.push('<div class="expand"></div><div class="name">' + jade.escape(null == (jade_interp = model.default_option.name) ? "" : jade_interp) + "</div>");
            } else {
                buf.push('<div class="checkbox"></div><div class="name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + "</div>");
            }
            buf.push('<div class="price single">' + jade.escape(null == (jade_interp = model.display_price) ? "" : jade_interp) + "</div></div>");
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "require" in locals_for_with ? locals_for_with.require : typeof require !== "undefined" ? require : undefined);
        return buf.join("");
    };

    // lightbox/purchase-option.jade compiled template
    templatizer["lightbox"]["purchase-option"] = function tmpl_lightbox_purchase_option(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, require) {
            var ua = require("app/utils/ua");
            var classes = [];
            if (model.no_qty) classes.push("no-qty");
            buf.push("<div" + jade.cls([ "purchase-option", classes ], [ null, true ]) + '><div class="head"><div class="name">' + jade.escape(null == (jade_interp = model.display_name) ? "" : jade_interp));
            if (model.description) {
                buf.push('<div class="info"></div>');
            }
            buf.push('</div><div class="price single">' + jade.escape(null == (jade_interp = model.display_price) ? "" : jade_interp) + "</div>");
            if (!model.no_qty) {
                buf.push('<div class="qty"><input type="number" name="qty" value="0"/><div class="minus"></div><div class="plus"></div></div>');
            } else {
                buf.push('<div class="checkbox"></div>');
            }
            buf.push('<div class="price final">' + jade.escape(null == (jade_interp = model.formatMoney(0)) ? "" : jade_interp) + '</div></div><div class="addons"></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "require" in locals_for_with ? locals_for_with.require : typeof require !== "undefined" ? require : undefined);
        return buf.join("");
    };

    // nav.jade compiled template
    templatizer["nav"] = function tmpl_nav() {
        return '<nav><div class="content"><div class="body"><div class="right"><a href="#" class="search"><span>Search Events</span></a><a href="#favorites" class="favs"><span>Favorites</span></a><a href="#cart" class="cart"><span>Cart</span></a><a href="#" class="menu"><span>Menu</span></a></div><a href="#" class="logo"><img src="/img/logo.png"/></a></div></div></nav>';
    };

    // pages/event.jade compiled template
    templatizer["pages"]["event"] = function tmpl_pages_event(locals) {
        var buf = [];
        var jade_mixins = {};
        var jade_interp;
        var locals_for_with = locals || {};
        (function(model, pageType) {
            buf.push("<div" + jade.cls([ "page", "event", "inactive", pageType === "event-category" ? "category" : "" ], [ null, null, null, true ]) + '><div class="page-scroller"><div class="page-scroll"><div class="page-content"><div class="content"> ');
            if (model) {
                buf.push('<div class="header"><h1 data-hook="event-name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</h1><div class="page-path"> <span class="wide-only">You are here:</span><a href="#">Home</a><span>&gt;</span>');
                if (pageType === "event") {
                    buf.push('<span data-hook="event-name">' + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + "</span>");
                } else {
                    buf.push('<a data-hook="event-name"' + jade.attr("href", "#events/" + model.id, true, false) + ">" + jade.escape(null == (jade_interp = model.name) ? "" : jade_interp) + '</a><span>&gt;</span><span data-hook="category-name"></span>');
                }
                buf.push('</div></div><div class="events-group"><div class="categories group-events"></div></div>');
            }
            buf.push('</div></div><div class="page-footer"> </div></div></div></div>');
        }).call(this, "model" in locals_for_with ? locals_for_with.model : typeof model !== "undefined" ? model : undefined, "pageType" in locals_for_with ? locals_for_with.pageType : typeof pageType !== "undefined" ? pageType : undefined);
        return buf.join("");
    };

    // pages/events.jade compiled template
    templatizer["pages"]["events"] = function tmpl_pages_events() {
        return '<div class="page events complete reloading inactive"><div class="page-scroller"><div class="page-scroll"><div class="page-content"><div class="content"> <div class="header"><div class="filters"><div class="inner"><a href="#" data-filter="showcased" class="showcased">Showcased<i> Events</i></a><a href="#" data-filter="current" class="current">Current</a><a href="#" data-filter="upcoming" class="upcoming">Upcoming</a><a href="#" data-filter="all" class="all active">All<i class="all"> Events</i></a></div></div><div class="sorting"><div class="inner"><span>Sort events by:</span><a href="#" data-sort="name" class="name"><i>Event </i>Name</a><a href="#" data-sort="date" class="date active"><i>Event </i>Date</a><a href="#" data-sort="type" class="type"><i>Event </i>Type</a></div></div><div class="clear"></div><div class="label"><div class="icon"></div><div class="loader grey-bg"></div></div></div><div class="events-list"></div><div class="load-more"><div class="label">Load More Events</div><div class="loader"></div></div></div></div><div class="page-footer"></div></div></div></div>';
    };

    // popup.jade compiled template
    templatizer["popup"] = function tmpl_popup() {
        return '<div class="popup"><div class="close"></div><div class="contents"><div class="title"></div><div class="image"></div><div class="text"></div></div></div>';
    };

    return templatizer;
}));
