// app/cart/discount-code-model.js
"use strict";

var BaseModel = require('app/base/base-model');

var app = require('app');

module.exports = BaseModel.extend({
  props: {
    discount_code: ['string', true, ''],
    email: ['string', true, ''],
    valid: ['boolean', true, false],
    message: 'any',
    type: 'any',
    remaining: ['number', true, ''],
    tiers: ['array', true],
    include_products: ['boolean', true, false],
    pay_for_tax: ['boolean', true, false],
    apply_pre_tax_calc: ['boolean', true, false],
  },
  
  session: {
    loading: ['boolean', true, false],
  },
  
  derived: {
    event: {
      fn: function () {
        return this.parent.event;
      },
    },
  },
  
  url: function () {
    return app.apiUrl +'cart/code/'+ this.event.id +'/'+ 
        encodeURIComponent(this.discount_code)+
        (this.email ? '/'+ encodeURIComponent(this.email) : '')+
        '?url='+ app.url;
  },
  
  initialize: function () {
    BaseModel.prototype.initialize.apply(this, arguments);
    
    this.listenTo(this, 'request', function () {
      this.loading = true;
    });
    this.listenTo(this, 'sync', function () {
      this.loading = false;
      console.log('promo sync', this.toJSON());
    });
    
    this.listenTo(this, 'error', function () {
      this.loading = false;
      console.log('promo sync error', this.toJSON());
    });
  },
  
  parse: function (data) {
    console.log('parse', data);
    return data.data;
  },
});
