// app/cart/photo/addon/cart-addon-model.js
/*==========================================================================*\
  A photo's purchase addon in cart
\*==========================================================================*/
"use strict";

var BaseModel = require('app/base/base-model');

var app = require('app');

module.exports = BaseModel.extend({
  // idAttribute: 'id',
  props: {
    id: 'number', // ID of the addon
    oid: 'number', // selected option ID
  },
  derived: {
    cart: {
      fn: function () {
        return this.collection.parent.cart;
      },
    },
    addon: {
      deps: ['id'],
      fn: function () {
        return app.addons.get(this.id);
      },
    },
    price: {
      cache: false,
      // deps: ['oid', 'id'],
      fn: function () {
        var sizeId = this.collection.parent.size_id;
        var price;
        if (!this.addon) { console.log('no addon', this); return 0; }
        if (this.addon.options.length <= 1) {
          if (this.addon.options.length) {
            price = this.addon.options.at(0).getPrice(sizeId);
          }
          else {
            price = this.addon.price;
          }
        }
        else {
          price = this.oid ? 
                  this.addon.options.get(this.oid).getPrice(sizeId) : 0;
        }
        
        var e = this.cart.event;
        var discount = e.sales.allProducts;
        var discountAmount = (discount && discount.discount) 
                             ? (discount.discount * price) : 0;
        if (discountAmount) {
          price = Math.round(price - discountAmount);
        }
        
        return price;
        // return this.oid ? this.addon.options.get(this.oid).price 
        //                 : this.addon.price;
      },
    },
    shipping_price: {
      // deps: ['oid', 'id'],
      cache: false,
      fn: function () {
        var option = this.addon.options.get(this.oid);
        var size = this.collection.parent;
        var shipId = this.cart.shipping_type;
        return option.sizes[size.size_id].shipping.get(shipId).price;
      },
    },
  },
});
